From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: TheMeinerLP <p.glanz@madfix.me>
Date: Tue, 13 Jun 2023 22:50:57 +0200
Subject: [PATCH] Fix gradle submodule for patched version


diff --git a/build.gradle.kts b/build.gradle.kts
index 32f8a037754849d3ad31e4cab64b5f9ff57e9ce9..c102b750aded543fe7a712775922a181aaaa1169 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -3,6 +3,28 @@ plugins {
     id("minestom.publishing-conventions")
     id("minestom.native-conventions")
     alias(libs.plugins.blossom)
+    // publishing
+    id("org.ajoberstar.grgit") version "5.2.0"
+    signing
+}
+
+var baseVersion by extra("1.1.1")
+var versionExtension by extra("")
+var snapshot by extra("")
+
+group = "net.onelitefeather.microtus"
+
+
+ext {
+    val git: Grgit = Grgit.open {
+        dir = File("$rootDir/.git")
+    }
+    val revision = git.head().abbreviatedId
+    if (snapshot.isNotEmpty()) {
+        versionExtension = "%s+%s".format(Locale.ROOT, snapshot, revision)
+    } else {
+        versionExtension = ""
+    }
 }
 
 allprojects {
diff --git a/demo/build.gradle.kts b/demo/build.gradle.kts
index 60b18683dd01e7f349f7f22465db2bde969c9690..97f211b6509ee687978c78cf58480b8b214f9e0f 100644
--- a/demo/build.gradle.kts
+++ b/demo/build.gradle.kts
@@ -4,14 +4,12 @@ plugins {
     application
     id("minestom.common-conventions")
     id("minestom.native-conventions")
-    id("com.github.johnrengelman.shadow") version ("7.1.1")
+    id("com.github.johnrengelman.shadow") version ("8.1.1")
 }
 
 application {
     mainClass.set("net.minestom.demo.Main")
     // This is included because Shadow is buggy. Wait for https://github.com/johnrengelman/shadow/issues/613 to befixed.
-    @Suppress("DEPRECATION")
-    mainClassName = "net.minestom.demo.Main"
 }
 
 dependencies {
diff --git a/jcstress-tests/build.gradle.kts b/jcstress-tests/build.gradle.kts
index dc62cdcae7dc1e23bc9d72303673edf77f607422..0ffb43098ea0772c4daa84aa5adce400d417508e 100644
--- a/jcstress-tests/build.gradle.kts
+++ b/jcstress-tests/build.gradle.kts
@@ -1,13 +1,15 @@
 plugins {
-    id("io.github.reyerizo.gradle.jcstress") version "0.8.13"
+    // id("io.github.reyerizo.gradle.jcstress") version "0.8.13"
     id("minestom.common-conventions")
 }
 
 dependencies {
-    jcstressImplementation(rootProject)
-    jcstress(libs.jcstress.core)
+/*    jcstressImplementation(rootProject)
+    jcstress(libs.jcstress.core)*/
 }
 
+/*
 jcstress {
     verbose = "true"
 }
+*/
diff --git a/settings.gradle.kts b/settings.gradle.kts
index 2230bb02892d10d6756722105337ae6434cc8a35..3fb63e467ef2d6fe82c96f53718e5b8292c36ad4 100644
--- a/settings.gradle.kts
+++ b/settings.gradle.kts
@@ -1,5 +1,4 @@
 enableFeaturePreview("TYPESAFE_PROJECT_ACCESSORS")
-enableFeaturePreview("VERSION_CATALOGS")
 
 dependencyResolutionManagement {
     repositories {
