From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Phillipp Glanz <p.glanz@madfix.me>
Date: Thu, 30 Nov 2023 12:38:07 +0100
Subject: [PATCH] Improve event handling


diff --git a/src/main/java/net/minestom/server/entity/Player.java b/src/main/java/net/minestom/server/entity/Player.java
index 33f4fe1d6b1af922d294924d58acfa1eeb3036d1..60cd81d8e7354b0813188b60b923bcf61aea142a 100644
--- a/src/main/java/net/minestom/server/entity/Player.java
+++ b/src/main/java/net/minestom/server/entity/Player.java
@@ -30,6 +30,7 @@ import net.minestom.server.entity.fakeplayer.FakePlayer;
 import net.minestom.server.entity.metadata.PlayerMeta;
 import net.minestom.server.entity.vehicle.PlayerVehicleInformation;
 import net.minestom.server.event.EventDispatcher;
+import net.minestom.server.event.inventory.InventoryCloseEvent;
 import net.minestom.server.event.inventory.InventoryOpenEvent;
 import net.minestom.server.event.item.ItemDropEvent;
 import net.minestom.server.event.item.ItemUpdateStateEvent;
@@ -1305,13 +1306,22 @@ public class Player extends LivingEntity implements CommandSender, Localizable,
         return gameMode;
     }
 
+    // Microtus start - Add gamemode change event
     /**
      * Changes the player {@link GameMode}
      *
      * @param gameMode the new player GameMode
-     */
-    public void setGameMode(@NotNull GameMode gameMode) {
-        this.gameMode = gameMode;
+     * @return true if the gamemode was changed successfully, false otherwise (cancelled by event)
+     */
+    public boolean setGameMode(@NotNull GameMode gameMode) {
+        PlayerGameModeChangeEvent playerGameModeChangeEvent = new PlayerGameModeChangeEvent(this, gameMode);
+        EventDispatcher.call(playerGameModeChangeEvent);
+        if (playerGameModeChangeEvent.isCancelled()) {
+            // Abort
+            return false;
+        }
+        this.gameMode = playerGameModeChangeEvent.getNewGameMode();
+        // Microtus end - Add gamemode change event
         // Condition to prevent sending the packets before spawning the player
         if (isActive()) {
             sendPacket(new ChangeGameStatePacket(ChangeGameStatePacket.Reason.CHANGE_GAMEMODE, gameMode.id()));
@@ -1342,8 +1352,10 @@ public class Player extends LivingEntity implements CommandSender, Localizable,
         if (isActive()) {
             refreshAbilities();
         }
+        return true; // Microtus - Add gamemode change event
     }
 
+
     /**
      * Gets if this player is in creative. Used for code readability.
      *
@@ -1500,6 +1512,10 @@ public class Player extends LivingEntity implements CommandSender, Localizable,
         }
 
         if (openInventory == getOpenInventory()) {
+            // Microtus start - 1.19.3 trigger inventory close
+            InventoryCloseEvent closeEvent = new InventoryCloseEvent(openInventory, this);
+            EventDispatcher.call(closeEvent);
+            // Microtus end - 1.19.3 trigger inventory close
             CloseWindowPacket closeWindowPacket;
             if (openInventory == null) {
                 closeWindowPacket = new CloseWindowPacket((byte) 0);
diff --git a/src/main/java/net/minestom/server/event/player/PlayerGameModeChangeEvent.java b/src/main/java/net/minestom/server/event/player/PlayerGameModeChangeEvent.java
new file mode 100644
index 0000000000000000000000000000000000000000..8ced8497feac0a3aa2b625ab7606567d671055c2
--- /dev/null
+++ b/src/main/java/net/minestom/server/event/player/PlayerGameModeChangeEvent.java
@@ -0,0 +1,55 @@
+package net.minestom.server.event.player;
+
+import net.minestom.server.entity.GameMode;
+import net.minestom.server.entity.Player;
+import net.minestom.server.event.trait.CancellableEvent;
+import net.minestom.server.event.trait.PlayerInstanceEvent;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * Called when the gamemode of a player is being modified.
+ */
+public class PlayerGameModeChangeEvent implements PlayerInstanceEvent, CancellableEvent {
+    private final Player player;
+    private GameMode newGameMode;
+
+    private boolean cancelled;
+
+    public PlayerGameModeChangeEvent(@NotNull Player player, @NotNull GameMode newGameMode) {
+        this.player = player;
+        this.newGameMode = newGameMode;
+    }
+
+    /**
+     * Gets the target gamemode.
+     *
+     * @return the target gamemode
+     */
+    public @NotNull GameMode getNewGameMode() {
+        return newGameMode;
+    }
+
+    /**
+     * Changes the target gamemode.
+     *
+     * @param newGameMode the new target gamemode
+     */
+    public void setNewGameMode(@NotNull GameMode newGameMode) {
+        this.newGameMode = newGameMode;
+    }
+
+    @Override
+    public boolean isCancelled() {
+        return cancelled;
+    }
+
+    @Override
+    public void setCancelled(boolean cancel) {
+        this.cancelled = cancel;
+    }
+
+    @Override
+    public @NotNull Player getPlayer() {
+        return player;
+    }
+}
diff --git a/src/test/java/net/minestom/server/entity/player/PlayerIntegrationTest.java b/src/test/java/net/minestom/server/entity/player/PlayerIntegrationTest.java
index bd9b1ef7cb79058fd74299b3be653821fca6a064..da0bc9a7a883153117b069041305b95be35b8a89 100644
--- a/src/test/java/net/minestom/server/entity/player/PlayerIntegrationTest.java
+++ b/src/test/java/net/minestom/server/entity/player/PlayerIntegrationTest.java
@@ -34,17 +34,41 @@ public class PlayerIntegrationTest {
         var connection = env.createConnection();
         var player = connection.connect(instance, new Pos(0, 42, 0)).join();
         assertEquals(instance, player.getInstance());
+        // Microtus start - Add gamemode change event
+
+        // Abilities
+        {
+            player.setGameMode(GameMode.CREATIVE);
+            assertAbilities(player, true, false, true, true);
+            player.setGameMode(GameMode.SPECTATOR);
+            assertAbilities(player, true, true, true, false);
+            player.setGameMode(GameMode.CREATIVE);
+            assertAbilities(player, true, true, true, true);
+            player.setGameMode(GameMode.ADVENTURE);
+            assertAbilities(player, false, false, false, false);
+            player.setGameMode(GameMode.SURVIVAL);
+            assertAbilities(player, false, false, false, false);
+        }
 
-        player.setGameMode(GameMode.CREATIVE);
-        assertAbilities(player, true, false, true, true);
-        player.setGameMode(GameMode.SPECTATOR);
-        assertAbilities(player, true, true, true, false);
-        player.setGameMode(GameMode.CREATIVE);
-        assertAbilities(player, true, true, true, true);
-        player.setGameMode(GameMode.ADVENTURE);
-        assertAbilities(player, false, false, false, false);
-        player.setGameMode(GameMode.SURVIVAL);
-        assertAbilities(player, false, false, false, false);
+        var listener = env.listen(net.minestom.server.event.player.PlayerGameModeChangeEvent.class);
+        // Normal change
+        {
+            listener.followup();
+            assertTrue(player.setGameMode(GameMode.ADVENTURE));
+        }
+        // Change target gamemode event
+        {
+            listener.followup(event -> event.setNewGameMode(GameMode.SPECTATOR));
+            assertTrue(player.setGameMode(GameMode.CREATIVE));
+            assertEquals(GameMode.SPECTATOR, player.getGameMode());
+        }
+        // Cancel event
+        {
+            listener.followup(event -> event.setCancelled(true));
+            assertFalse(player.setGameMode(GameMode.CREATIVE));
+            assertEquals(GameMode.SPECTATOR, player.getGameMode());
+        }
+        // Microtus end - Add gamemode change event
     }
 
     @Test
