From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: TheMeinerLP <p.glanz@madfix.me>
Date: Wed, 14 Jun 2023 09:06:38 +0200
Subject: [PATCH] Add better differentiation of shift click. Add left or right


diff --git a/src/main/java/net/minestom/server/inventory/Inventory.java b/src/main/java/net/minestom/server/inventory/Inventory.java
index be9c688fa6192c6b190a96e74ffd9bc7bbe6e3ba..9d80f6331c1320603680411d1eff3d710c38e6c3 100644
--- a/src/main/java/net/minestom/server/inventory/Inventory.java
+++ b/src/main/java/net/minestom/server/inventory/Inventory.java
@@ -258,7 +258,7 @@ public non-sealed class Inventory extends AbstractInventory implements Viewable
     }
 
     @Override
-    public boolean shiftClick(@NotNull Player player, int slot) {
+    public boolean shiftClick(@NotNull Player player, int slot, int button) { // Microtus
         final PlayerInventory playerInventory = player.getInventory();
         final boolean isInWindow = isClickInWindow(slot);
         final int clickSlot = isInWindow ? slot : PlayerInventoryUtils.convertSlot(slot, offset);
@@ -268,7 +268,7 @@ public non-sealed class Inventory extends AbstractInventory implements Viewable
                 isInWindow ? this : playerInventory,
                 isInWindow ? playerInventory : this,
                 0, isInWindow ? playerInventory.getInnerSize() : getInnerSize(), 1,
-                player, clickSlot, clicked, cursor);
+                player, clickSlot, clicked, cursor, button); // Microtus
         if (clickResult.isCancel()) {
             updateAll(player);
             return false;
diff --git a/src/main/java/net/minestom/server/inventory/InventoryClickHandler.java b/src/main/java/net/minestom/server/inventory/InventoryClickHandler.java
index 5fa175416eb75cba7a6f945fd3ed9dd6cae1ce88..550f50ae2fd72f71b4601b19b6a3b45712f0a62e 100644
--- a/src/main/java/net/minestom/server/inventory/InventoryClickHandler.java
+++ b/src/main/java/net/minestom/server/inventory/InventoryClickHandler.java
@@ -38,9 +38,10 @@ public sealed interface InventoryClickHandler permits AbstractInventory {
      *
      * @param player the player who clicked
      * @param slot   the slot number
+     * @param button the primary mouse or secondary mouse button
      * @return true if the click hasn't been cancelled, false otherwise
      */
-    boolean shiftClick(@NotNull Player player, int slot); // shift + left/right click have the same behavior
+    boolean shiftClick(@NotNull Player player, int slot, int button); // shift + left/right click have the same behavior // Microtus
 
     /**
      * Called when a {@link Player} held click in the inventory
diff --git a/src/main/java/net/minestom/server/inventory/PlayerInventory.java b/src/main/java/net/minestom/server/inventory/PlayerInventory.java
index bb2158419ccf7d18e4bdb99cbcbb134d15446166..40c66742f437ad88e333c87965162b3f94c7ef3a 100644
--- a/src/main/java/net/minestom/server/inventory/PlayerInventory.java
+++ b/src/main/java/net/minestom/server/inventory/PlayerInventory.java
@@ -243,7 +243,7 @@ public non-sealed class PlayerInventory extends AbstractInventory implements Equ
     }
 
     @Override
-    public boolean shiftClick(@NotNull Player player, int slot) {
+    public boolean shiftClick(@NotNull Player player, int slot, int button) { // Microtus
         final int convertedSlot = convertPlayerInventorySlot(slot, OFFSET);
         final ItemStack cursor = getCursorItem();
         final ItemStack clicked = getItemStack(convertedSlot);
@@ -253,7 +253,7 @@ public non-sealed class PlayerInventory extends AbstractInventory implements Equ
         final InventoryClickResult clickResult = clickProcessor.shiftClick(
                 this, this,
                 start, end, 1,
-                player, convertedSlot, clicked, cursor);
+                player, convertedSlot, clicked, cursor, button); // Microtus
         if (clickResult.isCancel()) {
             update();
             return false;
diff --git a/src/main/java/net/minestom/server/inventory/click/ClickType.java b/src/main/java/net/minestom/server/inventory/click/ClickType.java
index c6fe8eda8d551a023a8843f6e332b3c3cd4cd618..bf5600760e409ef9a0d1e952cb859e0e39f1521b 100644
--- a/src/main/java/net/minestom/server/inventory/click/ClickType.java
+++ b/src/main/java/net/minestom/server/inventory/click/ClickType.java
@@ -6,8 +6,10 @@ public enum ClickType {
     RIGHT_CLICK,
     CHANGE_HELD,
 
-    START_SHIFT_CLICK,
-    SHIFT_CLICK,
+    START_SHIFT_LEFT_CLICK, // Microtus
+    START_SHIFT_RIGHT_CLICK, // Microtus
+    SHIFT_LEFT_CLICK, // Microtus
+    SHIFT_RIGHT_CLICK, // Microtus
 
     START_LEFT_DRAGGING,
     START_RIGHT_DRAGGING,
diff --git a/src/main/java/net/minestom/server/inventory/click/InventoryClickProcessor.java b/src/main/java/net/minestom/server/inventory/click/InventoryClickProcessor.java
index 634a9ef2aa4e1a05f7ed28f756db12ff3d24077b..ead4ef6fed576403e9368f23ffe393668a505ce1 100644
--- a/src/main/java/net/minestom/server/inventory/click/InventoryClickProcessor.java
+++ b/src/main/java/net/minestom/server/inventory/click/InventoryClickProcessor.java
@@ -119,10 +119,12 @@ public final class InventoryClickProcessor {
     public @NotNull InventoryClickResult shiftClick(@NotNull AbstractInventory inventory, @NotNull AbstractInventory targetInventory,
                                                     int start, int end, int step,
                                                     @NotNull Player player, int slot,
-                                                    @NotNull ItemStack clicked, @NotNull ItemStack cursor) {
-        InventoryClickResult clickResult = startCondition(player, inventory, slot, ClickType.START_SHIFT_CLICK, clicked, cursor);
+                                                    @NotNull ItemStack clicked, @NotNull ItemStack cursor, int button) { // Microtus
+        ClickType startShiftClick = button == 0 ? ClickType.START_SHIFT_LEFT_CLICK : ClickType.START_SHIFT_RIGHT_CLICK; // Microtus
+        InventoryClickResult clickResult = startCondition(player, inventory, slot, startShiftClick, clicked, cursor); // Microtus
         if (clickResult.isCancel()) return clickResult;
         if (clicked.isAir()) return clickResult.cancelled();
+        ClickType shiftClick = button == 0 ? ClickType.SHIFT_LEFT_CLICK : ClickType.SHIFT_RIGHT_CLICK; // Microtus
 
         // Handle armor equip
         if (inventory instanceof PlayerInventory && targetInventory instanceof PlayerInventory) {
@@ -133,7 +135,7 @@ public final class InventoryClickProcessor {
                 final ItemStack currentArmor = player.getEquipment(equipmentSlot);
                 if (currentArmor.isAir()) {
                     final int armorSlot = equipmentSlot.armorSlot();
-                    InventoryClickResult result = startCondition(player, targetInventory, armorSlot, ClickType.SHIFT_CLICK, clicked, cursor);
+                    InventoryClickResult result = startCondition(player, targetInventory, armorSlot, shiftClick, clicked, cursor); // Microtus
                     if (result.isCancel()) return clickResult;
                     result.setClicked(ItemStack.AIR);
                     result.setCursor(cursor);
@@ -147,7 +149,7 @@ public final class InventoryClickProcessor {
         final var pair = TransactionType.ADD.process(targetInventory, clicked, (index, itemStack) -> {
             if (inventory == targetInventory && index == slot)
                 return false; // Prevent item lose/duplication
-            InventoryClickResult result = startCondition(player, targetInventory, index, ClickType.SHIFT_CLICK, itemStack, cursor);
+            InventoryClickResult result = startCondition(player, targetInventory, index, shiftClick, itemStack, cursor); // Microtus
             if (result.isCancel()) {
                 return false;
             }
@@ -159,7 +161,7 @@ public final class InventoryClickProcessor {
         final Map<Integer, ItemStack> itemChangesMap = pair.right();
         itemChangesMap.forEach((Integer s, ItemStack itemStack) -> {
             targetInventory.setItemStack(s, itemStack);
-            callClickEvent(player, targetInventory, s, ClickType.SHIFT_CLICK, itemStack, cursor);
+            callClickEvent(player, targetInventory, s, shiftClick, itemStack, cursor); // Microtus
         });
         clickResult.setClicked(itemResult);
         return clickResult;
diff --git a/src/main/java/net/minestom/server/listener/WindowListener.java b/src/main/java/net/minestom/server/listener/WindowListener.java
index dfdf2e6751fc14b110b4c1a9f293bfe26ee3d9f0..10b139213594c420b01a65fe731396303b81da65 100644
--- a/src/main/java/net/minestom/server/listener/WindowListener.java
+++ b/src/main/java/net/minestom/server/listener/WindowListener.java
@@ -48,7 +48,7 @@ public class WindowListener {
                 }
             }
         } else if (clickType == ClientClickWindowPacket.ClickType.QUICK_MOVE) {
-            successful = inventory.shiftClick(player, slot);
+            successful = inventory.shiftClick(player, slot, button); // Microtus
         } else if (clickType == ClientClickWindowPacket.ClickType.SWAP) {
             successful = inventory.changeHeld(player, slot, button);
         } else if (clickType == ClientClickWindowPacket.ClickType.CLONE) {
