From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: TheMeinerLP <p.glanz@madfix.me>
Date: Wed, 14 Jun 2023 13:34:16 +0200
Subject: [PATCH] Add 1.19.3 recipe categories


diff --git a/src/main/java/net/minestom/server/network/packet/server/play/DeclareRecipesPacket.java b/src/main/java/net/minestom/server/network/packet/server/play/DeclareRecipesPacket.java
index 33ff8b72862496bf5a5c41216bb4ad947d4f5dd6..a3592735becafe263e41aa55de5d563cf34ebf48 100644
--- a/src/main/java/net/minestom/server/network/packet/server/play/DeclareRecipesPacket.java
+++ b/src/main/java/net/minestom/server/network/packet/server/play/DeclareRecipesPacket.java
@@ -4,6 +4,8 @@ import net.minestom.server.item.ItemStack;
 import net.minestom.server.network.NetworkBuffer;
 import net.minestom.server.network.packet.server.ServerPacket;
 import net.minestom.server.network.packet.server.ServerPacketIdentifier;
+import net.minestom.server.recipe.CookingRecipeCategory;
+import net.minestom.server.recipe.CraftingRecipeCategory;
 import org.jetbrains.annotations.NotNull;
 
 import java.util.ArrayList;
@@ -57,16 +59,18 @@ public record DeclareRecipesPacket(@NotNull List<DeclaredRecipe> recipes) implem
     }
 
     public record DeclaredShapelessCraftingRecipe(String recipeId, String group,
+                                                  CraftingRecipeCategory category, // Microtus - RecipeCategories
                                                   List<Ingredient> ingredients,
                                                   ItemStack result) implements DeclaredRecipe {
         private DeclaredShapelessCraftingRecipe(@NotNull NetworkBuffer reader) {
-            this(reader.read(STRING), reader.read(STRING),
+            this(reader.read(STRING), reader.read(STRING), reader.readEnum(CraftingRecipeCategory.class), // Microtus - RecipeCategories
                     reader.readCollection(Ingredient::new), reader.read(ITEM));
         }
 
         @Override
         public void write(@NotNull NetworkBuffer writer) {
             writer.write(STRING, group);
+            writer.writeEnum(CraftingRecipeCategory.class, category); // Microtus - RecipeCategories
             writer.writeCollection(ingredients);
             writer.write(ITEM, result);
         }
@@ -78,14 +82,16 @@ public record DeclareRecipesPacket(@NotNull List<DeclaredRecipe> recipes) implem
     }
 
     public record DeclaredShapedCraftingRecipe(@NotNull String recipeId, int width, int height,
-                                               @NotNull String group, @NotNull List<Ingredient> ingredients,
+                                               @NotNull String group,
+                                               CraftingRecipeCategory category, // Microtus - RecipeCategories
+                                               @NotNull List<Ingredient> ingredients,
                                                @NotNull ItemStack result) implements DeclaredRecipe {
         public DeclaredShapedCraftingRecipe {
             ingredients = List.copyOf(ingredients);
         }
 
         private DeclaredShapedCraftingRecipe(DeclaredShapedCraftingRecipe packet) {
-            this(packet.recipeId, packet.width, packet.height, packet.group, packet.ingredients, packet.result);
+            this(packet.recipeId, packet.width, packet.height, packet.group, packet.category, packet.ingredients, packet.result); // Microtus - RecipeCategories
         }
 
         public DeclaredShapedCraftingRecipe(@NotNull NetworkBuffer reader) {
@@ -98,12 +104,13 @@ public record DeclareRecipesPacket(@NotNull List<DeclaredRecipe> recipes) implem
             int width = reader.read(VAR_INT);
             int height = reader.read(VAR_INT);
             String group = reader.read(STRING);
+            CraftingRecipeCategory category = reader.readEnum(CraftingRecipeCategory.class); // Microtus - RecipeCategories
             List<Ingredient> ingredients = new ArrayList<>();
             for (int slot = 0; slot < width * height; slot++) {
                 ingredients.add(new Ingredient(reader));
             }
             ItemStack result = reader.read(ITEM);
-            return new DeclaredShapedCraftingRecipe(recipeId, width, height, group, ingredients, result);
+            return new DeclaredShapedCraftingRecipe(recipeId, width, height, group, category, ingredients, result); // Microtus - RecipeCategories
         }
 
         @Override
@@ -111,6 +118,7 @@ public record DeclareRecipesPacket(@NotNull List<DeclaredRecipe> recipes) implem
             writer.write(VAR_INT, width);
             writer.write(VAR_INT, height);
             writer.write(STRING, group);
+            writer.writeEnum(CraftingRecipeCategory.class, category); // Microtus - RecipeCategories
             for (Ingredient ingredient : ingredients) {
                 ingredient.write(writer);
             }
@@ -124,17 +132,20 @@ public record DeclareRecipesPacket(@NotNull List<DeclaredRecipe> recipes) implem
     }
 
     public record DeclaredSmeltingRecipe(@NotNull String recipeId, @NotNull String group,
-                                         @NotNull Ingredient ingredient, @NotNull ItemStack result,
+                                         @NotNull Ingredient ingredient,
+                                         @NotNull CookingRecipeCategory category, // Microtus - RecipeCategories
+                                         @NotNull ItemStack result,
                                          float experience, int cookingTime) implements DeclaredRecipe {
         public DeclaredSmeltingRecipe(@NotNull NetworkBuffer reader) {
             this(reader.read(STRING), reader.read(STRING),
-                    new Ingredient(reader), reader.read(ITEM),
+                    new Ingredient(reader), reader.readEnum(CookingRecipeCategory.class), reader.read(ITEM), // Microtus - RecipeCategories
                     reader.read(FLOAT), reader.read(VAR_INT));
         }
 
         @Override
         public void write(@NotNull NetworkBuffer writer) {
             writer.write(STRING, group);
+            writer.writeEnum(CookingRecipeCategory.class, category); // Microtus - RecipeCategories
             writer.write(ingredient);
             writer.write(ITEM, result);
             writer.write(FLOAT, experience);
@@ -148,17 +159,20 @@ public record DeclareRecipesPacket(@NotNull List<DeclaredRecipe> recipes) implem
     }
 
     public record DeclaredBlastingRecipe(@NotNull String recipeId, @NotNull String group,
-                                         @NotNull Ingredient ingredient, @NotNull ItemStack result,
+                                         @NotNull Ingredient ingredient,
+                                         @NotNull CookingRecipeCategory category, // Microtus - RecipeCategories
+                                         @NotNull ItemStack result,
                                          float experience, int cookingTime) implements DeclaredRecipe {
         public DeclaredBlastingRecipe(@NotNull NetworkBuffer reader) {
             this(reader.read(STRING), reader.read(STRING),
-                    new Ingredient(reader), reader.read(ITEM),
+                    new Ingredient(reader), reader.readEnum(CookingRecipeCategory.class), reader.read(ITEM), // Microtus - RecipeCategories
                     reader.read(FLOAT), reader.read(VAR_INT));
         }
 
         @Override
         public void write(@NotNull NetworkBuffer writer) {
             writer.write(STRING, group);
+            writer.writeEnum(CookingRecipeCategory.class, category); // Microtus - RecipeCategories
             writer.write(ingredient);
             writer.write(ITEM, result);
             writer.write(FLOAT, experience);
@@ -172,17 +186,20 @@ public record DeclareRecipesPacket(@NotNull List<DeclaredRecipe> recipes) implem
     }
 
     public record DeclaredSmokingRecipe(@NotNull String recipeId, @NotNull String group,
-                                        @NotNull Ingredient ingredient, @NotNull ItemStack result,
+                                        @NotNull Ingredient ingredient,
+                                        @NotNull CookingRecipeCategory category, // Microtus - RecipeCategories
+                                        @NotNull ItemStack result,
                                         float experience, int cookingTime) implements DeclaredRecipe {
         public DeclaredSmokingRecipe(@NotNull NetworkBuffer reader) {
             this(reader.read(STRING), reader.read(STRING),
-                    new Ingredient(reader), reader.read(ITEM),
+                    new Ingredient(reader), reader.readEnum(CookingRecipeCategory.class), reader.read(ITEM), // Microtus - RecipeCategories
                     reader.read(FLOAT), reader.read(VAR_INT));
         }
 
         @Override
         public void write(@NotNull NetworkBuffer writer) {
             writer.write(STRING, group);
+            writer.writeEnum(CookingRecipeCategory.class, category); // Microtus - RecipeCategories
             writer.write(ingredient);
             writer.write(ITEM, result);
             writer.write(FLOAT, experience);
@@ -196,17 +213,20 @@ public record DeclareRecipesPacket(@NotNull List<DeclaredRecipe> recipes) implem
     }
 
     public record DeclaredCampfireCookingRecipe(@NotNull String recipeId, @NotNull String group,
-                                                @NotNull Ingredient ingredient, @NotNull ItemStack result,
+                                                @NotNull Ingredient ingredient,
+                                                @NotNull CookingRecipeCategory category, // Microtus - RecipeCategories
+                                                @NotNull ItemStack result,
                                                 float experience, int cookingTime) implements DeclaredRecipe {
         public DeclaredCampfireCookingRecipe(@NotNull NetworkBuffer reader) {
             this(reader.read(STRING), reader.read(STRING),
-                    new Ingredient(reader), reader.read(ITEM),
+                    new Ingredient(reader), reader.readEnum(CookingRecipeCategory.class), reader.read(ITEM), // Microtus - RecipeCategories
                     reader.read(FLOAT), reader.read(VAR_INT));
         }
 
         @Override
         public void write(@NotNull NetworkBuffer writer) {
             writer.write(STRING, group);
+            writer.writeEnum(CookingRecipeCategory.class, category); // Microtus - RecipeCategories
             writer.write(ingredient);
             writer.write(ITEM, result);
             writer.write(FLOAT, experience);
diff --git a/src/main/java/net/minestom/server/recipe/BlastingRecipe.java b/src/main/java/net/minestom/server/recipe/BlastingRecipe.java
index 6a6bd4ae7670421d59c123b6fd74652dea6e9805..0079416f97b251b4f6491e3010e7a0bc5ae5c08a 100644
--- a/src/main/java/net/minestom/server/recipe/BlastingRecipe.java
+++ b/src/main/java/net/minestom/server/recipe/BlastingRecipe.java
@@ -6,6 +6,7 @@ import org.jetbrains.annotations.NotNull;
 
 public abstract class BlastingRecipe extends Recipe {
     private String group;
+    private CookingRecipeCategory category;  // Microtus - RecipeCategories
     private DeclareRecipesPacket.Ingredient ingredient;
     private ItemStack result;
     private float experience;
@@ -14,12 +15,14 @@ public abstract class BlastingRecipe extends Recipe {
     protected BlastingRecipe(
             @NotNull String recipeId,
             @NotNull String group,
+            @NotNull CookingRecipeCategory category,  // Microtus - RecipeCategories
             @NotNull ItemStack result,
             float experience,
             int cookingTime
     ) {
         super(Type.BLASTING, recipeId);
         this.group = group;
+        this.category = category;  // Microtus - RecipeCategories
         this.result = result;
         this.experience = experience;
         this.cookingTime = cookingTime;
@@ -34,6 +37,17 @@ public abstract class BlastingRecipe extends Recipe {
         this.group = group;
     }
 
+    // Microtus start - RecipeCategories
+    @NotNull
+    public CookingRecipeCategory getCategory() {
+        return category;
+    }
+
+    public void setCategory(@NotNull CookingRecipeCategory category) {
+        this.category = category;
+    }
+    // Microtus end - RecipeCategories
+
     @NotNull
     public DeclareRecipesPacket.Ingredient getIngredient() {
         return ingredient;
diff --git a/src/main/java/net/minestom/server/recipe/CampfireCookingRecipe.java b/src/main/java/net/minestom/server/recipe/CampfireCookingRecipe.java
index 9c0d72ad6fb077dbb8ac72b5f7d4e5b829fd69aa..d47bf192508e19e7d1f6229c7fe05338329fa9ea 100644
--- a/src/main/java/net/minestom/server/recipe/CampfireCookingRecipe.java
+++ b/src/main/java/net/minestom/server/recipe/CampfireCookingRecipe.java
@@ -6,6 +6,7 @@ import org.jetbrains.annotations.NotNull;
 
 public abstract class CampfireCookingRecipe extends Recipe {
     private String group;
+    private CookingRecipeCategory category;  // Microtus - RecipeCategories
     private DeclareRecipesPacket.Ingredient ingredient;
     private ItemStack result;
     private float experience;
@@ -14,12 +15,14 @@ public abstract class CampfireCookingRecipe extends Recipe {
     protected CampfireCookingRecipe(
             @NotNull String recipeId,
             @NotNull String group,
+            @NotNull CookingRecipeCategory category, // Microtus - RecipeCategories
             @NotNull ItemStack result,
             float experience,
             int cookingTime
     ) {
         super(Type.CAMPFIRE_COOKING, recipeId);
         this.group = group;
+        this.category = category; // Microtus - RecipeCategories
         this.result = result;
         this.experience = experience;
         this.cookingTime = cookingTime;
@@ -34,6 +37,17 @@ public abstract class CampfireCookingRecipe extends Recipe {
         this.group = group;
     }
 
+    // Microtus start - RecipeCategories
+    @NotNull
+    public CookingRecipeCategory getCategory() {
+        return category;
+    }
+
+    public void setCategory(@NotNull CookingRecipeCategory category) {
+        this.category = category;
+    }
+    // Microtus end - RecipeCategories
+
     @NotNull
     public DeclareRecipesPacket.Ingredient getIngredient() {
         return ingredient;
diff --git a/src/main/java/net/minestom/server/recipe/CookingRecipeCategory.java b/src/main/java/net/minestom/server/recipe/CookingRecipeCategory.java
new file mode 100644
index 0000000000000000000000000000000000000000..ce6a1c4d1de4bd1a85be7321071b239683a3e028
--- /dev/null
+++ b/src/main/java/net/minestom/server/recipe/CookingRecipeCategory.java
@@ -0,0 +1,8 @@
+package net.minestom.server.recipe;
+// Microtus start - RecipeCategories
+public enum CookingRecipeCategory {
+    FOOD,
+    BLOCKS,
+    MISC
+}
+// Microtus end - RecipeCategories
\ No newline at end of file
diff --git a/src/main/java/net/minestom/server/recipe/CraftingRecipeCategory.java b/src/main/java/net/minestom/server/recipe/CraftingRecipeCategory.java
new file mode 100644
index 0000000000000000000000000000000000000000..cfb0454668bb86ca97f0e023e548cb1636359c73
--- /dev/null
+++ b/src/main/java/net/minestom/server/recipe/CraftingRecipeCategory.java
@@ -0,0 +1,9 @@
+package net.minestom.server.recipe;
+// Microtus start - RecipeCategories
+public enum CraftingRecipeCategory {
+    BUILDING,
+    REDSTONE,
+    EQUIPMENT,
+    MISC
+}
+// Microtus end - RecipeCategories
\ No newline at end of file
diff --git a/src/main/java/net/minestom/server/recipe/RecipeManager.java b/src/main/java/net/minestom/server/recipe/RecipeManager.java
index a5b004e03a0f3228cc8a81610195b17ae6062eea..9251a817797a25ac77a3a61d3f93b83cefd61627 100644
--- a/src/main/java/net/minestom/server/recipe/RecipeManager.java
+++ b/src/main/java/net/minestom/server/recipe/RecipeManager.java
@@ -50,6 +50,7 @@ public class RecipeManager {
                             new DeclareRecipesPacket.DeclaredShapelessCraftingRecipe(
                                     shapelessRecipe.getRecipeId(),
                                     shapelessRecipe.getGroup(),
+                                    shapelessRecipe.getCategory(), // Microtus - RecipeCategories
                                     shapelessRecipe.getIngredients(),
                                     shapelessRecipe.getResult()));
                 }
@@ -61,6 +62,7 @@ public class RecipeManager {
                                     shapedRecipe.getWidth(),
                                     shapedRecipe.getHeight(),
                                     shapedRecipe.getGroup(),
+                                    shapedRecipe.getCategory(), // Microtus - RecipeCategories
                                     shapedRecipe.getIngredients(),
                                     shapedRecipe.getResult()));
                 }
@@ -71,6 +73,7 @@ public class RecipeManager {
                                     smeltingRecipe.getRecipeId(),
                                     smeltingRecipe.getGroup(),
                                     smeltingRecipe.getIngredient(),
+                                    smeltingRecipe.getCategory(), // Microtus - RecipeCategories
                                     smeltingRecipe.getResult(),
                                     smeltingRecipe.getExperience(),
                                     smeltingRecipe.getCookingTime()));
@@ -82,6 +85,7 @@ public class RecipeManager {
                                     blastingRecipe.getRecipeId(),
                                     blastingRecipe.getGroup(),
                                     blastingRecipe.getIngredient(),
+                                    blastingRecipe.getCategory(), // Microtus - RecipeCategories
                                     blastingRecipe.getResult(),
                                     blastingRecipe.getExperience(),
                                     blastingRecipe.getCookingTime()));
@@ -93,6 +97,7 @@ public class RecipeManager {
                                     smokingRecipe.getRecipeId(),
                                     smokingRecipe.getGroup(),
                                     smokingRecipe.getIngredient(),
+                                    smokingRecipe.getCategory(), // Microtus - RecipeCategories
                                     smokingRecipe.getResult(),
                                     smokingRecipe.getExperience(),
                                     smokingRecipe.getCookingTime()));
@@ -104,6 +109,7 @@ public class RecipeManager {
                                     campfireCookingRecipe.getRecipeId(),
                                     campfireCookingRecipe.getGroup(),
                                     campfireCookingRecipe.getIngredient(),
+                                    campfireCookingRecipe.getCategory(), // Microtus - RecipeCategories
                                     campfireCookingRecipe.getResult(),
                                     campfireCookingRecipe.getExperience(),
                                     campfireCookingRecipe.getCookingTime()));
diff --git a/src/main/java/net/minestom/server/recipe/ShapedRecipe.java b/src/main/java/net/minestom/server/recipe/ShapedRecipe.java
index f8f66c99d6cb4065c550912dea8982c9e3da5fca..c33d5193147f709dea6104975bb7d23530fc2a9b 100644
--- a/src/main/java/net/minestom/server/recipe/ShapedRecipe.java
+++ b/src/main/java/net/minestom/server/recipe/ShapedRecipe.java
@@ -14,18 +14,21 @@ public abstract class ShapedRecipe extends Recipe {
     private final int height;
     private String group;
     private final List<DeclareRecipesPacket.Ingredient> ingredients;
+    private CraftingRecipeCategory category; // Microtus - RecipeCategories
     private ItemStack result;
 
     protected ShapedRecipe(@NotNull String recipeId,
                            int width,
                            int height,
                            @NotNull String group,
+                           @NotNull CraftingRecipeCategory category, // Microtus - RecipeCategories
                            @Nullable List<DeclareRecipesPacket.Ingredient> ingredients,
                            @NotNull ItemStack result) {
         super(Type.SHAPED, recipeId);
         this.width = width;
         this.height = height;
         this.group = group;
+        this.category = category; // Microtus - RecipeCategories
         this.ingredients = Objects.requireNonNullElseGet(ingredients, LinkedList::new);
         this.result = result;
     }
@@ -43,6 +46,17 @@ public abstract class ShapedRecipe extends Recipe {
         return group;
     }
 
+    // Microtus start - RecipeCategories
+    @NotNull
+    public CraftingRecipeCategory getCategory() {
+        return category;
+    }
+
+    public void setCategory(@NotNull CraftingRecipeCategory category) {
+        this.category = category;
+    }
+    // Microtus end - RecipeCategories
+
     public void setGroup(@NotNull String group) {
         this.group = group;
     }
diff --git a/src/main/java/net/minestom/server/recipe/ShapelessRecipe.java b/src/main/java/net/minestom/server/recipe/ShapelessRecipe.java
index 28ca42ab2089d4ae2bcf1b6f0c4c714e107ceed2..fadd16ab841f4dea8eb277f5997b77563319b35c 100644
--- a/src/main/java/net/minestom/server/recipe/ShapelessRecipe.java
+++ b/src/main/java/net/minestom/server/recipe/ShapelessRecipe.java
@@ -11,17 +11,20 @@ import java.util.Objects;
 
 public abstract class ShapelessRecipe extends Recipe {
     private String group;
+    private CraftingRecipeCategory category; // Microtus - RecipeCategories
     private final List<DeclareRecipesPacket.Ingredient> ingredients;
     private ItemStack result;
 
     protected ShapelessRecipe(
             @NotNull String recipeId,
             @NotNull String group,
+            @NotNull CraftingRecipeCategory category, // Microtus - RecipeCategories
             @Nullable List<DeclareRecipesPacket.Ingredient> ingredients,
             @NotNull ItemStack result
     ) {
         super(Type.SHAPELESS, recipeId);
         this.group = group;
+        this.category = category; // Microtus - RecipeCategories
         this.ingredients = Objects.requireNonNullElseGet(ingredients, LinkedList::new);
         this.result = result;
     }
@@ -35,6 +38,17 @@ public abstract class ShapelessRecipe extends Recipe {
         this.group = group;
     }
 
+    // Microtus start - RecipeCategories
+    @NotNull
+    public CraftingRecipeCategory getCategory() {
+        return category;
+    }
+
+    public void setCategory(@NotNull CraftingRecipeCategory category) {
+        this.category = category;
+    }
+    // Microtus end - RecipeCategories
+
     public void addIngredient(DeclareRecipesPacket.Ingredient ingredient) {
         ingredients.add(ingredient);
     }
diff --git a/src/main/java/net/minestom/server/recipe/SmeltingRecipe.java b/src/main/java/net/minestom/server/recipe/SmeltingRecipe.java
index 6ba216f1c77d22be646cf83b88d93f684fad16f0..1625df110906dc10cc9df9802af9b25b903d2681 100644
--- a/src/main/java/net/minestom/server/recipe/SmeltingRecipe.java
+++ b/src/main/java/net/minestom/server/recipe/SmeltingRecipe.java
@@ -6,6 +6,7 @@ import org.jetbrains.annotations.NotNull;
 
 public abstract class SmeltingRecipe extends Recipe {
     private String group;
+    private CookingRecipeCategory category; // Microtus - RecipeCategories
     private DeclareRecipesPacket.Ingredient ingredient;
     private ItemStack result;
     private float experience;
@@ -14,12 +15,14 @@ public abstract class SmeltingRecipe extends Recipe {
     protected SmeltingRecipe(
             @NotNull String recipeId,
             @NotNull String group,
+            @NotNull CookingRecipeCategory category, // Microtus - RecipeCategories
             @NotNull ItemStack result,
             float experience,
             int cookingTime
     ) {
         super(Type.SMELTING, recipeId);
         this.group = group;
+        this.category = category;  // Microtus - RecipeCategories
         this.result = result;
         this.experience = experience;
         this.cookingTime = cookingTime;
@@ -33,6 +36,17 @@ public abstract class SmeltingRecipe extends Recipe {
     public void setGroup(@NotNull String group) {
         this.group = group;
     }
+    // Microtus start - RecipeCategories
+    @NotNull
+    public CookingRecipeCategory getCategory() {
+        return category;
+    }
+
+    public void setCategory(@NotNull CookingRecipeCategory category) {
+        this.category = category;
+    }
+
+    // Microtus end - RecipeCategories
 
     @NotNull
     public DeclareRecipesPacket.Ingredient getIngredient() {
diff --git a/src/main/java/net/minestom/server/recipe/SmokingRecipe.java b/src/main/java/net/minestom/server/recipe/SmokingRecipe.java
index 278dc8457a98ec32e1b69bef09e061bf23c2a6c7..7887305d43b55dd6cce3376023cbebd0d4859fce 100644
--- a/src/main/java/net/minestom/server/recipe/SmokingRecipe.java
+++ b/src/main/java/net/minestom/server/recipe/SmokingRecipe.java
@@ -6,6 +6,7 @@ import org.jetbrains.annotations.NotNull;
 
 public abstract class SmokingRecipe extends Recipe {
     private String group;
+    private CookingRecipeCategory category; // Microtus - RecipeCategories
     private DeclareRecipesPacket.Ingredient ingredient;
     private ItemStack result;
     private float experience;
@@ -14,12 +15,14 @@ public abstract class SmokingRecipe extends Recipe {
     protected SmokingRecipe(
             @NotNull String recipeId,
             @NotNull String group,
+            @NotNull CookingRecipeCategory category, // Microtus - RecipeCategories
             @NotNull ItemStack result,
             float experience,
             int cookingTime
     ) {
         super(Type.SMOKING, recipeId);
         this.group = group;
+        this.category = category; // Microtus - RecipeCategories
         this.result = result;
         this.experience = experience;
         this.cookingTime = cookingTime;
@@ -34,6 +37,18 @@ public abstract class SmokingRecipe extends Recipe {
         this.group = group;
     }
 
+    // Microtus start - RecipeCategories
+    @NotNull
+    public CookingRecipeCategory getCategory() {
+        return category;
+    }
+
+    public void setCategory(@NotNull CookingRecipeCategory category) {
+        this.category = category;
+    }
+
+    // Microtus end - RecipeCategories
+
     @NotNull
     public DeclareRecipesPacket.Ingredient getIngredient() {
         return ingredient;
diff --git a/src/test/java/net/minestom/server/network/PacketWriteReadTest.java b/src/test/java/net/minestom/server/network/PacketWriteReadTest.java
index a2be94cfe04bb14ab994b45097e2068cee418eca..f23020126745417521cc687d2eac2bc1539549c8 100644
--- a/src/test/java/net/minestom/server/network/PacketWriteReadTest.java
+++ b/src/test/java/net/minestom/server/network/PacketWriteReadTest.java
@@ -95,6 +95,7 @@ public class PacketWriteReadTest {
                 List.of(new DeclareRecipesPacket.DeclaredShapelessCraftingRecipe(
                                 "minecraft:sticks",
                                 "sticks",
+                                net.minestom.server.recipe.CraftingRecipeCategory.MISC, // Microtus - RecipeCategories
                                 List.of(new Ingredient(List.of(ItemStack.of(Material.OAK_PLANKS)))),
                                 ItemStack.of(Material.STICK)
                         ),
@@ -103,6 +104,7 @@ public class PacketWriteReadTest {
                                 1,
                                 2,
                                 "",
+                                net.minestom.server.recipe.CraftingRecipeCategory.MISC, // Microtus - RecipeCategories
                                 List.of(new Ingredient(List.of(ItemStack.of(Material.COAL))),
                                         new Ingredient(List.of(ItemStack.of(Material.STICK)))),
                                 ItemStack.of(Material.TORCH)
