From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Phillipp Glanz <p.glanz@madfix.me>
Date: Thu, 30 Nov 2023 12:52:30 +0100
Subject: [PATCH] Improve entity metadata handling


diff --git a/src/main/java/net/minestom/server/entity/metadata/animal/FrogMeta.java b/src/main/java/net/minestom/server/entity/metadata/animal/FrogMeta.java
index cfb2914cea7591a1d69ab6b1e147bafbacc7671b..f8ebe95a2eb670337d47180a8943217efbaad292 100644
--- a/src/main/java/net/minestom/server/entity/metadata/animal/FrogMeta.java
+++ b/src/main/java/net/minestom/server/entity/metadata/animal/FrogMeta.java
@@ -13,29 +13,64 @@ public class FrogMeta extends AnimalMeta {
         super(entity, metadata);
     }
 
-    public Variant getVariant() {
+    /**
+     * Get the current {@link Variant} from a frog.
+     * @return the given variant entry
+     */
+    public @NotNull Variant getVariant() {
         return super.metadata.getIndex(OFFSET, Variant.TEMPERATE);
     }
 
+    /**
+     * Set the variant for a frog.
+     * @param value the value to set
+     */
     public void setVariant(@NotNull Variant value) {
         super.metadata.setIndex(OFFSET, Metadata.FrogVariant(value));
     }
 
-    public @Nullable Integer getTongueTarget() {
-        return super.metadata.getIndex(OFFSET + 1, null);
+    /**
+     * Get the current tongue target value.
+     * @return the given value
+     */
+    public int getTongueTarget() {
+        return super.metadata.getIndex(OFFSET + 1, 0);
     }
 
+    /**
+     * Set's the current tongue target back to the default value from the protocol.
+     */
+    public void resetTongueTarget() {
+        super.metadata.setIndex(OFFSET + 1, Metadata.VarInt(0));
+    }
 
-    public void setTongueTarget(@Nullable Integer value) {
+    /**
+     * Set the target value for a tongue.
+     * @param value the target to set
+     */
+    public void setTongueTarget(int value) {
         super.metadata.setIndex(OFFSET + 1, Metadata.OptVarInt(value));
     }
 
-
+    /**
+     * The enum contains all variants from a frog which are currently implemented in the game.
+     */
     public enum Variant {
         TEMPERATE,
         WARM,
         COLD;
 
-        private final static FrogMeta.Variant[] VALUES = values();
+        private static final FrogMeta.Variant[] VALUES = values();
+
+        // Microtus start - meta update
+        /**
+         * Add method to get a variant from a frog over the ordinal id
+         * @param id the ordinal id
+         * @return the entry which matches with the id
+         */
+        public static @Nullable FrogMeta.Variant getVariant(int id) {
+            return id >= 0 && id <= VALUES.length ? VALUES[id] : null;
+        }
+        // Microtus end - meta update
     }
 }
diff --git a/src/main/java/net/minestom/server/entity/metadata/animal/GoatMeta.java b/src/main/java/net/minestom/server/entity/metadata/animal/GoatMeta.java
index de2a71b81a4b9787ec31b5213d0c00bc1befdb6b..6d78f6cf69a410d58317945439070c7a9b7dac85 100644
--- a/src/main/java/net/minestom/server/entity/metadata/animal/GoatMeta.java
+++ b/src/main/java/net/minestom/server/entity/metadata/animal/GoatMeta.java
@@ -20,6 +20,7 @@ public class GoatMeta extends AnimalMeta {
         metadata.setIndex(OFFSET, Metadata.Boolean(screaming));
     }
 
+    // Microtus start - meta update
     public boolean hasLeftHorn() {
         return metadata.getIndex(OFFSET + 1, true);
     }
@@ -35,4 +36,5 @@ public class GoatMeta extends AnimalMeta {
     public void setRightHorn(boolean rightHorn) {
         metadata.setIndex(OFFSET + 2, Metadata.Boolean(rightHorn));
     }
+    // Microtus end - meta update
 }
diff --git a/src/main/java/net/minestom/server/entity/metadata/monster/WardenMeta.java b/src/main/java/net/minestom/server/entity/metadata/monster/WardenMeta.java
index 261f6b584d5db0fe60c9e4774860736b35d5163c..4354c9ce62a2a9053ed880d094134cdae4357cd9 100644
--- a/src/main/java/net/minestom/server/entity/metadata/monster/WardenMeta.java
+++ b/src/main/java/net/minestom/server/entity/metadata/monster/WardenMeta.java
@@ -4,21 +4,40 @@ import net.minestom.server.entity.Entity;
 import net.minestom.server.entity.Metadata;
 import org.jetbrains.annotations.NotNull;
 
+// Microtus - entity meta update
+/**
+ * This metadata implementation can be used for a warden entity.
+ * @author theEvilReaper
+ * @version 1.0.0
+ * @since 1.0.0
+ **/
 public class WardenMeta extends MonsterMeta {
 
     public static final byte OFFSET = MonsterMeta.MAX_OFFSET;
     public static final byte MAX_OFFSET = OFFSET + 1;
 
+    /**
+     * Creates a new metadata which can be set to a warden entity.
+     * @param entity the warden entity reference
+     * @param metadata the reference to a {@link Metadata}
+     */
     public WardenMeta(@NotNull Entity entity, @NotNull Metadata metadata) {
         super(entity, metadata);
     }
 
-    public int getAngerLevel() {
-        return super.metadata.getIndex(OFFSET, 0);
-    }
-
+    /**
+     * Set the anger level for a warden.
+     * @param value the level to set
+     */
     public void setAngerLevel(int value) {
         super.metadata.setIndex(OFFSET, Metadata.VarInt(value));
     }
 
+    /**
+     * Returns the given anger level from the meta or zero as default value if no value is set.
+     * @return the given level
+     */
+    public int getAngerLevel() {
+        return super.metadata.getIndex(OFFSET, 0);
+    }
 }
diff --git a/src/main/java/net/minestom/server/entity/metadata/monster/raider/PillagerMeta.java b/src/main/java/net/minestom/server/entity/metadata/monster/raider/PillagerMeta.java
index 8096e5e2ab77e099192e49cde16eafce6a745eb9..1cf1168ecd108e70e3bc827c4a6aa244897ae8e8 100644
--- a/src/main/java/net/minestom/server/entity/metadata/monster/raider/PillagerMeta.java
+++ b/src/main/java/net/minestom/server/entity/metadata/monster/raider/PillagerMeta.java
@@ -12,12 +12,21 @@ public class PillagerMeta extends AbstractIllagerMeta {
         super(entity, metadata);
     }
 
-    public boolean isChargingCrossbow() {
-        return super.metadata.getIndex(OFFSET, false);
-    }
-
+    // Microtus start - meta update
+    /**
+     * Set the state if a Pillager charge his crossbow.
+     * @param value the value to set
+     */
     public void setChargingCrossbow(boolean value) {
         super.metadata.setIndex(OFFSET, Metadata.Boolean(value));
     }
 
+    /**
+     * Returns a boolean value if a Pillager is charging his crossbow.
+     * @return true when yes otherwise false
+     */
+    public boolean isChargingCrossbow() {
+        return super.metadata.getIndex(OFFSET, false);
+    }
+    // Microtus end - meta update
 }
diff --git a/src/main/java/net/minestom/server/entity/metadata/other/BoatMeta.java b/src/main/java/net/minestom/server/entity/metadata/other/BoatMeta.java
index 57e3e40ec7d89c5484e02f3d17b1d255fb24b986..84a81a7b326efed0b37f244cadb8b5d6c117ae98 100644
--- a/src/main/java/net/minestom/server/entity/metadata/other/BoatMeta.java
+++ b/src/main/java/net/minestom/server/entity/metadata/other/BoatMeta.java
@@ -81,7 +81,7 @@ public class BoatMeta extends EntityMeta {
         MANGROVE,
         BAMBOO;
 
-        private final static Type[] VALUES = values();
+        private static final Type[] VALUES = values();
     }
 
 }
diff --git a/src/main/java/net/minestom/server/entity/metadata/water/DolphinMeta.java b/src/main/java/net/minestom/server/entity/metadata/water/DolphinMeta.java
index 6dca1c672effa503e3a7ba81b9da84001126cd45..1e39f9b4c70bbaa74b7aec1769619a2ab46948c4 100644
--- a/src/main/java/net/minestom/server/entity/metadata/water/DolphinMeta.java
+++ b/src/main/java/net/minestom/server/entity/metadata/water/DolphinMeta.java
@@ -6,36 +6,66 @@ import net.minestom.server.entity.Entity;
 import net.minestom.server.entity.Metadata;
 import org.jetbrains.annotations.NotNull;
 
+// Microtus start - meta update
 public class DolphinMeta extends WaterAnimalMeta {
+
     public static final byte OFFSET = WaterAnimalMeta.MAX_OFFSET;
     public static final byte MAX_OFFSET = OFFSET + 3;
 
+    /**
+     * Creates a new meta data for a dolphin.
+     * @param entity the involved entity
+     * @param metadata the base metadata
+     */
     public DolphinMeta(@NotNull Entity entity, @NotNull Metadata metadata) {
         super(entity, metadata);
     }
 
-    @NotNull
-    public Point getTreasurePosition() {
+    /**
+     * Get the given destination point for a treasure.
+     * @return the point or {@link Vec#ZERO} when no point is set
+     */
+    public @NotNull Point getTreasurePosition() {
         return super.metadata.getIndex(OFFSET, Vec.ZERO);
     }
 
+    /**
+     * Set the destination point to lead a player to a treasure.
+     * @param value the point to set
+     */
     public void setTreasurePosition(@NotNull Point value) {
         super.metadata.setIndex(OFFSET, Metadata.Position(value));
     }
 
-    public boolean isHasFish() {
+    /**
+     * Returns an indicator if a Dolphin received a fish from a player.
+     * @return true when yes otherwise false
+     */
+    public boolean hasFish() {
         return super.metadata.getIndex(OFFSET + 1, false);
     }
 
+    /**
+     * Set the indicator if a Dolphin got a fish from a player.
+     * @param value the value to set
+     */
     public void setHasFish(boolean value) {
         super.metadata.setIndex(OFFSET + 1, Metadata.Boolean(value));
     }
 
+    /**
+     * Get the current moisture level.
+     * @return the given level
+     */
     public int getMoistureLevel() {
         return super.metadata.getIndex(OFFSET + 2, 2400);
     }
 
-    public void setMoistureLevel(int value) {
-        super.metadata.setIndex(OFFSET + 2, Metadata.VarInt(value));
+    /**
+     * Updates the given moisture level.
+     * @param level the level to set
+     */
+    public void setMoistureLevel(int level) {
+        super.metadata.setIndex(OFFSET + 2, Metadata.VarInt(level));
     }
 }
diff --git a/src/main/java/net/minestom/server/entity/metadata/water/fish/TadpoleMeta.java b/src/main/java/net/minestom/server/entity/metadata/water/fish/TadpoleMeta.java
index 8cf3e9c86854ae20115dc05d54a619e418e850da..ad9152ca44eb6521458aa20a4484278c9334614e 100644
--- a/src/main/java/net/minestom/server/entity/metadata/water/fish/TadpoleMeta.java
+++ b/src/main/java/net/minestom/server/entity/metadata/water/fish/TadpoleMeta.java
@@ -4,12 +4,18 @@ import net.minestom.server.entity.Entity;
 import net.minestom.server.entity.Metadata;
 import org.jetbrains.annotations.NotNull;
 
-public class TadpoleMeta extends AbstractFishMeta {
-    public static final byte OFFSET = AbstractFishMeta.MAX_OFFSET;
-    public static final byte MAX_OFFSET = OFFSET + 0;
+// Microtus - entity meta update
+/**
+ * Metadata implementation for a Tadpole entity.
+ */
+public final class TadpoleMeta extends AbstractFishMeta {
 
+    /**
+     * Creates a new reference from this meta.
+     * @param entity the involved entity
+     * @param metadata the involved metadata
+     */
     public TadpoleMeta(@NotNull Entity entity, @NotNull Metadata metadata) {
         super(entity, metadata);
     }
-
 }
diff --git a/src/main/java/net/minestom/server/network/NetworkBufferTypes.java b/src/main/java/net/minestom/server/network/NetworkBufferTypes.java
index d86c8fb731eccb95cf60cf45fc28dd2f18d8c19c..13cc9079a34832e9bae32d275b0e9842bd761164 100644
--- a/src/main/java/net/minestom/server/network/NetworkBufferTypes.java
+++ b/src/main/java/net/minestom/server/network/NetworkBufferTypes.java
@@ -571,7 +571,7 @@ final class NetworkBufferTypes {
             },
             buffer -> {
                 final int ordinal = buffer.read(VAR_INT);
-                return FrogMeta.Variant.values()[ordinal];
+                return FrogMeta.Variant.getVariant(ordinal); // Microtus - meta update
             });
     static final TypeImpl<SnifferMeta.State> SNIFFER_STATE = new TypeImpl<>(SnifferMeta.State.class,
             (buffer, value) -> {
diff --git a/src/test/java/net/minestom/server/entity/metadata/animal/FrogMetaTest.java b/src/test/java/net/minestom/server/entity/metadata/animal/FrogMetaTest.java
new file mode 100644
index 0000000000000000000000000000000000000000..dff4f20dac5cfb5dc02c815ef381f91fc093fa97
--- /dev/null
+++ b/src/test/java/net/minestom/server/entity/metadata/animal/FrogMetaTest.java
@@ -0,0 +1,18 @@
+package net.minestom.server.entity.metadata.animal;
+
+import net.minestom.server.entity.metadata.animal.FrogMeta.Variant;
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+// Microtus - meta update + add test for the variant read method
+class FrogMetaTest {
+
+    @Test
+    void testFrogVariantRead() {
+        assertEquals(Variant.WARM, Variant.getVariant(Variant.WARM.ordinal()));
+        assertEquals(Variant.TEMPERATE, Variant.getVariant(0));
+        assertNull(Variant.getVariant(-1));
+        assertNull(Variant.getVariant(100));
+    }
+}
\ No newline at end of file
