From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: TheMeinerLP <p.glanz@madfix.me>
Date: Wed, 14 Jun 2023 09:38:27 +0200
Subject: [PATCH] Add github publishing


diff --git a/build-logic/src/main/kotlin/minestom.publishing-conventions.gradle.kts b/build-logic/src/main/kotlin/minestom.publishing-conventions.gradle.kts
index b63a95323b4e20af1c36f29f24c623d978016696..30d97cbeb3737c0fdfeb469d3a3cdca7e6fb643c 100644
--- a/build-logic/src/main/kotlin/minestom.publishing-conventions.gradle.kts
+++ b/build-logic/src/main/kotlin/minestom.publishing-conventions.gradle.kts
@@ -10,23 +10,17 @@ indra {
         testWith(17)
     }
 
-    github("Minestom", "Minestom") {
+    publishAllTo("github", "https://maven.pkg.github.com/OneLiteFeatherNET/Microtus")
+
+    github("OneLiteFeatherNET", "Microtus") {
         ci(true)
     }
-    apache2License()
-
-    configurePublications {
-        pom {
-            developers {
-                developer {
-                    id.set("TheMode")
-                    name.set("TheMode")
-                }
-                developer {
-                    id.set("jglrxavpok")
-                    name.set("jglrxavpok")
-                }
-            }
-        }
+    license {
+        name("APGL-3.0")
+        url("https://github.com/OneLiteFeatherNET/Microtus/blob/master/LICENSE")
     }
+}
+
+indraSonatype {
+    useAlternateSonatypeOSSHost("s01")
 }
\ No newline at end of file
diff --git a/build.gradle.kts b/build.gradle.kts
index 32f8a037754849d3ad31e4cab64b5f9ff57e9ce9..66606515d3244b280b8aa90351f2abb1413c443b 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -1,13 +1,40 @@
+import org.ajoberstar.grgit.Grgit
+import java.util.*
+
 plugins {
     `java-library`
     id("minestom.publishing-conventions")
     id("minestom.native-conventions")
     alias(libs.plugins.blossom)
+    // publishing
+    id("org.ajoberstar.grgit") version "5.2.0"
+}
+
+var baseVersion by extra("1.1.0")
+var versionExtension by extra("")
+var snapshot by extra("-SNAPSHOT")
+
+group = "net.onelitefeather.microtus"
+
+
+ext {
+    val git: Grgit = Grgit.open {
+        dir = File("$rootDir/.git")
+    }
+    val revision = git.head().abbreviatedId
+    if (snapshot.isNotEmpty()) {
+        versionExtension = "%s+%s".format(Locale.ROOT, snapshot, revision)
+    } else {
+        versionExtension = ""
+    }
+
 }
 
+version = "%s%s".format(Locale.ROOT, baseVersion, versionExtension)
+
 allprojects {
-    group = "net.minestom.server"
-    version = "1.0"
+    group = "net.onelitefeather.microtus"
+    version = rootProject.version
     description = "Lightweight and multi-threaded Minecraft server implementation"
 }
 
@@ -34,7 +61,7 @@ tasks {
             addStringOption("-release", "17")
             // Links to external javadocs
             links("https://docs.oracle.com/en/java/javase/17/docs/api/")
-            links("https://jd.adventure.kyori.net/api/${libs.versions.adventure.get()}/")
+            // links("https://jd.adventure.kyori.net/api/${libs.versions.adventure.get()}/")
         }
     }
     withType<Zip> {
diff --git a/testing/build.gradle.kts b/testing/build.gradle.kts
index 4ea0e6bc68bea05e0345261a861fc9fcae9461fe..d8d8282ab8663995d7dc9b7ab44282856cc42704 100644
--- a/testing/build.gradle.kts
+++ b/testing/build.gradle.kts
@@ -1,21 +1,10 @@
 plugins {
     id("java-library")
-    `maven-publish`
+    id("net.kyori.indra")
+    id("net.kyori.indra.publishing")
 }
 
-publishing {
-    publications {
-        create<MavenPublication>("maven") {
-            groupId = "net.minestom.testing"
-            artifactId = "testing"
-            version = "1.0"
-
-            from(components["java"])
-        }
-    }
-}
-
-group = "net.minestom.testing"
+group = "net.onelitefeather.microtus.testing"
 version = "1.0"
 
 repositories {
@@ -36,4 +25,21 @@ dependencies {
 
 tasks.getByName<Test>("test") {
     useJUnitPlatform()
+}
+
+indra {
+    javaVersions {
+        target(17)
+        testWith(17)
+    }
+
+    publishAllTo("github", "https://maven.pkg.github.com/OneLiteFeatherNET/Microtus")
+
+    github("OneLiteFeatherNET", "Microtus") {
+        ci(true)
+    }
+    license {
+        name("APGL-3.0")
+        url("https://github.com/OneLiteFeatherNET/Microtus/blob/master/LICENSE")
+    }
 }
\ No newline at end of file
