From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Phillipp Glanz <p.glanz@madfix.me>
Date: Thu, 30 Nov 2023 12:49:37 +0100
Subject: [PATCH] Improve hashcode and equals integration


diff --git a/src/main/java/net/minestom/server/entity/Entity.java b/src/main/java/net/minestom/server/entity/Entity.java
index daf0ece5bc96be7862d265f89ae8edee6958fc13..ff0b45ffe8ac3ce607de0c4312b11451014a11a5 100644
--- a/src/main/java/net/minestom/server/entity/Entity.java
+++ b/src/main/java/net/minestom/server/entity/Entity.java
@@ -477,6 +477,21 @@ public class Entity implements Viewable, Tickable, Schedulable, Snapshotable, Ev
         return viewers;
     }
 
+    //Microtus start - equals and hashcode
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        Entity entity = (Entity) o;
+        return Objects.equals(uuid, entity.uuid);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(uuid);
+    }
+    //Microtus end - equals and hashcode
+
     /**
      * Gets if this entity's viewers (surrounding players) can be predicted from surrounding chunks.
      */
diff --git a/src/main/java/net/minestom/server/instance/Instance.java b/src/main/java/net/minestom/server/instance/Instance.java
index b7c5affa63ebe29d30996718c3402d8348137f7d..d6dd0549db22ee58c4057012fcc46a3aba8f087b 100644
--- a/src/main/java/net/minestom/server/instance/Instance.java
+++ b/src/main/java/net/minestom/server/instance/Instance.java
@@ -451,6 +451,21 @@ public abstract class Instance implements Block.Getter, Block.Setter,
         return new TimeUpdatePacket(worldAge, time);
     }
 
+    //Microtus start - equals and hashcode
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        Instance instance = (Instance) o;
+        return Objects.equals(uniqueId, instance.uniqueId);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(uniqueId);
+    }
+    //Microtus end - equals and hashcode
+
     /**
      * Gets the instance {@link WorldBorder};
      *
diff --git a/src/main/java/net/minestom/server/utils/NamespaceID.java b/src/main/java/net/minestom/server/utils/NamespaceID.java
index 35b34bfe35067d28eef3c66539d9fd4f0fa68c3f..e68ff90079bea1debfa543b14f6df39396359234 100644
--- a/src/main/java/net/minestom/server/utils/NamespaceID.java
+++ b/src/main/java/net/minestom/server/utils/NamespaceID.java
@@ -67,12 +67,14 @@ public final class NamespaceID implements CharSequence, Key {
     public boolean equals(Object o) {
         if (this == o) return true;
         if (!(o instanceof final Key that)) return false;
-        return Objects.equals(this.full, that.asString());
+        return Objects.equals(this.domain, that.namespace()) && Objects.equals(this.path, that.value()); // Microtus - Fix namespace hashcode and equals
     }
 
     @Override
     public int hashCode() {
-        return full.hashCode();
+        int result = this.domain.hashCode(); // Microtus start - Fix namespace hashcode and equals
+        result = (31 * result) + this.path.hashCode();
+        return result; // Microtus end - Fix namespace hashcode and equals
     }
 
     @Override
diff --git a/src/test/java/net/minestom/server/entity/EntityEqualsTest.java b/src/test/java/net/minestom/server/entity/EntityEqualsTest.java
new file mode 100644
index 0000000000000000000000000000000000000000..e11088a617e6da7b058b6c6a8a8ffb311ceebe48
--- /dev/null
+++ b/src/test/java/net/minestom/server/entity/EntityEqualsTest.java
@@ -0,0 +1,13 @@
+package net.minestom.server.entity;
+
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+class EntityEqualsTest {
+
+    @Test
+    void testEntityEquals() {
+        assertNotEquals(new Entity(EntityTypes.BEE), new Entity(EntityTypes.BEE));
+    }
+}
diff --git a/src/test/java/net/minestom/server/instance/InstanceEqualsTest.java b/src/test/java/net/minestom/server/instance/InstanceEqualsTest.java
new file mode 100644
index 0000000000000000000000000000000000000000..305be82a87d514f18510d13364f0574f193a093a
--- /dev/null
+++ b/src/test/java/net/minestom/server/instance/InstanceEqualsTest.java
@@ -0,0 +1,16 @@
+package net.minestom.server.instance;
+
+import net.minestom.testing.Env;
+import net.minestom.testing.EnvTest;
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+@EnvTest
+class InstanceEqualsTest {
+
+    @Test
+    void testInstanceEquals(Env env) {
+        assertNotEquals(env.createFlatInstance(), env.createFlatInstance());
+    }
+}
diff --git a/src/test/java/net/minestom/server/utils/NamespaceIDTest.java b/src/test/java/net/minestom/server/utils/NamespaceIDTest.java
index 2a41bf2cd11d1eb766850ef32b909c94d2a68597..5bae45cd92299a01210869345899018fa4c646af 100644
--- a/src/test/java/net/minestom/server/utils/NamespaceIDTest.java
+++ b/src/test/java/net/minestom/server/utils/NamespaceIDTest.java
@@ -27,6 +27,16 @@ public class NamespaceIDTest {
         assertNotEquals(namespace, NamespaceID.from("minecraft:any2"));
         assertEquals(namespace, Key.key("minecraft:any"));
     }
+    // Microtus start - Fix namespace hashcode and equals
+    @Test
+    public void hashCodeConsistentWithEquals() {
+        var namespace = NamespaceID.from("minecraft:any");
+        var key = Key.key("minecraft:any");
+
+        assertEquals(namespace, key);
+        assertEquals(namespace.hashCode(), key.hashCode());
+    }
+    // Microtus end - Fix namespace hashcode and equals
 
     @Test
     public void atMostOneColon() {
