From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: theEvilReaper <steffenwx@gmail.com>
Date: Mon, 19 Jun 2023 19:39:05 +0200
Subject: [PATCH] Add missing InventoryCloseEvent call in the closeInventory
 method


diff --git a/src/main/java/net/minestom/server/entity/Player.java b/src/main/java/net/minestom/server/entity/Player.java
index 93171392c717319c92321f7613b68fbc4356b49d..b3703129dad9aefbbaf521004d6a405abb656d74 100644
--- a/src/main/java/net/minestom/server/entity/Player.java
+++ b/src/main/java/net/minestom/server/entity/Player.java
@@ -30,6 +30,7 @@ import net.minestom.server.entity.fakeplayer.FakePlayer;
 import net.minestom.server.entity.metadata.PlayerMeta;
 import net.minestom.server.entity.vehicle.PlayerVehicleInformation;
 import net.minestom.server.event.EventDispatcher;
+import net.minestom.server.event.inventory.InventoryCloseEvent;
 import net.minestom.server.event.inventory.InventoryOpenEvent;
 import net.minestom.server.event.item.ItemDropEvent;
 import net.minestom.server.event.item.ItemUpdateStateEvent;
@@ -1515,6 +1516,10 @@ public class Player extends LivingEntity implements CommandSender, Localizable,
         }
 
         if (openInventory == getOpenInventory()) {
+            // Microtus start - 1.19.3 trigger inventory close
+            InventoryCloseEvent closeEvent = new InventoryCloseEvent(openInventory, this);
+            EventDispatcher.call(closeEvent);
+            // Microtus end - 1.19.3 trigger inventory close
             CloseWindowPacket closeWindowPacket;
             if (openInventory == null) {
                 closeWindowPacket = new CloseWindowPacket((byte) 0);
