From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Phillipp Glanz <p.glanz@madfix.me>
Date: Thu, 30 Nov 2023 12:52:30 +0100
Subject: [PATCH] Improve entity metadata handling


diff --git a/src/main/java/net/minestom/server/entity/Entity.java b/src/main/java/net/minestom/server/entity/Entity.java
index ff0b45ffe8ac3ce607de0c4312b11451014a11a5..8e7ad8420da6168ea2444dae90e5bc68bb32c176 100644
--- a/src/main/java/net/minestom/server/entity/Entity.java
+++ b/src/main/java/net/minestom/server/entity/Entity.java
@@ -1747,6 +1747,7 @@ public class Entity implements Viewable, Tickable, Schedulable, Snapshotable, Ev
         DYING,
         CROAKING,
         USING_TONGUE,
+        SITTING, // Microtus - Add missing entity pose
         ROARING,
         SNIFFING,
         EMERGING,
diff --git a/src/main/java/net/minestom/server/entity/EntityTypeImpl.java b/src/main/java/net/minestom/server/entity/EntityTypeImpl.java
index 23754c19cfb7a8d2dc65c319b2573d90bcffa60e..69330b33f092e5a13e1f4e0a6d6950e45109e147 100644
--- a/src/main/java/net/minestom/server/entity/EntityTypeImpl.java
+++ b/src/main/java/net/minestom/server/entity/EntityTypeImpl.java
@@ -33,10 +33,7 @@ import net.minestom.server.entity.metadata.water.AxolotlMeta;
 import net.minestom.server.entity.metadata.water.DolphinMeta;
 import net.minestom.server.entity.metadata.water.GlowSquidMeta;
 import net.minestom.server.entity.metadata.water.SquidMeta;
-import net.minestom.server.entity.metadata.water.fish.CodMeta;
-import net.minestom.server.entity.metadata.water.fish.PufferfishMeta;
-import net.minestom.server.entity.metadata.water.fish.SalmonMeta;
-import net.minestom.server.entity.metadata.water.fish.TropicalFishMeta;
+import net.minestom.server.entity.metadata.water.fish.*;
 import net.minestom.server.registry.Registry;
 import org.jetbrains.annotations.NotNull;
 
@@ -88,7 +85,7 @@ record EntityTypeImpl(Registry.EntityEntry registry) implements EntityType {
                 entry("minecraft:blaze", BlazeMeta::new),
                 entry("minecraft:breeze", BreezeMeta::new), // Microtus - 1.20
                 entry("minecraft:boat", BoatMeta::new),
-                entry("minecraft:chest_boat", EntityMeta::new), // TODO dedicated metadata
+                entry("minecraft:chest_boat", BoatMeta::new), // Microtus - entity meta update
                 entry("minecraft:cat", CatMeta::new),
                 entry("minecraft:block_display", BlockDisplayMeta::new), // Microtus - 1.19.4 update
                 entry("minecraft:cave_spider", CaveSpiderMeta::new),
@@ -176,7 +173,7 @@ record EntityTypeImpl(Registry.EntityEntry registry) implements EntityType {
                 entry("minecraft:squid", SquidMeta::new),
                 entry("minecraft:stray", StrayMeta::new),
                 entry("minecraft:strider", StriderMeta::new),
-                entry("minecraft:tadpole", EntityMeta::new), // TODO dedicated metadata
+                entry("minecraft:tadpole", TadpoleMeta::new), // Microtus - entity meta update
                 entry("minecraft:egg", ThrownEggMeta::new),
                 entry("minecraft:ender_pearl", ThrownEnderPearlMeta::new),
                 entry("minecraft:experience_bottle", ThrownExperienceBottleMeta::new),
@@ -189,7 +186,7 @@ record EntityTypeImpl(Registry.EntityEntry registry) implements EntityType {
                 entry("minecraft:villager", VillagerMeta::new),
                 entry("minecraft:vindicator", VindicatorMeta::new),
                 entry("minecraft:wandering_trader", WanderingTraderMeta::new),
-                entry("minecraft:warden", EntityMeta::new), // TODO dedicated metadata
+                entry("minecraft:warden", WardenMeta::new), // Microtus - entity meta update
                 entry("minecraft:witch", WitchMeta::new),
                 entry("minecraft:wither", WitherMeta::new),
                 entry("minecraft:wither_skeleton", WitherSkeletonMeta::new),
@@ -204,7 +201,7 @@ record EntityTypeImpl(Registry.EntityEntry registry) implements EntityType {
                 entry("minecraft:player", PlayerMeta::new),
                 entry("minecraft:fishing_bobber", FishingHookMeta::new),
                 // Microtus start - 1.20 update
-                entry("minecraft:camel", EntityMeta::new) // TODO dedicated metadata
+                entry("minecraft:camel", CamelMeta::new) // Microtus - entity meta update
                 // Microtus end - 1.20 update
         );
     }
diff --git a/src/main/java/net/minestom/server/entity/metadata/animal/FrogMeta.java b/src/main/java/net/minestom/server/entity/metadata/animal/FrogMeta.java
index a98c919b7939627199f292f63061baefea70f591..f8ebe95a2eb670337d47180a8943217efbaad292 100644
--- a/src/main/java/net/minestom/server/entity/metadata/animal/FrogMeta.java
+++ b/src/main/java/net/minestom/server/entity/metadata/animal/FrogMeta.java
@@ -3,6 +3,7 @@ package net.minestom.server.entity.metadata.animal;
 import net.minestom.server.entity.Entity;
 import net.minestom.server.entity.Metadata;
 import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
 
 public class FrogMeta extends AnimalMeta {
     public static final byte OFFSET = AnimalMeta.MAX_OFFSET;
@@ -12,27 +13,64 @@ public class FrogMeta extends AnimalMeta {
         super(entity, metadata);
     }
 
-    public Variant getVariant() {
+    /**
+     * Get the current {@link Variant} from a frog.
+     * @return the given variant entry
+     */
+    public @NotNull Variant getVariant() {
         return super.metadata.getIndex(OFFSET, Variant.TEMPERATE);
     }
 
+    /**
+     * Set the variant for a frog.
+     * @param value the value to set
+     */
     public void setVariant(@NotNull Variant value) {
         super.metadata.setIndex(OFFSET, Metadata.FrogVariant(value));
     }
 
+    /**
+     * Get the current tongue target value.
+     * @return the given value
+     */
     public int getTongueTarget() {
         return super.metadata.getIndex(OFFSET + 1, 0);
     }
 
+    /**
+     * Set's the current tongue target back to the default value from the protocol.
+     */
+    public void resetTongueTarget() {
+        super.metadata.setIndex(OFFSET + 1, Metadata.VarInt(0));
+    }
+
+    /**
+     * Set the target value for a tongue.
+     * @param value the target to set
+     */
     public void setTongueTarget(int value) {
         super.metadata.setIndex(OFFSET + 1, Metadata.OptVarInt(value));
     }
 
+    /**
+     * The enum contains all variants from a frog which are currently implemented in the game.
+     */
     public enum Variant {
         TEMPERATE,
         WARM,
         COLD;
 
-        private final static FrogMeta.Variant[] VALUES = values();
+        private static final FrogMeta.Variant[] VALUES = values();
+
+        // Microtus start - meta update
+        /**
+         * Add method to get a variant from a frog over the ordinal id
+         * @param id the ordinal id
+         * @return the entry which matches with the id
+         */
+        public static @Nullable FrogMeta.Variant getVariant(int id) {
+            return id >= 0 && id <= VALUES.length ? VALUES[id] : null;
+        }
+        // Microtus end - meta update
     }
 }
diff --git a/src/main/java/net/minestom/server/entity/metadata/animal/GoatMeta.java b/src/main/java/net/minestom/server/entity/metadata/animal/GoatMeta.java
index f830d79b46083bd671eaee0ba964c3d483f1b48f..6d78f6cf69a410d58317945439070c7a9b7dac85 100644
--- a/src/main/java/net/minestom/server/entity/metadata/animal/GoatMeta.java
+++ b/src/main/java/net/minestom/server/entity/metadata/animal/GoatMeta.java
@@ -6,7 +6,7 @@ import org.jetbrains.annotations.NotNull;
 
 public class GoatMeta extends AnimalMeta {
     public static final byte OFFSET = AnimalMeta.MAX_OFFSET;
-    public static final byte MAX_OFFSET = OFFSET + 1;
+    public static final byte MAX_OFFSET = OFFSET + 3;
 
     public GoatMeta(@NotNull Entity entity, @NotNull Metadata metadata) {
         super(entity, metadata);
@@ -19,4 +19,22 @@ public class GoatMeta extends AnimalMeta {
     public void setScreaming(boolean screaming) {
         metadata.setIndex(OFFSET, Metadata.Boolean(screaming));
     }
+
+    // Microtus start - meta update
+    public boolean hasLeftHorn() {
+        return metadata.getIndex(OFFSET + 1, true);
+    }
+
+    public void setLeftHorn(boolean leftHorn) {
+        metadata.setIndex(OFFSET + 1, Metadata.Boolean(leftHorn));
+    }
+
+    public boolean hasRightHorn() {
+        return metadata.getIndex(OFFSET + 2, true);
+    }
+
+    public void setRightHorn(boolean rightHorn) {
+        metadata.setIndex(OFFSET + 2, Metadata.Boolean(rightHorn));
+    }
+    // Microtus end - meta update
 }
diff --git a/src/main/java/net/minestom/server/entity/metadata/monster/WardenMeta.java b/src/main/java/net/minestom/server/entity/metadata/monster/WardenMeta.java
new file mode 100644
index 0000000000000000000000000000000000000000..4354c9ce62a2a9053ed880d094134cdae4357cd9
--- /dev/null
+++ b/src/main/java/net/minestom/server/entity/metadata/monster/WardenMeta.java
@@ -0,0 +1,43 @@
+package net.minestom.server.entity.metadata.monster;
+
+import net.minestom.server.entity.Entity;
+import net.minestom.server.entity.Metadata;
+import org.jetbrains.annotations.NotNull;
+
+// Microtus - entity meta update
+/**
+ * This metadata implementation can be used for a warden entity.
+ * @author theEvilReaper
+ * @version 1.0.0
+ * @since 1.0.0
+ **/
+public class WardenMeta extends MonsterMeta {
+
+    public static final byte OFFSET = MonsterMeta.MAX_OFFSET;
+    public static final byte MAX_OFFSET = OFFSET + 1;
+
+    /**
+     * Creates a new metadata which can be set to a warden entity.
+     * @param entity the warden entity reference
+     * @param metadata the reference to a {@link Metadata}
+     */
+    public WardenMeta(@NotNull Entity entity, @NotNull Metadata metadata) {
+        super(entity, metadata);
+    }
+
+    /**
+     * Set the anger level for a warden.
+     * @param value the level to set
+     */
+    public void setAngerLevel(int value) {
+        super.metadata.setIndex(OFFSET, Metadata.VarInt(value));
+    }
+
+    /**
+     * Returns the given anger level from the meta or zero as default value if no value is set.
+     * @return the given level
+     */
+    public int getAngerLevel() {
+        return super.metadata.getIndex(OFFSET, 0);
+    }
+}
diff --git a/src/main/java/net/minestom/server/entity/metadata/monster/raider/PillagerMeta.java b/src/main/java/net/minestom/server/entity/metadata/monster/raider/PillagerMeta.java
index 5755ed4b1f6f1073698856bdc565b3e7943f1044..1cf1168ecd108e70e3bc827c4a6aa244897ae8e8 100644
--- a/src/main/java/net/minestom/server/entity/metadata/monster/raider/PillagerMeta.java
+++ b/src/main/java/net/minestom/server/entity/metadata/monster/raider/PillagerMeta.java
@@ -6,10 +6,27 @@ import org.jetbrains.annotations.NotNull;
 
 public class PillagerMeta extends AbstractIllagerMeta {
     public static final byte OFFSET = AbstractIllagerMeta.MAX_OFFSET;
-    public static final byte MAX_OFFSET = OFFSET + 0;
+    public static final byte MAX_OFFSET = OFFSET + 1;
 
     public PillagerMeta(@NotNull Entity entity, @NotNull Metadata metadata) {
         super(entity, metadata);
     }
 
+    // Microtus start - meta update
+    /**
+     * Set the state if a Pillager charge his crossbow.
+     * @param value the value to set
+     */
+    public void setChargingCrossbow(boolean value) {
+        super.metadata.setIndex(OFFSET, Metadata.Boolean(value));
+    }
+
+    /**
+     * Returns a boolean value if a Pillager is charging his crossbow.
+     * @return true when yes otherwise false
+     */
+    public boolean isChargingCrossbow() {
+        return super.metadata.getIndex(OFFSET, false);
+    }
+    // Microtus end - meta update
 }
diff --git a/src/main/java/net/minestom/server/entity/metadata/other/BoatMeta.java b/src/main/java/net/minestom/server/entity/metadata/other/BoatMeta.java
index 6be4483182ed904913c9e76c8cf58baa733db9a0..313774679c7c4921bf89216c23a31cc30dc92a7b 100644
--- a/src/main/java/net/minestom/server/entity/metadata/other/BoatMeta.java
+++ b/src/main/java/net/minestom/server/entity/metadata/other/BoatMeta.java
@@ -76,9 +76,13 @@ public class BoatMeta extends EntityMeta {
         BIRCH,
         JUNGLE,
         ACACIA,
-        DARK_OAK;
+        // Microtus start - meta update
+        CHERRY,
+        DARK_OAK,
+        MANGROVE,
+        BAMBOO;
+        // Microtus end - meta update
 
-        private final static Type[] VALUES = values();
+        private static final Type[] VALUES = values();
     }
-
 }
diff --git a/src/main/java/net/minestom/server/entity/metadata/water/DolphinMeta.java b/src/main/java/net/minestom/server/entity/metadata/water/DolphinMeta.java
index 54902744760ff77762bfdab3f5a71cf9117b35fa..1e39f9b4c70bbaa74b7aec1769619a2ab46948c4 100644
--- a/src/main/java/net/minestom/server/entity/metadata/water/DolphinMeta.java
+++ b/src/main/java/net/minestom/server/entity/metadata/water/DolphinMeta.java
@@ -6,37 +6,66 @@ import net.minestom.server.entity.Entity;
 import net.minestom.server.entity.Metadata;
 import org.jetbrains.annotations.NotNull;
 
+// Microtus start - meta update
 public class DolphinMeta extends WaterAnimalMeta {
+
     public static final byte OFFSET = WaterAnimalMeta.MAX_OFFSET;
     public static final byte MAX_OFFSET = OFFSET + 3;
 
+    /**
+     * Creates a new meta data for a dolphin.
+     * @param entity the involved entity
+     * @param metadata the base metadata
+     */
     public DolphinMeta(@NotNull Entity entity, @NotNull Metadata metadata) {
         super(entity, metadata);
     }
 
-    @NotNull
-    public Point getTreasurePosition() {
+    /**
+     * Get the given destination point for a treasure.
+     * @return the point or {@link Vec#ZERO} when no point is set
+     */
+    public @NotNull Point getTreasurePosition() {
         return super.metadata.getIndex(OFFSET, Vec.ZERO);
     }
 
+    /**
+     * Set the destination point to lead a player to a treasure.
+     * @param value the point to set
+     */
     public void setTreasurePosition(@NotNull Point value) {
         super.metadata.setIndex(OFFSET, Metadata.Position(value));
     }
 
-    public boolean isCanFindTreasure() {
+    /**
+     * Returns an indicator if a Dolphin received a fish from a player.
+     * @return true when yes otherwise false
+     */
+    public boolean hasFish() {
         return super.metadata.getIndex(OFFSET + 1, false);
     }
 
-    public void setCanFindTreasure(boolean value) {
+    /**
+     * Set the indicator if a Dolphin got a fish from a player.
+     * @param value the value to set
+     */
+    public void setHasFish(boolean value) {
         super.metadata.setIndex(OFFSET + 1, Metadata.Boolean(value));
     }
 
-    public boolean isHasFish() {
-        return super.metadata.getIndex(OFFSET + 2, false);
+    /**
+     * Get the current moisture level.
+     * @return the given level
+     */
+    public int getMoistureLevel() {
+        return super.metadata.getIndex(OFFSET + 2, 2400);
     }
 
-    public void setHasFish(boolean value) {
-        super.metadata.setIndex(OFFSET + 2, Metadata.Boolean(value));
+    /**
+     * Updates the given moisture level.
+     * @param level the level to set
+     */
+    public void setMoistureLevel(int level) {
+        super.metadata.setIndex(OFFSET + 2, Metadata.VarInt(level));
     }
-
 }
diff --git a/src/main/java/net/minestom/server/entity/metadata/water/fish/TadpoleMeta.java b/src/main/java/net/minestom/server/entity/metadata/water/fish/TadpoleMeta.java
new file mode 100644
index 0000000000000000000000000000000000000000..ad9152ca44eb6521458aa20a4484278c9334614e
--- /dev/null
+++ b/src/main/java/net/minestom/server/entity/metadata/water/fish/TadpoleMeta.java
@@ -0,0 +1,21 @@
+package net.minestom.server.entity.metadata.water.fish;
+
+import net.minestom.server.entity.Entity;
+import net.minestom.server.entity.Metadata;
+import org.jetbrains.annotations.NotNull;
+
+// Microtus - entity meta update
+/**
+ * Metadata implementation for a Tadpole entity.
+ */
+public final class TadpoleMeta extends AbstractFishMeta {
+
+    /**
+     * Creates a new reference from this meta.
+     * @param entity the involved entity
+     * @param metadata the involved metadata
+     */
+    public TadpoleMeta(@NotNull Entity entity, @NotNull Metadata metadata) {
+        super(entity, metadata);
+    }
+}
diff --git a/src/main/java/net/minestom/server/network/NetworkBufferTypes.java b/src/main/java/net/minestom/server/network/NetworkBufferTypes.java
index 7d7f4cab472337ffcc522e5c4a56dfdfa2db74e2..b5df65dce1976421d60597fc82e72f9fd02a952a 100644
--- a/src/main/java/net/minestom/server/network/NetworkBufferTypes.java
+++ b/src/main/java/net/minestom/server/network/NetworkBufferTypes.java
@@ -571,7 +571,7 @@ final class NetworkBufferTypes {
             },
             buffer -> {
                 final int ordinal = buffer.read(VAR_INT);
-                return FrogMeta.Variant.values()[ordinal];
+                return FrogMeta.Variant.getVariant(ordinal); // Microtus - meta update
             });
     // Microtus start - 1.19.4 update
     static final TypeImpl<SnifferMeta.State> SNIFFER_STATE = new TypeImpl<>(SnifferMeta.State.class,
diff --git a/src/test/java/net/minestom/server/entity/metadata/animal/FrogMetaTest.java b/src/test/java/net/minestom/server/entity/metadata/animal/FrogMetaTest.java
new file mode 100644
index 0000000000000000000000000000000000000000..dff4f20dac5cfb5dc02c815ef381f91fc093fa97
--- /dev/null
+++ b/src/test/java/net/minestom/server/entity/metadata/animal/FrogMetaTest.java
@@ -0,0 +1,18 @@
+package net.minestom.server.entity.metadata.animal;
+
+import net.minestom.server.entity.metadata.animal.FrogMeta.Variant;
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+// Microtus - meta update + add test for the variant read method
+class FrogMetaTest {
+
+    @Test
+    void testFrogVariantRead() {
+        assertEquals(Variant.WARM, Variant.getVariant(Variant.WARM.ordinal()));
+        assertEquals(Variant.TEMPERATE, Variant.getVariant(0));
+        assertNull(Variant.getVariant(-1));
+        assertNull(Variant.getVariant(100));
+    }
+}
\ No newline at end of file
