From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Phillipp Glanz <6745190+TheMeinerLP@users.noreply.github.com>
Date: Mon, 8 Jan 2024 13:42:41 +0100
Subject: [PATCH] Add better biome system


diff --git a/code-generators/src/main/java/net/minestom/codegen/Generators.java b/code-generators/src/main/java/net/minestom/codegen/Generators.java
index aae4d324f16ff4b5204d7bb3ffb110b43e40b1fc..a49a65590a80320e17bfa28461f891f0bb33ac70 100644
--- a/code-generators/src/main/java/net/minestom/codegen/Generators.java
+++ b/code-generators/src/main/java/net/minestom/codegen/Generators.java
@@ -1,5 +1,6 @@
 package net.minestom.codegen;
 
+import net.minestom.codegen.biome.BiomeGenerator;
 import net.minestom.codegen.color.DyeColorGenerator;
 import net.minestom.codegen.fluid.FluidGenerator;
 import org.slf4j.Logger;
@@ -21,6 +22,7 @@ public class Generators {
 
         // Generate DyeColors
         new DyeColorGenerator(resource("dye_colors.json"), outputFolder).generate();
+        new BiomeGenerator(resource("biomes.json"), outputFolder).generate();
 
 
         var generator = new CodeGenerator(outputFolder);
diff --git a/code-generators/src/main/java/net/minestom/codegen/MinestomCodeGenerator.java b/code-generators/src/main/java/net/minestom/codegen/MinestomCodeGenerator.java
index a56248de382badce98656ce4006452beb8990cc8..69eb7e67dd5253bca7e7948335a8f2f2acd849f1 100644
--- a/code-generators/src/main/java/net/minestom/codegen/MinestomCodeGenerator.java
+++ b/code-generators/src/main/java/net/minestom/codegen/MinestomCodeGenerator.java
@@ -3,6 +3,7 @@ package net.minestom.codegen;
 import com.google.gson.Gson;
 import com.google.gson.GsonBuilder;
 import com.squareup.javapoet.JavaFile;
+import java.util.regex.Pattern;
 import org.jetbrains.annotations.NotNull;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
@@ -15,6 +16,7 @@ import java.util.Locale;
 public abstract class MinestomCodeGenerator {
     private static final Logger LOGGER = LoggerFactory.getLogger(MinestomCodeGenerator.class);
     protected static final Gson GSON = new GsonBuilder().setPrettyPrinting().disableHtmlEscaping().create();
+    protected static final Pattern minecraftNamespaceRegex = Pattern.compile("minecraft:");
 
     public abstract void generate();
 
@@ -28,7 +30,8 @@ public abstract class MinestomCodeGenerator {
         }
     }
 
+
     protected static String toConstant(String namespace) {
-        return namespace.replace("minecraft:", "").toUpperCase(Locale.ROOT);
+        return minecraftNamespaceRegex.matcher(namespace).replaceAll("").toUpperCase(Locale.ROOT);
     }
 }
diff --git a/code-generators/src/main/java/net/minestom/codegen/biome/BiomeGenerator.java b/code-generators/src/main/java/net/minestom/codegen/biome/BiomeGenerator.java
new file mode 100644
index 0000000000000000000000000000000000000000..ca13484b90d51dd3c72b9369f299f7af20088505
--- /dev/null
+++ b/code-generators/src/main/java/net/minestom/codegen/biome/BiomeGenerator.java
@@ -0,0 +1,145 @@
+package net.minestom.codegen.biome;
+
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import com.squareup.javapoet.AnnotationSpec;
+import com.squareup.javapoet.ArrayTypeName;
+import com.squareup.javapoet.ClassName;
+import com.squareup.javapoet.CodeBlock;
+import com.squareup.javapoet.FieldSpec;
+import com.squareup.javapoet.JavaFile;
+import com.squareup.javapoet.MethodSpec;
+import com.squareup.javapoet.ParameterSpec;
+import com.squareup.javapoet.TypeName;
+import com.squareup.javapoet.TypeSpec;
+import java.awt.Color;
+import java.io.File;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.util.List;
+import javax.lang.model.element.Modifier;
+import net.minestom.codegen.MinestomCodeGenerator;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@ApiStatus.Internal
+@ApiStatus.NonExtendable
+public final class BiomeGenerator extends MinestomCodeGenerator {
+
+    private static final String BIOME_KEY = "biome";
+    private static final String BIOME_FILE_PACKAGE = "net.minestom.server.world.biomes"; // Microtus - Biomes
+    private static final String CLASS_NAME = "Biomes"; // Microtus - Biomes
+    private static final Logger LOGGER = LoggerFactory.getLogger(BiomeGenerator.class);
+    private final InputStream biomeFile;
+    private final File outputFolder;
+
+    public BiomeGenerator(@Nullable InputStream biomeFile, @NotNull File outputFolder) {
+        this.biomeFile = biomeFile;
+        this.outputFolder = outputFolder;
+    }
+
+    @Override
+    public void generate() {
+        if (biomeFile == null) {
+            LOGGER.error("Failed to find biome.json.");
+            LOGGER.error("Stopped code generation for biomes.");
+            return;
+        }
+        if (!outputFolder.exists() && !outputFolder.mkdirs()) {
+            LOGGER.error("Output folder for code generation does not exist and could not be created.");
+            return;
+        }
+
+        JsonElement biomes = GSON.fromJson(new InputStreamReader(biomeFile), JsonElement.class);
+        ClassName biomeCN = ClassName.get(BIOME_FILE_PACKAGE, CLASS_NAME);
+        ClassName biomeCn = ClassName.get(BIOME_FILE_PACKAGE, "Biome");
+        ClassName categoryCn = biomeCn.nestedClass("Category");
+        ClassName minecraftServerCn = ClassName.get("net.minestom.server", "MinecraftServer");
+        ClassName namespaceIDCn = ClassName.get("net.minestom.server.utils", "NamespaceID");
+        ClassName biomeEffectsCn = ClassName.get(BIOME_FILE_PACKAGE, "BiomeEffects");
+        // Dye Color Enum
+        TypeSpec.Builder biomesEnum = TypeSpec.enumBuilder(biomeCN)
+                .addModifiers(Modifier.PUBLIC).addJavadoc("AUTOGENERATED by " + getClass().getSimpleName());
+
+        // Fields
+        biomesEnum.addFields(
+                List.of(
+                        FieldSpec.builder(biomeCn, BIOME_KEY, Modifier.PRIVATE, Modifier.FINAL).build(),
+                        FieldSpec.builder(ArrayTypeName.of(biomeCN), "VALUES", Modifier.PRIVATE, Modifier.STATIC, Modifier.FINAL).initializer(CLASS_NAME + ".values()").build()  // Microtus - Banner and shield meta
+                )
+        );
+
+        // Methods
+        biomesEnum.addMethods(
+                List.of(
+                        // Constructor
+                        MethodSpec.constructorBuilder()
+                                .addParameter(
+                                        ParameterSpec.builder(biomeCn, BIOME_KEY).addAnnotation(NotNull.class).build()
+                                )
+                                .addStatement("this.$1L = $1L", BIOME_KEY)
+                                .build(),
+                        MethodSpec.methodBuilder(BIOME_KEY)
+                                .addModifiers(Modifier.PUBLIC)
+                                .returns(biomeCn.annotated(AnnotationSpec.builder(NotNull.class).build()))
+                                .addStatement("return this.$L", BIOME_KEY)
+                                .build(),
+                        MethodSpec.methodBuilder("getValue")
+                                .addModifiers(Modifier.PUBLIC, Modifier.STATIC)
+                                .addParameter(ParameterSpec.builder(TypeName.INT, "id").build())
+                                .addAnnotation(Nullable.class)
+                                .returns(ClassName.get(BIOME_FILE_PACKAGE, CLASS_NAME))
+                                .addCode("return VALUES[$L];", "id")
+                                .build(),
+                        MethodSpec.methodBuilder("registerBiomes")
+                                .addModifiers(Modifier.PUBLIC, Modifier.STATIC)
+                                .beginControlFlow("for (var $L : VALUES)", BIOME_KEY)
+                                .addStatement("$1T.getBiomeManager().addBiome($2L.$2L())", minecraftServerCn, BIOME_KEY)
+                                .endControlFlow()
+                                .build()
+                )
+        );
+        final JsonObject biomesAsJsonObject = biomes.getAsJsonObject();
+        biomesAsJsonObject.keySet().forEach(biome -> {
+            var biomeObject = biomesAsJsonObject.get(biome).getAsJsonObject();
+            var temperature = biomeObject.get("temperature").getAsFloat();
+            var downfall = biomeObject.get("downfall").getAsFloat();
+            var hasPrecipitation = biomeObject.get("has_precipitation").getAsBoolean();
+            var fogColor = Integer.toHexString(new Color(biomeObject.get("fogColor").getAsInt()).getRGB()).substring(2);
+            var waterColor = Integer.toHexString(new Color(biomeObject.get("waterColor").getAsInt()).getRGB()).substring(2);
+            var waterFogColor = Integer.toHexString(new Color(biomeObject.get("waterFogColor").getAsInt()).getRGB()).substring(2);
+            var skyColor = Integer.toHexString(new Color(biomeObject.get("skyColor").getAsInt()).getRGB()).substring(2);
+            var effectCodeBlock = CodeBlock.builder();
+            effectCodeBlock = effectCodeBlock.add("$T.builder().fogColor(0x$L).waterColor(0x$L).waterFogColor(0x$L).skyColor(0x$L)", biomeEffectsCn, fogColor, waterColor, waterFogColor, skyColor);
+            if (biomeObject.has("grassColor")) {
+                var grassColor = Integer.toHexString(new Color(biomeObject.get("grassColor").getAsInt()).getRGB()).substring(2);
+                effectCodeBlock = effectCodeBlock.add(".grassColor(0x$L)", grassColor);
+            }
+            if (biomeObject.has("foliageColor")) {
+                var foliageColor = Integer.toHexString(new Color(biomeObject.get("foliageColor").getAsInt()).getRGB()).substring(2);
+                effectCodeBlock = effectCodeBlock.add(".foliageColor(0x$L)", foliageColor);
+            }
+            effectCodeBlock = effectCodeBlock.add(".build()");
+            var namespacedCode = CodeBlock.of("$T.from($S)", namespaceIDCn, biome.replaceFirst("minecraft:", ""));
+            biomesEnum.addEnumConstant(toConstant(biome), TypeSpec.anonymousClassBuilder(
+                            "$T.builder().name($L).category($T.NONE).depth(0.125F).scale(0.05F).temperature($Lf).downfall($Lf).showPrecipitation($L).effects($L).build()",
+                            biomeCn, namespacedCode, categoryCn, temperature, downfall, hasPrecipitation, effectCodeBlock.build()
+                    ).build()
+            );
+        });
+
+        // Write files to outputFolder
+        writeFiles(
+                List.of(
+                        JavaFile.builder(BIOME_FILE_PACKAGE, biomesEnum.build())
+                                .indent("    ")
+                                .skipJavaLangImports(true)
+                                .build()
+                ),
+                outputFolder
+        );
+    }
+}
diff --git a/src/main/java/net/minestom/server/ServerFlag.java b/src/main/java/net/minestom/server/ServerFlag.java
index ea61869059d877f59ac9c613891c5c502c603198..3c126640a14d073e37562a66ab6eb3465be7c472 100644
--- a/src/main/java/net/minestom/server/ServerFlag.java
+++ b/src/main/java/net/minestom/server/ServerFlag.java
@@ -39,6 +39,9 @@ public final class ServerFlag {
     public static final @Nullable String STACKING_RULE = System.getProperty("minestom.stacking-rule");
     public static final int WORLD_BORDER_SIZE = Integer.getInteger("minestom.world-border-size", 29999984);
 
+    // Biomes
+    public static final boolean BIOMES_ENABLED = System.getProperty("minestom.attributes.disabled") == null;
+
     // Maps
     public static final @NotNull String MAP_RGB_MAPPING = System.getProperty("minestom.map.rgbmapping", "lazy");
     public static final @Nullable String MAP_RGB_REDUCTION = System.getProperty("minestom.map.rgbreduction"); // Only used if rgb mapping is "approximate"
diff --git a/src/main/java/net/minestom/server/ServerProcessImpl.java b/src/main/java/net/minestom/server/ServerProcessImpl.java
index fd5af2b71286497addf6a41f3df8d151dfb6720e..aac3f8749abfd970760049e81861bf29643c4608 100644
--- a/src/main/java/net/minestom/server/ServerProcessImpl.java
+++ b/src/main/java/net/minestom/server/ServerProcessImpl.java
@@ -44,6 +44,7 @@ import net.minestom.server.utils.PropertyUtils;
 import net.minestom.server.utils.collection.MappedCollection;
 import net.minestom.server.world.DimensionTypeManager;
 import net.minestom.server.world.biomes.BiomeManager;
+import net.minestom.server.world.biomes.Biomes;
 import org.jetbrains.annotations.NotNull;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
@@ -248,6 +249,11 @@ final class ServerProcessImpl implements ServerProcess {
 
         LOGGER.info(MinecraftServer.getBrandName() + " server started successfully.");
 
+        if (ServerFlag.BIOMES_ENABLED) {
+            Biomes.registerBiomes();
+        }
+        LOGGER.info("Register Biomes({})", biome.unmodifiableCollection().size());
+
         if (MinecraftServer.isTerminalEnabled()) {
             MinestomTerminal.start();
         }
diff --git a/src/main/java/net/minestom/server/world/biomes/BiomeParticle.java b/src/main/java/net/minestom/server/world/biomes/BiomeParticle.java
index f2f3edeb37600c2b603c292a3d7179f61848f0c2..ce20735425d344574ad17001f410bd41db1dbda3 100644
--- a/src/main/java/net/minestom/server/world/biomes/BiomeParticle.java
+++ b/src/main/java/net/minestom/server/world/biomes/BiomeParticle.java
@@ -1,70 +1,15 @@
 package net.minestom.server.world.biomes;
 
-import net.minestom.server.instance.block.Block;
-import net.minestom.server.item.ItemStack;
-import net.minestom.server.utils.NamespaceID;
+import net.minestom.server.world.biomes.particle.BiomeOption;
 import org.jglrxavpok.hephaistos.nbt.NBT;
 import org.jglrxavpok.hephaistos.nbt.NBTCompound;
 
 import java.util.Map;
 
-public record BiomeParticle(float probability, Option option) {
+public record BiomeParticle(float probability, BiomeOption option) {
     public NBTCompound toNbt() {
         return NBT.Compound(Map.of(
                 "probability", NBT.Float(probability),
                 "options", option.toNbt()));
     }
-
-    public interface Option {
-        NBTCompound toNbt();
-    }
-
-    public record BlockOption(Block block) implements Option {
-        //TODO also can be falling_dust
-        private static final String type = "block";
-
-        @Override
-        public NBTCompound toNbt() {
-            return NBT.Compound(nbtCompound -> {
-                nbtCompound.setString("type", type);
-                nbtCompound.setString("Name", block.name());
-                Map<String, String> propertiesMap = block.properties();
-                if (propertiesMap.size() != 0) {
-                    nbtCompound.set("Properties", NBT.Compound(p -> propertiesMap.forEach(p::setString)));
-                }
-            });
-        }
-    }
-
-    public record DustOption(float red, float green, float blue, float scale) implements Option {
-        private static final String type = "dust";
-
-        @Override
-        public NBTCompound toNbt() {
-            return NBT.Compound(Map.of(
-                    "type", NBT.String(type),
-                    "r", NBT.Float(red),
-                    "g", NBT.Float(green),
-                    "b", NBT.Float(blue),
-                    "scale", NBT.Float(scale)));
-        }
-    }
-
-    public record ItemOption(ItemStack item) implements Option {
-        private static final String type = "item";
-
-        @Override
-        public NBTCompound toNbt() {
-            //todo test count might be wrong type
-            NBTCompound nbtCompound = item.meta().toNBT();
-            return nbtCompound.modify(n -> n.setString("type", type));
-        }
-    }
-
-    public record NormalOption(NamespaceID type) implements Option {
-        @Override
-        public NBTCompound toNbt() {
-            return NBT.Compound(Map.of("type", NBT.String(type.toString())));
-        }
-    }
 }
diff --git a/src/main/java/net/minestom/server/world/biomes/particle/BiomeOption.java b/src/main/java/net/minestom/server/world/biomes/particle/BiomeOption.java
new file mode 100644
index 0000000000000000000000000000000000000000..07ffda2ba5d3b7664db1d1ee9f850a0504397636
--- /dev/null
+++ b/src/main/java/net/minestom/server/world/biomes/particle/BiomeOption.java
@@ -0,0 +1,16 @@
+package net.minestom.server.world.biomes.particle;
+
+import net.minestom.server.particle.Particle;
+import org.jetbrains.annotations.NotNull;
+import org.jglrxavpok.hephaistos.nbt.NBTCompound;
+
+/**
+ * Just a generic interface to hold data about biome particles
+ * @author TheMeinerLP
+ */
+public interface BiomeOption {
+    @NotNull
+    NBTCompound toNbt();
+    @NotNull
+    Particle getType();
+}
diff --git a/src/main/java/net/minestom/server/world/biomes/particle/BlockOption.java b/src/main/java/net/minestom/server/world/biomes/particle/BlockOption.java
new file mode 100644
index 0000000000000000000000000000000000000000..117d9a1ec79de53bf3e2df2a48c8a8a0010d3672
--- /dev/null
+++ b/src/main/java/net/minestom/server/world/biomes/particle/BlockOption.java
@@ -0,0 +1,32 @@
+package net.minestom.server.world.biomes.particle;
+
+import net.minestom.server.instance.block.Block;
+import net.minestom.server.particle.Particle;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.NotNull;
+import org.jglrxavpok.hephaistos.nbt.NBT;
+import org.jglrxavpok.hephaistos.nbt.NBTCompound;
+
+import java.util.Map;
+
+@ApiStatus.Experimental
+public record BlockOption(@NotNull Block block) implements BiomeOption {
+
+    @Override
+    public @NotNull NBTCompound toNbt() {
+        return NBT.Compound(nbtCompound -> {
+            nbtCompound.setString("type", getType().namespace().toString());
+            nbtCompound.setString("Name", block.name());
+            Map<String, String> propertiesMap = block.properties();
+            if (!propertiesMap.isEmpty()) {
+                nbtCompound.set("Properties", NBT.Compound(p -> propertiesMap.forEach(p::setString)));
+            }
+        });
+    }
+
+    @Override
+    public @NotNull Particle getType() {
+        return Particle.BLOCK;
+    }
+
+}
diff --git a/src/main/java/net/minestom/server/world/biomes/particle/DefaultOption.java b/src/main/java/net/minestom/server/world/biomes/particle/DefaultOption.java
new file mode 100644
index 0000000000000000000000000000000000000000..1ddad4d7ea400653786608669689a84ed3078143
--- /dev/null
+++ b/src/main/java/net/minestom/server/world/biomes/particle/DefaultOption.java
@@ -0,0 +1,20 @@
+package net.minestom.server.world.biomes.particle;
+
+import net.minestom.server.particle.Particle;
+import org.jetbrains.annotations.NotNull;
+import org.jglrxavpok.hephaistos.nbt.NBT;
+import org.jglrxavpok.hephaistos.nbt.NBTCompound;
+
+import java.util.Map;
+
+public record DefaultOption(@NotNull Particle particle) implements BiomeOption {
+    @Override
+    public @NotNull NBTCompound toNbt() {
+        return NBT.Compound(Map.of("type", NBT.String(getType().namespace().toString())));
+    }
+
+    @Override
+    public @NotNull Particle getType() {
+        return particle;
+    }
+}
diff --git a/src/main/java/net/minestom/server/world/biomes/particle/DustOption.java b/src/main/java/net/minestom/server/world/biomes/particle/DustOption.java
new file mode 100644
index 0000000000000000000000000000000000000000..0a84f495820afa685c08fd97bb454865ba976564
--- /dev/null
+++ b/src/main/java/net/minestom/server/world/biomes/particle/DustOption.java
@@ -0,0 +1,34 @@
+package net.minestom.server.world.biomes.particle;
+
+import net.minestom.server.particle.Particle;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.NotNull;
+import org.jglrxavpok.hephaistos.nbt.NBT;
+import org.jglrxavpok.hephaistos.nbt.NBTCompound;
+
+import java.util.Map;
+
+/**
+ * Particle effect which used based on the color values.
+ * @param red can between 0 and 1 and is divided by 255
+ * @param green can between 0 and 1 and is divided by 255
+ * @param blue can between 0 and 1 and is divided by 255
+ * @param scale can between 0.01 to 4
+ */
+@ApiStatus.Experimental
+public record DustOption(float red, float green, float blue, float scale) implements BiomeOption {
+    @Override
+    public @NotNull NBTCompound toNbt() {
+        return NBT.Compound(Map.of(
+                "type", NBT.String(getType().namespace().toString()),
+                "Red", NBT.Float(red),
+                "Green", NBT.Float(green),
+                "Blue", NBT.Float(blue),
+                "Scale", NBT.Float(scale)));
+    }
+
+    @Override
+    public @NotNull Particle getType() {
+        return Particle.DUST;
+    }
+}
diff --git a/src/main/java/net/minestom/server/world/biomes/particle/ItemOption.java b/src/main/java/net/minestom/server/world/biomes/particle/ItemOption.java
new file mode 100644
index 0000000000000000000000000000000000000000..a977cd052c10fd1b540b7934251d62e550361ca1
--- /dev/null
+++ b/src/main/java/net/minestom/server/world/biomes/particle/ItemOption.java
@@ -0,0 +1,27 @@
+package net.minestom.server.world.biomes.particle;
+
+import net.minestom.server.item.ItemStack;
+import net.minestom.server.particle.Particle;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.NotNull;
+import org.jglrxavpok.hephaistos.nbt.NBTCompound;
+
+/**
+ * Particle effect which used based on an item inside the biome.
+ * @param item to bes used
+ */
+@ApiStatus.Experimental
+public record ItemOption(@NotNull ItemStack item) implements BiomeOption {
+
+    @Override
+    public @NotNull NBTCompound toNbt() {
+        NBTCompound nbtCompound = item.meta().toNBT();
+        return nbtCompound.modify(n -> n.setString("type", getType().namespace().toString()));
+    }
+
+    @Override
+    public @NotNull Particle getType() {
+        return Particle.ITEM;
+    }
+
+}
diff --git a/src/test/java/net/minestom/server/instance/AnvilLoaderIntegrationTest.java b/src/test/java/net/minestom/server/instance/AnvilLoaderIntegrationTest.java
index d03ed2b6a9865c0eb5b3f6610f15733d259b2e55..ca4d776c70684d554d7f941b3fd3a33258f278e1 100644
--- a/src/test/java/net/minestom/server/instance/AnvilLoaderIntegrationTest.java
+++ b/src/test/java/net/minestom/server/instance/AnvilLoaderIntegrationTest.java
@@ -1,9 +1,11 @@
 package net.minestom.server.instance;
 
+import net.minestom.server.MinecraftServer;
 import net.minestom.server.instance.block.Block;
 import net.minestom.server.network.NetworkBuffer;
 import net.minestom.server.utils.NamespaceID;
 import net.minestom.server.world.biomes.Biome;
+import net.minestom.server.world.biomes.Biomes;
 import net.minestom.testing.Env;
 import net.minestom.testing.EnvTest;
 import org.junit.jupiter.api.AfterAll;
@@ -49,6 +51,7 @@ class AnvilLoaderIntegrationTest {
 
     @Test
     void loadHouse(Env env) {
+        MinecraftServer.getBiomeManager().addBiome(Biomes.PLAINS.biome());
         // load a world that contains only a basic house and make sure it is loaded properly
 
         AnvilLoader chunkLoader = new AnvilLoader(worldFolder) {
@@ -144,6 +147,7 @@ class AnvilLoaderIntegrationTest {
 
     @Test
     void loadAndSaveChunk(Env env) throws InterruptedException {
+        MinecraftServer.getBiomeManager().addBiome(Biomes.PLAINS.biome());
         Instance instance = env.createFlatInstance(new AnvilLoader(worldFolder) {
             // Force loads inside current thread
             @Override
