From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: theEvilReaper <steffenwx@gmail.com>
Date: Mon, 17 Jul 2023 20:35:27 +0200
Subject: [PATCH] Implement equals and hashcode


diff --git a/src/main/java/net/minestom/server/entity/Entity.java b/src/main/java/net/minestom/server/entity/Entity.java
index daf0ece5bc96be7862d265f89ae8edee6958fc13..ff0b45ffe8ac3ce607de0c4312b11451014a11a5 100644
--- a/src/main/java/net/minestom/server/entity/Entity.java
+++ b/src/main/java/net/minestom/server/entity/Entity.java
@@ -477,6 +477,21 @@ public class Entity implements Viewable, Tickable, Schedulable, Snapshotable, Ev
         return viewers;
     }
 
+    //Microtus start - equals and hashcode
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        Entity entity = (Entity) o;
+        return Objects.equals(uuid, entity.uuid);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(uuid);
+    }
+    //Microtus end - equals and hashcode
+
     /**
      * Gets if this entity's viewers (surrounding players) can be predicted from surrounding chunks.
      */
diff --git a/src/main/java/net/minestom/server/instance/Instance.java b/src/main/java/net/minestom/server/instance/Instance.java
index b7c5affa63ebe29d30996718c3402d8348137f7d..d6dd0549db22ee58c4057012fcc46a3aba8f087b 100644
--- a/src/main/java/net/minestom/server/instance/Instance.java
+++ b/src/main/java/net/minestom/server/instance/Instance.java
@@ -451,6 +451,21 @@ public abstract class Instance implements Block.Getter, Block.Setter,
         return new TimeUpdatePacket(worldAge, time);
     }
 
+    //Microtus start - equals and hashcode
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        Instance instance = (Instance) o;
+        return Objects.equals(uniqueId, instance.uniqueId);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(uniqueId);
+    }
+    //Microtus end - equals and hashcode
+
     /**
      * Gets the instance {@link WorldBorder};
      *
