From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: theEvilReaper <steffenwx@gmail.com>
Date: Wed, 26 Jul 2023 20:30:19 +0200
Subject: [PATCH] Add metadata implementation for some entities


diff --git a/src/main/java/net/minestom/server/entity/Entity.java b/src/main/java/net/minestom/server/entity/Entity.java
index ff0b45ffe8ac3ce607de0c4312b11451014a11a5..8e7ad8420da6168ea2444dae90e5bc68bb32c176 100644
--- a/src/main/java/net/minestom/server/entity/Entity.java
+++ b/src/main/java/net/minestom/server/entity/Entity.java
@@ -1747,6 +1747,7 @@ public class Entity implements Viewable, Tickable, Schedulable, Snapshotable, Ev
         DYING,
         CROAKING,
         USING_TONGUE,
+        SITTING, // Microtus - Add missing entity pose
         ROARING,
         SNIFFING,
         EMERGING,
diff --git a/src/main/java/net/minestom/server/entity/EntityTypeImpl.java b/src/main/java/net/minestom/server/entity/EntityTypeImpl.java
index 9f4cfa4a898ca348d60883ec140a874c2b9cd8fa..1bad471ee2fda90766e11ba3ab1f8ee9445bcbeb 100644
--- a/src/main/java/net/minestom/server/entity/EntityTypeImpl.java
+++ b/src/main/java/net/minestom/server/entity/EntityTypeImpl.java
@@ -33,10 +33,7 @@ import net.minestom.server.entity.metadata.water.AxolotlMeta;
 import net.minestom.server.entity.metadata.water.DolphinMeta;
 import net.minestom.server.entity.metadata.water.GlowSquidMeta;
 import net.minestom.server.entity.metadata.water.SquidMeta;
-import net.minestom.server.entity.metadata.water.fish.CodMeta;
-import net.minestom.server.entity.metadata.water.fish.PufferfishMeta;
-import net.minestom.server.entity.metadata.water.fish.SalmonMeta;
-import net.minestom.server.entity.metadata.water.fish.TropicalFishMeta;
+import net.minestom.server.entity.metadata.water.fish.*;
 import net.minestom.server.registry.Registry;
 import org.jetbrains.annotations.NotNull;
 
@@ -87,7 +84,7 @@ record EntityTypeImpl(Registry.EntityEntry registry) implements EntityType {
                 entry("minecraft:bee", BeeMeta::new),
                 entry("minecraft:blaze", BlazeMeta::new),
                 entry("minecraft:boat", BoatMeta::new),
-                entry("minecraft:chest_boat", EntityMeta::new), // TODO dedicated metadata
+                entry("minecraft:chest_boat", BoatMeta::new), // Microtus - entity meta update
                 entry("minecraft:cat", CatMeta::new),
                 entry("minecraft:block_display", BlockDisplayMeta::new), // Microtus - 1.19.4 update
                 entry("minecraft:cave_spider", CaveSpiderMeta::new),
@@ -175,7 +172,7 @@ record EntityTypeImpl(Registry.EntityEntry registry) implements EntityType {
                 entry("minecraft:squid", SquidMeta::new),
                 entry("minecraft:stray", StrayMeta::new),
                 entry("minecraft:strider", StriderMeta::new),
-                entry("minecraft:tadpole", EntityMeta::new), // TODO dedicated metadata
+                entry("minecraft:tadpole", TadpoleMeta::new), // Microtus - entity meta update
                 entry("minecraft:egg", ThrownEggMeta::new),
                 entry("minecraft:ender_pearl", ThrownEnderPearlMeta::new),
                 entry("minecraft:experience_bottle", ThrownExperienceBottleMeta::new),
@@ -188,7 +185,7 @@ record EntityTypeImpl(Registry.EntityEntry registry) implements EntityType {
                 entry("minecraft:villager", VillagerMeta::new),
                 entry("minecraft:vindicator", VindicatorMeta::new),
                 entry("minecraft:wandering_trader", WanderingTraderMeta::new),
-                entry("minecraft:warden", EntityMeta::new), // TODO dedicated metadata
+                entry("minecraft:warden", WardenMeta::new), // Microtus - entity meta update
                 entry("minecraft:witch", WitchMeta::new),
                 entry("minecraft:wither", WitherMeta::new),
                 entry("minecraft:wither_skeleton", WitherSkeletonMeta::new),
@@ -200,7 +197,10 @@ record EntityTypeImpl(Registry.EntityEntry registry) implements EntityType {
                 entry("minecraft:zombie_villager", ZombieVillagerMeta::new),
                 entry("minecraft:zombified_piglin", ZombifiedPiglinMeta::new),
                 entry("minecraft:player", PlayerMeta::new),
-                entry("minecraft:fishing_bobber", FishingHookMeta::new)
+                entry("minecraft:fishing_bobber", FishingHookMeta::new),
+                // Microtus start - 1.20 update
+                entry("minecraft:camel", CamelMeta::new) // Microtus - entity meta update
+                // Microtus end - 1.20 update
         );
     }
 }
diff --git a/src/main/java/net/minestom/server/entity/metadata/monster/WardenMeta.java b/src/main/java/net/minestom/server/entity/metadata/monster/WardenMeta.java
new file mode 100644
index 0000000000000000000000000000000000000000..4354c9ce62a2a9053ed880d094134cdae4357cd9
--- /dev/null
+++ b/src/main/java/net/minestom/server/entity/metadata/monster/WardenMeta.java
@@ -0,0 +1,43 @@
+package net.minestom.server.entity.metadata.monster;
+
+import net.minestom.server.entity.Entity;
+import net.minestom.server.entity.Metadata;
+import org.jetbrains.annotations.NotNull;
+
+// Microtus - entity meta update
+/**
+ * This metadata implementation can be used for a warden entity.
+ * @author theEvilReaper
+ * @version 1.0.0
+ * @since 1.0.0
+ **/
+public class WardenMeta extends MonsterMeta {
+
+    public static final byte OFFSET = MonsterMeta.MAX_OFFSET;
+    public static final byte MAX_OFFSET = OFFSET + 1;
+
+    /**
+     * Creates a new metadata which can be set to a warden entity.
+     * @param entity the warden entity reference
+     * @param metadata the reference to a {@link Metadata}
+     */
+    public WardenMeta(@NotNull Entity entity, @NotNull Metadata metadata) {
+        super(entity, metadata);
+    }
+
+    /**
+     * Set the anger level for a warden.
+     * @param value the level to set
+     */
+    public void setAngerLevel(int value) {
+        super.metadata.setIndex(OFFSET, Metadata.VarInt(value));
+    }
+
+    /**
+     * Returns the given anger level from the meta or zero as default value if no value is set.
+     * @return the given level
+     */
+    public int getAngerLevel() {
+        return super.metadata.getIndex(OFFSET, 0);
+    }
+}
diff --git a/src/main/java/net/minestom/server/entity/metadata/water/fish/TadpoleMeta.java b/src/main/java/net/minestom/server/entity/metadata/water/fish/TadpoleMeta.java
new file mode 100644
index 0000000000000000000000000000000000000000..ad9152ca44eb6521458aa20a4484278c9334614e
--- /dev/null
+++ b/src/main/java/net/minestom/server/entity/metadata/water/fish/TadpoleMeta.java
@@ -0,0 +1,21 @@
+package net.minestom.server.entity.metadata.water.fish;
+
+import net.minestom.server.entity.Entity;
+import net.minestom.server.entity.Metadata;
+import org.jetbrains.annotations.NotNull;
+
+// Microtus - entity meta update
+/**
+ * Metadata implementation for a Tadpole entity.
+ */
+public final class TadpoleMeta extends AbstractFishMeta {
+
+    /**
+     * Creates a new reference from this meta.
+     * @param entity the involved entity
+     * @param metadata the involved metadata
+     */
+    public TadpoleMeta(@NotNull Entity entity, @NotNull Metadata metadata) {
+        super(entity, metadata);
+    }
+}
