From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: theEvilReaper <steffenwx@gmail.com>
Date: Mon, 17 Jul 2023 20:35:27 +0200
Subject: [PATCH] Implement equals and hashcode


diff --git a/src/main/java/net/minestom/server/entity/Entity.java b/src/main/java/net/minestom/server/entity/Entity.java
index daf0ece5bc96be7862d265f89ae8edee6958fc13..ff0b45ffe8ac3ce607de0c4312b11451014a11a5 100644
--- a/src/main/java/net/minestom/server/entity/Entity.java
+++ b/src/main/java/net/minestom/server/entity/Entity.java
@@ -477,6 +477,21 @@ public class Entity implements Viewable, Tickable, Schedulable, Snapshotable, Ev
         return viewers;
     }
 
+    //Microtus start - equals and hashcode
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        Entity entity = (Entity) o;
+        return Objects.equals(uuid, entity.uuid);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(uuid);
+    }
+    //Microtus end - equals and hashcode
+
     /**
      * Gets if this entity's viewers (surrounding players) can be predicted from surrounding chunks.
      */
diff --git a/src/main/java/net/minestom/server/instance/Instance.java b/src/main/java/net/minestom/server/instance/Instance.java
index 58bf4617c2a3b03f9c67a17573edc656be097f28..d6ef6fa393cb47481e998da81978d472aecd385f 100644
--- a/src/main/java/net/minestom/server/instance/Instance.java
+++ b/src/main/java/net/minestom/server/instance/Instance.java
@@ -460,6 +460,21 @@ public abstract class Instance implements Block.Getter, Block.Setter,
         return new TimeUpdatePacket(worldAge, time);
     }
 
+    //Microtus start - equals and hashcode
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        Instance instance = (Instance) o;
+        return Objects.equals(uniqueId, instance.uniqueId);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(uniqueId);
+    }
+    //Microtus end - equals and hashcode
+
     /**
      * Gets the instance {@link WorldBorder};
      *
diff --git a/src/test/java/net/minestom/server/entity/EntityEqualsTest.java b/src/test/java/net/minestom/server/entity/EntityEqualsTest.java
new file mode 100644
index 0000000000000000000000000000000000000000..e11088a617e6da7b058b6c6a8a8ffb311ceebe48
--- /dev/null
+++ b/src/test/java/net/minestom/server/entity/EntityEqualsTest.java
@@ -0,0 +1,13 @@
+package net.minestom.server.entity;
+
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+class EntityEqualsTest {
+
+    @Test
+    void testEntityEquals() {
+        assertNotEquals(new Entity(EntityTypes.BEE), new Entity(EntityTypes.BEE));
+    }
+}
diff --git a/src/test/java/net/minestom/server/instance/InstanceEqualsTest.java b/src/test/java/net/minestom/server/instance/InstanceEqualsTest.java
new file mode 100644
index 0000000000000000000000000000000000000000..305be82a87d514f18510d13364f0574f193a093a
--- /dev/null
+++ b/src/test/java/net/minestom/server/instance/InstanceEqualsTest.java
@@ -0,0 +1,16 @@
+package net.minestom.server.instance;
+
+import net.minestom.testing.Env;
+import net.minestom.testing.EnvTest;
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+@EnvTest
+class InstanceEqualsTest {
+
+    @Test
+    void testInstanceEquals(Env env) {
+        assertNotEquals(env.createFlatInstance(), env.createFlatInstance());
+    }
+}
