From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Phillipp Glanz <p.glanz@madfix.me>
Date: Thu, 30 Nov 2023 13:00:31 +0100
Subject: [PATCH] Improve network handling


diff --git a/src/main/java/net/minestom/server/entity/Entity.java b/src/main/java/net/minestom/server/entity/Entity.java
index 8e7ad8420da6168ea2444dae90e5bc68bb32c176..3241e9577c38e301096cebf96cb72816d52bdcff 100644
--- a/src/main/java/net/minestom/server/entity/Entity.java
+++ b/src/main/java/net/minestom/server/entity/Entity.java
@@ -1340,7 +1340,10 @@ public class Entity implements Viewable, Tickable, Schedulable, Snapshotable, Ev
                     lastSyncedPosition, onGround), this);
         } else if (viewChange) {
             PacketUtils.prepareViewablePacket(chunk, new EntityHeadLookPacket(getEntityId(), position.yaw()), this);
-            PacketUtils.prepareViewablePacket(chunk, new EntityRotationPacket(getEntityId(), position.yaw(), position.pitch(), onGround), this);
+            // Microtus start - position sync fix
+            PacketUtils.prepareViewablePacket(chunk, EntityPositionAndRotationPacket.getPacket(getEntityId(), position,
+                    lastSyncedPosition, isOnGround()), this);
+            // Microtus end - position sync fix
         }
         this.lastSyncedPosition = position;
     }
diff --git a/src/main/java/net/minestom/server/network/NetworkBufferTypes.java b/src/main/java/net/minestom/server/network/NetworkBufferTypes.java
index 10d00f747f8f9247c0120d9952f89250662f3d15..8044908bb1e5109eb9437bba2ce109889358400d 100644
--- a/src/main/java/net/minestom/server/network/NetworkBufferTypes.java
+++ b/src/main/java/net/minestom/server/network/NetworkBufferTypes.java
@@ -237,7 +237,13 @@ final class NetworkBufferTypes {
                     buffer.nbtWriter = nbtWriter;
                 }
                 try {
-                    nbtWriter.writeNamed("", value);
+                    // Microtus start - update nbt sending and reading
+                    if (value == NBTEnd.INSTANCE) {
+                        buffer.write(BYTE, (byte) NBTType.TAG_End.getOrdinal());
+                    } else {
+                        nbtWriter.writeNamed("", value);
+                    }
+                    // Microtus end - update nbt sending and reading
                 } catch (IOException e) {
                     throw new RuntimeException(e);
                 }
@@ -323,7 +329,11 @@ final class NetworkBufferTypes {
                 buffer.write(BOOLEAN, true);
                 buffer.write(VAR_INT, value.material().id());
                 buffer.write(BYTE, (byte) value.amount());
-                buffer.write(NBT, value.meta().toNBT());
+                // Microtus start - update nbt send when the item has not data
+                // Vanilla does not write an empty object, just an end tag.
+                NBTCompound nbtData = value.meta().toNBT();
+                buffer.write(NBT, nbtData.isEmpty() ? NBTEnd.INSTANCE : nbtData);
+                // Microtus end - update nbt sending and reading
                 return -1;
             },
             buffer -> {
