From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: theEvilReaper <steffenwx@gmail.com>
Date: Thu, 10 Aug 2023 20:41:02 +0200
Subject: [PATCH] Update deprecated methods


diff --git a/src/main/java/net/minestom/server/advancements/notifications/Notification.java b/src/main/java/net/minestom/server/advancements/notifications/Notification.java
index 90660466e837326cb98d26032dd04b4f19247295..7b6b1169b4b65dd02d4b1f66900517995622f681 100644
--- a/src/main/java/net/minestom/server/advancements/notifications/Notification.java
+++ b/src/main/java/net/minestom/server/advancements/notifications/Notification.java
@@ -14,12 +14,22 @@ public record Notification(@NotNull Component title, @NotNull FrameType frameTyp
         this(title, frameType, ItemStack.of(icon));
     }
 
-    @Deprecated
+    /**
+     * Returns the given title as {@link Component} of the notification.
+     * @deprecated use the {@link #icon()} method instead.
+     * @return the icon of the notification
+     */
+    @Deprecated(forRemoval = true, since = "1.1.3") //Microtus - Update deprecation status
     public @NotNull Component getTitle() {
         return title;
     }
 
-    @Deprecated
+    /**
+     * Returns the given {@link FrameType} of the notification.
+     * @deprecated use the {@link #icon()} method instead.
+     * @return the icon of the notification
+     */
+    @Deprecated(forRemoval = true, since = "1.1.3") //Microtus - Update deprecation status
     public @NotNull FrameType getFrameType() {
         return frameType;
     }
diff --git a/src/main/java/net/minestom/server/adventure/AdventurePacketConvertor.java b/src/main/java/net/minestom/server/adventure/AdventurePacketConvertor.java
index 04244d43a556e33b5ce9cb4cf727433f976af35b..2d9fe8a7a8ebb0d6ef5af7ece430d769c158c164 100644
--- a/src/main/java/net/minestom/server/adventure/AdventurePacketConvertor.java
+++ b/src/main/java/net/minestom/server/adventure/AdventurePacketConvertor.java
@@ -10,7 +10,6 @@ import net.kyori.adventure.text.Component;
 import net.kyori.adventure.text.format.NamedTextColor;
 import net.kyori.adventure.title.Title;
 import net.kyori.adventure.title.TitlePart;
-import net.minestom.server.coordinate.Pos;
 import net.minestom.server.coordinate.Vec;
 import net.minestom.server.entity.Entity;
 import net.minestom.server.network.packet.server.ServerPacket;
@@ -150,7 +149,7 @@ public class AdventurePacketConvertor {
      * @return the packet
      * @deprecated Use {@link #createSoundPacket(Sound, Sound.Emitter)}
      */
-    @Deprecated(forRemoval = true)
+    @Deprecated(forRemoval = true, since = "1.1.3") //Microtus - Update deprecation status
     public static ServerPacket createEntitySoundPacket(@NotNull Sound sound, @NotNull Entity entity) {
         return createSoundPacket(sound, entity);
     }
diff --git a/src/main/java/net/minestom/server/color/Color.java b/src/main/java/net/minestom/server/color/Color.java
index 9fdaf30d8553660a75555228cbe63fcf99445fa6..9733f0b72d33b2866bf8ccf3782a5547840b420a 100644
--- a/src/main/java/net/minestom/server/color/Color.java
+++ b/src/main/java/net/minestom/server/color/Color.java
@@ -103,17 +103,32 @@ public record Color(int red, int green, int blue) implements RGBLike {
         return new Color(r, g, b);
     }
 
-    @Deprecated
+    /**
+     * Returns the red component of the color.
+     * @deprecated use {@link #red()} method instead.
+     * @return the red component of this color
+     */
+    @Deprecated(forRemoval = true, since = "1.1.3") //Microtus - Update deprecation status
     public int getRed() {
         return this.red;
     }
 
-    @Deprecated
+    /**
+     * Returns the green component of the color.
+     * @deprecated use the {@link #green()} method instead.
+     * @return the green component of this color
+     */
+    @Deprecated(forRemoval = true, since = "1.1.3") //Microtus - Update deprecation status
     public int getGreen() {
         return this.green;
     }
 
-    @Deprecated
+    /**
+     * Returns the blue component of the color.
+     * @deprecated use the {@link #blue()} method instead.
+     * @return the blue component of this color
+     */
+    @Deprecated(forRemoval = true, since = "1.1.3") //Microtus - Update deprecation status
     public int getBlue() {
         return this.blue;
     }
diff --git a/src/main/java/net/minestom/server/command/CommandSender.java b/src/main/java/net/minestom/server/command/CommandSender.java
index 20a22bac01f21fb14a5a2be6353d3fbb0abacb0d..5f09d7bc52fae017f8f10f6731e1f7818f14fbce 100644
--- a/src/main/java/net/minestom/server/command/CommandSender.java
+++ b/src/main/java/net/minestom/server/command/CommandSender.java
@@ -42,7 +42,7 @@ public interface CommandSender extends PermissionHandler, Audience, Taggable, Id
      *
      * @return true if 'this' is a player, false otherwise
      */
-    @Deprecated
+    @Deprecated(forRemoval = true, since = "1.1.3")
     default boolean isPlayer() {
         return false;
     }
@@ -54,7 +54,7 @@ public interface CommandSender extends PermissionHandler, Audience, Taggable, Id
      *
      * @return true if 'this' is the console, false otherwise
      */
-    @Deprecated
+    @Deprecated(forRemoval = true, since = "1.1.3")
     default boolean isConsole() {
         return false;
     }
@@ -66,7 +66,7 @@ public interface CommandSender extends PermissionHandler, Audience, Taggable, Id
      * @throws ClassCastException if 'this' is not a player
      * @see #isPlayer()
      */
-    @Deprecated
+    @Deprecated(forRemoval = true, since = "1.1.3")
     default Player asPlayer() {
         throw new ClassCastException("CommandSender is not a Player");
     }
@@ -78,7 +78,7 @@ public interface CommandSender extends PermissionHandler, Audience, Taggable, Id
      * @throws ClassCastException if 'this' is not a console sender
      * @see #isConsole()
      */
-    @Deprecated
+    @Deprecated(forRemoval = true, since = "1.1.3")
     default ConsoleSender asConsole() {
         throw new ClassCastException("CommandSender is not the ConsoleSender");
     }
diff --git a/src/main/java/net/minestom/server/entity/PlayerSkin.java b/src/main/java/net/minestom/server/entity/PlayerSkin.java
index fe85cf8ad30097aaef147977d8b59b163fc75760..72798e5ec55531a26217a15fa82b18af072408e7 100644
--- a/src/main/java/net/minestom/server/entity/PlayerSkin.java
+++ b/src/main/java/net/minestom/server/entity/PlayerSkin.java
@@ -64,7 +64,7 @@ public record PlayerSkin(String textures, String signature) {
     /**
      * @deprecated use {@link #textures()}
      */
-    @Deprecated
+    @Deprecated(forRemoval = true, since = "1.1.3") //Microtus - Update deprecation status
     public String getTextures() {
         return textures;
     }
@@ -72,7 +72,7 @@ public record PlayerSkin(String textures, String signature) {
     /**
      * @deprecated use {@link #signature()}
      */
-    @Deprecated
+    @Deprecated(forRemoval = true, since = "1.1.3") //Microtus - Update deprecation status
     public String getSignature() {
         return signature;
     }
diff --git a/src/main/java/net/minestom/server/entity/ai/target/ClosestEntityTarget.java b/src/main/java/net/minestom/server/entity/ai/target/ClosestEntityTarget.java
index 3775f1bf346d97b0a0f915f9ff9eb93698a128f3..c451537f5638b47f9a55212ad51aeda8764205ec 100644
--- a/src/main/java/net/minestom/server/entity/ai/target/ClosestEntityTarget.java
+++ b/src/main/java/net/minestom/server/entity/ai/target/ClosestEntityTarget.java
@@ -25,7 +25,7 @@ public class ClosestEntityTarget extends TargetSelector {
      * @deprecated Use {@link #ClosestEntityTarget(EntityCreature, double, Predicate)}
      */
     @SafeVarargs
-    @Deprecated
+    @Deprecated(forRemoval = true, since = "1.1.3") //Microtus - Update deprecation status
     public ClosestEntityTarget(@NotNull EntityCreature entityCreature, float range,
                                @NotNull Class<? extends LivingEntity>... entitiesTarget) {
         this(entityCreature, range, ent -> {
diff --git a/src/main/java/net/minestom/server/event/player/PlayerEatEvent.java b/src/main/java/net/minestom/server/event/player/PlayerEatEvent.java
index b45665fb719606ade2bc83318ba25866487a6dc4..54fd191ee02827e82946e2cf82533ee93ab94f59 100644
--- a/src/main/java/net/minestom/server/event/player/PlayerEatEvent.java
+++ b/src/main/java/net/minestom/server/event/player/PlayerEatEvent.java
@@ -27,7 +27,7 @@ public class PlayerEatEvent implements ItemEvent, PlayerInstanceEvent {
      * @return the food item
      * @deprecated use getItemStack() for the eaten item
      */
-    @Deprecated
+    @Deprecated(forRemoval = true, since = "1.1.3") //Microtus - Update deprecation status
     public @NotNull ItemStack getFoodItem() {
         return foodItem;
     }
diff --git a/src/main/java/net/minestom/server/event/player/PlayerPreEatEvent.java b/src/main/java/net/minestom/server/event/player/PlayerPreEatEvent.java
index 27de8eee60012568cfd85e612f9a7377085f9033..5511e29017aa54958b62f2e6f1fcfc53a3757a92 100644
--- a/src/main/java/net/minestom/server/event/player/PlayerPreEatEvent.java
+++ b/src/main/java/net/minestom/server/event/player/PlayerPreEatEvent.java
@@ -34,7 +34,7 @@ public class PlayerPreEatEvent implements ItemEvent, PlayerInstanceEvent, Cancel
      * @return the food item
      * @deprecated use getItemStack() for the eaten item
      */
-    @Deprecated
+    @Deprecated(forRemoval = true, since = "1.1.3") //Microtus - Update deprecation status
     public @NotNull ItemStack getFoodItem() {
         return foodItem;
     }
diff --git a/src/main/java/net/minestom/server/inventory/Inventory.java b/src/main/java/net/minestom/server/inventory/Inventory.java
index 9d80f6331c1320603680411d1eff3d710c38e6c3..ab57e8814738acb22cd9bfc8565a449038f1c4d8 100644
--- a/src/main/java/net/minestom/server/inventory/Inventory.java
+++ b/src/main/java/net/minestom/server/inventory/Inventory.java
@@ -53,9 +53,18 @@ public non-sealed class Inventory extends AbstractInventory implements Viewable
         this.offset = getSize();
     }
 
+    //Microtus start - Update deprecation status
+    /**
+     * Creates a new Inventory reference with a given size and title.
+     * The method is deprecated please use {@link Inventory#Inventory(InventoryType, Component)} instead.
+     * @param inventoryType the inventory type
+     * @param title the inventory title
+     */
+    @Deprecated(forRemoval = true, since = "1.1.3")
     public Inventory(@NotNull InventoryType inventoryType, @NotNull String title) {
         this(inventoryType, Component.text(title));
     }
+    //Microtus end - Update deprecation status
 
     private static byte generateId() {
         return (byte) ID_COUNTER.updateAndGet(i -> i + 1 >= 128 ? 1 : i + 1);
diff --git a/src/main/java/net/minestom/server/inventory/InventoryType.java b/src/main/java/net/minestom/server/inventory/InventoryType.java
index 6108c4283fc1128fddd208914095cd7739524f0f..4667b7c3a3fe0a49c3b4e00210c103155d85ea39 100644
--- a/src/main/java/net/minestom/server/inventory/InventoryType.java
+++ b/src/main/java/net/minestom/server/inventory/InventoryType.java
@@ -47,7 +47,7 @@ public enum InventoryType {
     /**
      * @deprecated use {@link #getSize()}
      */
-    @Deprecated
+    @Deprecated(forRemoval = true, since = "1.1.3") //Microtus - Update deprecation status
     public int getAdditionalSlot() {
         return size;
     }
diff --git a/src/main/java/net/minestom/server/inventory/type/AnvilInventory.java b/src/main/java/net/minestom/server/inventory/type/AnvilInventory.java
index 8a7c9f2c154e22d3df90c8ee20856bb0b2f3d4de..e8b36d90398e88b220bc3e8a50a2572803658404 100644
--- a/src/main/java/net/minestom/server/inventory/type/AnvilInventory.java
+++ b/src/main/java/net/minestom/server/inventory/type/AnvilInventory.java
@@ -14,9 +14,17 @@ public class AnvilInventory extends Inventory {
         super(InventoryType.ANVIL, title);
     }
 
+    //Microtus start - Update deprecation status
+    /**
+     * Creates a new {@link AnvilInventory} reference with a given size and title.
+     * The method is deprecated please use {@link AnvilInventory#(InventoryType, Component)} instead.
+     * @param title the inventory title
+     */
+    @Deprecated(forRemoval = true, since = "1.1.3")
     public AnvilInventory(@NotNull String title) {
         super(InventoryType.ANVIL, title);
     }
+    //Microtus end - Update deprecation status
 
     /**
      * Gets the anvil repair cost.
diff --git a/src/main/java/net/minestom/server/inventory/type/BeaconInventory.java b/src/main/java/net/minestom/server/inventory/type/BeaconInventory.java
index 30551dae67cde1d869a4bd6b8645d0db317cf1a4..6d87a30775ae31f3036bbae9a04a4e3fcb110781 100644
--- a/src/main/java/net/minestom/server/inventory/type/BeaconInventory.java
+++ b/src/main/java/net/minestom/server/inventory/type/BeaconInventory.java
@@ -17,9 +17,17 @@ public class BeaconInventory extends Inventory {
         super(InventoryType.BEACON, title);
     }
 
+    //Microtus start - Update deprecation status
+    /**
+     * Creates a new {@link BeaconInventory} reference with a given size and title.
+     * The method is deprecated please use {@link BeaconInventory#(InventoryType, Component)} instead.
+     * @param title the inventory title
+     */
+    @Deprecated(forRemoval = true, since = "1.1.3")
     public BeaconInventory(@NotNull String title) {
         super(InventoryType.BEACON, title);
     }
+    //Microtus end - Update deprecation status
 
     /**
      * Gets the beacon power level.
diff --git a/src/main/java/net/minestom/server/inventory/type/BrewingStandInventory.java b/src/main/java/net/minestom/server/inventory/type/BrewingStandInventory.java
index ed0bc7d783d17dcae72343138ffc3d36fad5b4d9..9b8536d92b6f86bb3636447dfb9ee52ad098f722 100644
--- a/src/main/java/net/minestom/server/inventory/type/BrewingStandInventory.java
+++ b/src/main/java/net/minestom/server/inventory/type/BrewingStandInventory.java
@@ -15,9 +15,17 @@ public class BrewingStandInventory extends Inventory {
         super(InventoryType.BREWING_STAND, title);
     }
 
+    //Microtus start - Update deprecation status
+    /**
+     * Creates a new {@link BrewingStandInventory} reference with a given size and title.
+     * The method is deprecated please use {@link BrewingStandInventory#(InventoryType, Component)} instead.
+     * @param title the inventory title
+     */
+    @Deprecated(forRemoval = true, since = "1.1.3")
     public BrewingStandInventory(@NotNull String title) {
         super(InventoryType.BREWING_STAND, title);
     }
+    //Microtus end - Update deprecation status
 
     /**
      * Gets the brewing stand brew time.
diff --git a/src/main/java/net/minestom/server/inventory/type/EnchantmentTableInventory.java b/src/main/java/net/minestom/server/inventory/type/EnchantmentTableInventory.java
index bd855c32a5171683627bf2729ab13ef0acebd934..3406dff2c0bf3f8fb8344d24d8a2af45700c288b 100644
--- a/src/main/java/net/minestom/server/inventory/type/EnchantmentTableInventory.java
+++ b/src/main/java/net/minestom/server/inventory/type/EnchantmentTableInventory.java
@@ -18,9 +18,17 @@ public class EnchantmentTableInventory extends Inventory {
         super(InventoryType.ENCHANTMENT, title);
     }
 
+    //Microtus start - Update deprecation status
+    /**
+     * Creates a new {@link EnchantmentTableInventory} reference with a given size and title.
+     * The method is deprecated please use {@link EnchantmentTableInventory#(InventoryType, Component)} instead.
+     * @param title the inventory title
+     */
+    @Deprecated(forRemoval = true, since = "1.1.3")
     public EnchantmentTableInventory(@NotNull String title) {
         super(InventoryType.ENCHANTMENT, title);
     }
+    //Microtus end - Update deprecation status
 
     /**
      * Gets the level requirement in a slot.
diff --git a/src/main/java/net/minestom/server/inventory/type/FurnaceInventory.java b/src/main/java/net/minestom/server/inventory/type/FurnaceInventory.java
index 9e357da9936b8eb3df5aa9e5d07ef6a39297a792..6af660d141332ab08e64de56f9cd60800249b030 100644
--- a/src/main/java/net/minestom/server/inventory/type/FurnaceInventory.java
+++ b/src/main/java/net/minestom/server/inventory/type/FurnaceInventory.java
@@ -17,9 +17,17 @@ public class FurnaceInventory extends Inventory {
         super(InventoryType.FURNACE, title);
     }
 
+    //Microtus start - Update deprecation status
+    /**
+     * Creates a new {@link FurnaceInventory} reference with a given size and title.
+     * The method is deprecated please use {@link FurnaceInventory#(InventoryType, Component)} instead.
+     * @param title the inventory title
+     */
+    @Deprecated(forRemoval = true, since = "1.1.3")
     public FurnaceInventory(@NotNull String title) {
         super(InventoryType.FURNACE, title);
     }
+    //Microtus end - Update deprecation status
 
     /**
      * Represents the amount of tick until the fire icon come empty.
diff --git a/src/main/java/net/minestom/server/inventory/type/VillagerInventory.java b/src/main/java/net/minestom/server/inventory/type/VillagerInventory.java
index 048fa58139363ac64588da40c50f25ebaa46eb0e..5c95004d2b09bb6906feb00d062add17369ffc7a 100644
--- a/src/main/java/net/minestom/server/inventory/type/VillagerInventory.java
+++ b/src/main/java/net/minestom/server/inventory/type/VillagerInventory.java
@@ -24,9 +24,17 @@ public class VillagerInventory extends Inventory {
         super(InventoryType.MERCHANT, title);
     }
 
+    //Microtus start - Update deprecation status
+    /**
+     * Creates a new {@link VillagerInventory} reference with a given size and title.
+     * The method is deprecated please use {@link VillagerInventory#(InventoryType, Component)} instead.
+     * @param title the inventory title
+     */
+    @Deprecated(forRemoval = true, since = "1.1.3")
     public VillagerInventory(@NotNull String title) {
         super(InventoryType.MERCHANT, title);
     }
+    //Microtus end - Update deprecation status
 
     public List<TradeListPacket.Trade> getTrades() {
         return Collections.unmodifiableList(trades);
diff --git a/src/main/java/net/minestom/server/item/ItemStack.java b/src/main/java/net/minestom/server/item/ItemStack.java
index 1642dbf93669d5c2b54bfb03741b4186ce5ff057..0e3f2a48d1a1c75368fc0c33905cc5b07cb27251 100644
--- a/src/main/java/net/minestom/server/item/ItemStack.java
+++ b/src/main/java/net/minestom/server/item/ItemStack.java
@@ -181,19 +181,19 @@ public sealed interface ItemStack extends TagReadable, HoverEventSource<HoverEve
     @NotNull NBTCompound toItemNBT();
 
 
-    @Deprecated
+    @Deprecated(forRemoval = true, since = "1.1.3") //Microtus - Update deprecation status
     @Contract(pure = true)
     default @NotNull Material getMaterial() {
         return material();
     }
 
-    @Deprecated
+    @Deprecated(forRemoval = true, since = "1.1.3") //Microtus - Update deprecation status
     @Contract(pure = true)
     default int getAmount() {
         return amount();
     }
 
-    @Deprecated
+    @Deprecated(forRemoval = true, since = "1.1.3") //Microtus - Update deprecation status
     @Contract(pure = true)
     default @NotNull ItemMeta getMeta() {
         return meta();
diff --git a/src/main/java/net/minestom/server/ping/ResponseData.java b/src/main/java/net/minestom/server/ping/ResponseData.java
index 706d9a7579f7da51ff0de98829c69d8b0534cd89..693edb9b906f347c81bda3d769ac2f17ed9db072 100644
--- a/src/main/java/net/minestom/server/ping/ResponseData.java
+++ b/src/main/java/net/minestom/server/ping/ResponseData.java
@@ -50,7 +50,7 @@ public class ResponseData {
      * @param name The name for the response data.
      * @deprecated This is named incorrectly, use {@link #setVersion(String)} instead
      */
-    @Deprecated
+    @Deprecated(forRemoval = true, since = "1.1.3") //Microtus - Update deprecation status
     public void setName(String name) {
         this.setVersion(name);
     }
@@ -133,7 +133,7 @@ public class ResponseData {
      * @param players the players
      * @deprecated Use {@link #addEntries(Collection)}}
      */
-    @Deprecated
+    @Deprecated(forRemoval = true, since = "1.1.3") //Microtus - Update deprecation status
     public void addPlayer(Iterable<Player> players) {
         for (Player player : players) {
             this.addPlayer(player);
@@ -146,7 +146,7 @@ public class ResponseData {
      * @param player the player
      * @deprecated Use {@link #addEntry(NamedAndIdentified)}
      */
-    @Deprecated
+    @Deprecated(forRemoval = true, since = "1.1.3") //Microtus - Update deprecation status
     public void addPlayer(Player player) {
         this.addEntry(player);
     }
@@ -158,7 +158,7 @@ public class ResponseData {
      * @param uuid The unique identifier of the player.
      * @deprecated Use {@link #addEntry(NamedAndIdentified)} with {@link NamedAndIdentified#of(String, UUID)}
      */
-    @Deprecated
+    @Deprecated(forRemoval = true, since = "1.1.3") //Microtus - Update deprecation status
     public void addPlayer(String name, UUID uuid) {
         this.addEntry(NamedAndIdentified.of(name, uuid));
     }
@@ -171,7 +171,7 @@ public class ResponseData {
      * @param name The name of the player.
      * @deprecated Use {@link #addEntry(NamedAndIdentified)} with {@link NamedAndIdentified#named(String)}
      */
-    @Deprecated
+    @Deprecated(forRemoval = true, since = "1.1.3") //Microtus - Update deprecation status
     public void addPlayer(String name) {
         this.addEntry(NamedAndIdentified.named(name));
     }
@@ -182,7 +182,7 @@ public class ResponseData {
      *
      * @deprecated Use {@link #clearEntries()}
      */
-    @Deprecated
+    @Deprecated(forRemoval = true, since = "1.1.3") //Microtus - Update deprecation status
     public void clearPlayers() {
         this.clearEntries();
     }
@@ -205,7 +205,7 @@ public class ResponseData {
      * @param description The description for the response data.
      * @deprecated Use {@link #setDescription(Component)}
      */
-    @Deprecated
+    @Deprecated(forRemoval = true, since = "1.1.3") //Microtus - Update deprecation status
     public void setDescription(String description) {
         this.description = LegacyComponentSerializer.legacySection().deserialize(description);
     }
diff --git a/src/main/java/net/minestom/server/scoreboard/BelowNameTag.java b/src/main/java/net/minestom/server/scoreboard/BelowNameTag.java
index b5817c96772fd399320bc61cafecf2101a55e867..b5c67427644748d5199ac42757c7360e5cb82d21 100644
--- a/src/main/java/net/minestom/server/scoreboard/BelowNameTag.java
+++ b/src/main/java/net/minestom/server/scoreboard/BelowNameTag.java
@@ -32,7 +32,7 @@ public class BelowNameTag implements Scoreboard {
      * @param value The value of the scoreboard
      * @deprecated Use {@link #BelowNameTag(String, Component)}
      */
-    @Deprecated
+    @Deprecated(forRemoval = true, since = "1.1.3") //Microtus - Update deprecation status
     public BelowNameTag(String name, String value) {
         this(name, Component.text(value));
     }
diff --git a/src/main/java/net/minestom/server/scoreboard/Scoreboard.java b/src/main/java/net/minestom/server/scoreboard/Scoreboard.java
index b14aebca46465331160051e55ceaab6439999ce4..9a1ad120480b8cf11106915185ef63cf8d445a3b 100644
--- a/src/main/java/net/minestom/server/scoreboard/Scoreboard.java
+++ b/src/main/java/net/minestom/server/scoreboard/Scoreboard.java
@@ -24,7 +24,7 @@ public interface Scoreboard extends Viewable, PacketGroupingAudience {
      * @return the creation objective packet
      * @deprecated Use {@link #getCreationObjectivePacket(Component, ScoreboardObjectivePacket.Type)}
      */
-    @Deprecated
+    @Deprecated(forRemoval = true, since = "1.1.3") //Microtus - Update deprecation status
     @NotNull
     default ScoreboardObjectivePacket getCreationObjectivePacket(String value, ScoreboardObjectivePacket.Type type) {
         return this.getCreationObjectivePacket(Component.text(value), type);
diff --git a/src/main/java/net/minestom/server/scoreboard/Sidebar.java b/src/main/java/net/minestom/server/scoreboard/Sidebar.java
index 09f1a7d2332a0eaf243c54f96d1e6c94b276e57c..efc5627faddcc6b688437df7febb6e2593348e93 100644
--- a/src/main/java/net/minestom/server/scoreboard/Sidebar.java
+++ b/src/main/java/net/minestom/server/scoreboard/Sidebar.java
@@ -58,7 +58,7 @@ public class Sidebar implements Scoreboard {
      * @param title The title of the sidebar
      * @deprecated Use {@link #Sidebar(Component)}
      */
-    @Deprecated
+    @Deprecated(forRemoval = true, since = "1.1.3") //Microtus - Update deprecation status
     public Sidebar(@NotNull String title) {
         this(Component.text(title));
     }
@@ -85,7 +85,7 @@ public class Sidebar implements Scoreboard {
      * @param title The new sidebar title
      * @deprecated Use {@link #setTitle(Component)}
      */
-    @Deprecated
+    @Deprecated(forRemoval = true, since = "1.1.3") //Microtus - Update deprecation status
     public void setTitle(@NotNull String title) {
         this.setTitle(Component.text(title));
     }
diff --git a/src/main/java/net/minestom/server/utils/NamespaceID.java b/src/main/java/net/minestom/server/utils/NamespaceID.java
index e68ff90079bea1debfa543b14f6df39396359234..988eb3f876c4720243876bdbea45e4ff5e4a6629 100644
--- a/src/main/java/net/minestom/server/utils/NamespaceID.java
+++ b/src/main/java/net/minestom/server/utils/NamespaceID.java
@@ -113,12 +113,12 @@ public final class NamespaceID implements CharSequence, Key {
         return this.full;
     }
 
-    @Deprecated
+    @Deprecated(forRemoval = true, since = "1.1.3") //Microtus - Update deprecation status
     public String getDomain() {
         return domain();
     }
 
-    @Deprecated
+    @Deprecated(forRemoval = true, since = "1.1.3") //Microtus - Update deprecation status
     public String getPath() {
         return path();
     }
