From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: theEvilReaper <steffenwx@gmail.com>
Date: Fri, 8 Dec 2023 13:09:07 +0100
Subject: [PATCH] Remove deprecated methods which already have a replacement


diff --git a/src/main/java/net/minestom/server/advancements/notifications/Notification.java b/src/main/java/net/minestom/server/advancements/notifications/Notification.java
index 7b6b1169b4b65dd02d4b1f66900517995622f681..15fee4d25369b8b3e76cfae3eccaa2e634e14821 100644
--- a/src/main/java/net/minestom/server/advancements/notifications/Notification.java
+++ b/src/main/java/net/minestom/server/advancements/notifications/Notification.java
@@ -9,28 +9,9 @@ import org.jetbrains.annotations.NotNull;
 /**
  * Represents a message which can be sent using the {@link NotificationCenter}.
  */
+// Microtus - remove deprecated methods
 public record Notification(@NotNull Component title, @NotNull FrameType frameType, @NotNull ItemStack icon) {
     public Notification(@NotNull Component title, @NotNull FrameType frameType, @NotNull Material icon) {
         this(title, frameType, ItemStack.of(icon));
     }
-
-    /**
-     * Returns the given title as {@link Component} of the notification.
-     * @deprecated use the {@link #icon()} method instead.
-     * @return the icon of the notification
-     */
-    @Deprecated(forRemoval = true, since = "1.1.3") //Microtus - Update deprecation status
-    public @NotNull Component getTitle() {
-        return title;
-    }
-
-    /**
-     * Returns the given {@link FrameType} of the notification.
-     * @deprecated use the {@link #icon()} method instead.
-     * @return the icon of the notification
-     */
-    @Deprecated(forRemoval = true, since = "1.1.3") //Microtus - Update deprecation status
-    public @NotNull FrameType getFrameType() {
-        return frameType;
-    }
 }
diff --git a/src/main/java/net/minestom/server/adventure/AdventurePacketConvertor.java b/src/main/java/net/minestom/server/adventure/AdventurePacketConvertor.java
index 2d9fe8a7a8ebb0d6ef5af7ece430d769c158c164..ded5231b0ac28efbebd5ff86e8a0f03a1eee5f7f 100644
--- a/src/main/java/net/minestom/server/adventure/AdventurePacketConvertor.java
+++ b/src/main/java/net/minestom/server/adventure/AdventurePacketConvertor.java
@@ -23,6 +23,7 @@ import java.util.Collection;
 /**
  * Utility methods to convert adventure enums to their packet values.
  */
+// Microtus - remove deprecated methods
 public class AdventurePacketConvertor {
     private static final Object2IntMap<NamedTextColor> NAMED_TEXT_COLOR_ID_MAP = new Object2IntArrayMap<>(16);
 
@@ -141,19 +142,6 @@ public class AdventurePacketConvertor {
         }
     }
 
-    /**
-     * Creates an entity sound packet from an Adventure sound.
-     *
-     * @param sound  the sound
-     * @param entity the entity the sound is coming from
-     * @return the packet
-     * @deprecated Use {@link #createSoundPacket(Sound, Sound.Emitter)}
-     */
-    @Deprecated(forRemoval = true, since = "1.1.3") //Microtus - Update deprecation status
-    public static ServerPacket createEntitySoundPacket(@NotNull Sound sound, @NotNull Entity entity) {
-        return createSoundPacket(sound, entity);
-    }
-
     /**
      * Creates a sound stop packet from a sound stop.
      *
diff --git a/src/main/java/net/minestom/server/color/Color.java b/src/main/java/net/minestom/server/color/Color.java
index 9733f0b72d33b2866bf8ccf3782a5547840b420a..2418b648c18f95bdea57ce56a3247234b4ac9620 100644
--- a/src/main/java/net/minestom/server/color/Color.java
+++ b/src/main/java/net/minestom/server/color/Color.java
@@ -10,6 +10,7 @@ import org.jetbrains.annotations.NotNull;
  * <p>
  * Colors must be in the range of 0-255.
  */
+// Microtus - remove deprecated methods
 public record Color(int red, int green, int blue) implements RGBLike {
     private static final int BIT_MASK = 0xff;
 
@@ -102,34 +103,4 @@ public record Color(int red, int green, int blue) implements RGBLike {
         b = Math.round(averageBlue * gainFactor);
         return new Color(r, g, b);
     }
-
-    /**
-     * Returns the red component of the color.
-     * @deprecated use {@link #red()} method instead.
-     * @return the red component of this color
-     */
-    @Deprecated(forRemoval = true, since = "1.1.3") //Microtus - Update deprecation status
-    public int getRed() {
-        return this.red;
-    }
-
-    /**
-     * Returns the green component of the color.
-     * @deprecated use the {@link #green()} method instead.
-     * @return the green component of this color
-     */
-    @Deprecated(forRemoval = true, since = "1.1.3") //Microtus - Update deprecation status
-    public int getGreen() {
-        return this.green;
-    }
-
-    /**
-     * Returns the blue component of the color.
-     * @deprecated use the {@link #blue()} method instead.
-     * @return the blue component of this color
-     */
-    @Deprecated(forRemoval = true, since = "1.1.3") //Microtus - Update deprecation status
-    public int getBlue() {
-        return this.blue;
-    }
 }
diff --git a/src/main/java/net/minestom/server/command/CommandSender.java b/src/main/java/net/minestom/server/command/CommandSender.java
index 5f09d7bc52fae017f8f10f6731e1f7818f14fbce..a0c10efbbc2e53d3ea2815faf6f8840cb144e73f 100644
--- a/src/main/java/net/minestom/server/command/CommandSender.java
+++ b/src/main/java/net/minestom/server/command/CommandSender.java
@@ -34,52 +34,4 @@ public interface CommandSender extends PermissionHandler, Audience, Taggable, Id
             sendMessage(message);
         }
     }
-
-    /**
-     * Gets if the sender is a {@link Player}.
-     * <p>
-     * Consider using {@code instanceof} instead.
-     *
-     * @return true if 'this' is a player, false otherwise
-     */
-    @Deprecated(forRemoval = true, since = "1.1.3")
-    default boolean isPlayer() {
-        return false;
-    }
-
-    /**
-     * Gets if the sender is a {@link ConsoleSender}.
-     * <p>
-     * Consider using {@code instanceof} instead.
-     *
-     * @return true if 'this' is the console, false otherwise
-     */
-    @Deprecated(forRemoval = true, since = "1.1.3")
-    default boolean isConsole() {
-        return false;
-    }
-
-    /**
-     * Casts this object to a {@link Player}.
-     * No checks are performed, {@link ClassCastException} can very much happen.
-     *
-     * @throws ClassCastException if 'this' is not a player
-     * @see #isPlayer()
-     */
-    @Deprecated(forRemoval = true, since = "1.1.3")
-    default Player asPlayer() {
-        throw new ClassCastException("CommandSender is not a Player");
-    }
-
-    /**
-     * Casts this object to a {@link ConsoleSender}.
-     * No checks are performed, {@link ClassCastException} can very much happen.
-     *
-     * @throws ClassCastException if 'this' is not a console sender
-     * @see #isConsole()
-     */
-    @Deprecated(forRemoval = true, since = "1.1.3")
-    default ConsoleSender asConsole() {
-        throw new ClassCastException("CommandSender is not the ConsoleSender");
-    }
 }
diff --git a/src/main/java/net/minestom/server/command/ConsoleSender.java b/src/main/java/net/minestom/server/command/ConsoleSender.java
index d97bc82ae1cd4f52268d16b394274d6a31b31656..c8f7392ec3fc95af089be33ff676182215637f68 100644
--- a/src/main/java/net/minestom/server/command/ConsoleSender.java
+++ b/src/main/java/net/minestom/server/command/ConsoleSender.java
@@ -45,16 +45,6 @@ public class ConsoleSender implements CommandSender {
         return permissions;
     }
 
-    @Override
-    public boolean isConsole() {
-        return true;
-    }
-
-    @Override
-    public ConsoleSender asConsole() {
-        return this;
-    }
-
     @Override
     public @NotNull TagHandler tagHandler() {
         return tagHandler;
diff --git a/src/main/java/net/minestom/server/entity/Player.java b/src/main/java/net/minestom/server/entity/Player.java
index 2b810253fc8992c659d306de2848df739cca432f..49bdbbfe174381fb8a7e468b8a24a1910b687e4b 100644
--- a/src/main/java/net/minestom/server/entity/Player.java
+++ b/src/main/java/net/minestom/server/entity/Player.java
@@ -2115,16 +2115,6 @@ public class Player extends LivingEntity implements CommandSender, Localizable,
         this.identity = Identity.identity(uuid);
     }
 
-    @Override
-    public boolean isPlayer() {
-        return true;
-    }
-
-    @Override
-    public Player asPlayer() {
-        return this;
-    }
-
     protected void sendChunkUpdates(Chunk newChunk) {
         if (chunkUpdateLimitChecker.addToHistory(newChunk)) {
             final int newX = newChunk.getChunkX();
diff --git a/src/main/java/net/minestom/server/entity/PlayerSkin.java b/src/main/java/net/minestom/server/entity/PlayerSkin.java
index 72798e5ec55531a26217a15fa82b18af072408e7..7230c9b2d052493b08e6df99b738c6ede5d112f5 100644
--- a/src/main/java/net/minestom/server/entity/PlayerSkin.java
+++ b/src/main/java/net/minestom/server/entity/PlayerSkin.java
@@ -14,6 +14,7 @@ import org.jetbrains.annotations.Nullable;
  * Can be applied to a player with {@link Player#setSkin(PlayerSkin)}
  * or in the linked event {@link net.minestom.server.event.player.PlayerSkinInitEvent}.
  */
+// Microtus - remove deprecated methods
 public record PlayerSkin(String textures, String signature) {
 
     /**
@@ -60,20 +61,4 @@ public record PlayerSkin(String textures, String signature) {
             return null;
         }
     }
-
-    /**
-     * @deprecated use {@link #textures()}
-     */
-    @Deprecated(forRemoval = true, since = "1.1.3") //Microtus - Update deprecation status
-    public String getTextures() {
-        return textures;
-    }
-
-    /**
-     * @deprecated use {@link #signature()}
-     */
-    @Deprecated(forRemoval = true, since = "1.1.3") //Microtus - Update deprecation status
-    public String getSignature() {
-        return signature;
-    }
 }
diff --git a/src/main/java/net/minestom/server/entity/ai/target/ClosestEntityTarget.java b/src/main/java/net/minestom/server/entity/ai/target/ClosestEntityTarget.java
index c451537f5638b47f9a55212ad51aeda8764205ec..1b05107e945189da3e963917f01fc028515f2f4b 100644
--- a/src/main/java/net/minestom/server/entity/ai/target/ClosestEntityTarget.java
+++ b/src/main/java/net/minestom/server/entity/ai/target/ClosestEntityTarget.java
@@ -13,32 +13,12 @@ import java.util.function.Predicate;
 /**
  * Target the closest targetable entity (based on the target predicate)
  */
+// Microtus - remove deprecated methods
 public class ClosestEntityTarget extends TargetSelector {
 
     private final double range;
     private final Predicate<Entity> targetPredicate;
 
-    /**
-     * @param entityCreature the entity (self)
-     * @param range          the maximum range the entity can target others within
-     * @param entitiesTarget the entities to target by class
-     * @deprecated Use {@link #ClosestEntityTarget(EntityCreature, double, Predicate)}
-     */
-    @SafeVarargs
-    @Deprecated(forRemoval = true, since = "1.1.3") //Microtus - Update deprecation status
-    public ClosestEntityTarget(@NotNull EntityCreature entityCreature, float range,
-                               @NotNull Class<? extends LivingEntity>... entitiesTarget) {
-        this(entityCreature, range, ent -> {
-            Class<? extends Entity> clazz = ent.getClass();
-            for (Class<? extends LivingEntity> targetClass : entitiesTarget) {
-                if (targetClass.isAssignableFrom(clazz)) {
-                    return true;
-                }
-            }
-            return false;
-        });
-    }
-
     /**
      * @param entityCreature  the entity (self)
      * @param range           the maximum range the entity can target others within
diff --git a/src/main/java/net/minestom/server/entity/metadata/LivingEntityMeta.java b/src/main/java/net/minestom/server/entity/metadata/LivingEntityMeta.java
index 60d7102b205bc04a60678fecc2da0009ce56437a..de4ff6d92dc8965fc6f0c7f77994623adacba55e 100644
--- a/src/main/java/net/minestom/server/entity/metadata/LivingEntityMeta.java
+++ b/src/main/java/net/minestom/server/entity/metadata/LivingEntityMeta.java
@@ -76,28 +76,6 @@ public class LivingEntityMeta extends EntityMeta {
         super.metadata.setIndex(OFFSET + 4, Metadata.VarInt(value));
     }
 
-    /**
-     * @deprecated
-     * This returns the bee stinger count, not the absorption heart count
-     * Use {@link #getBeeStingerCount()} instead
-     * @return The number of bee stingers in this entity
-     */
-    @Deprecated
-    public int getHealthAddedByAbsorption() {
-        return super.metadata.getIndex(OFFSET + 5, 0);
-    }
-
-    /**
-     * @deprecated
-     * This sets the bee stinger count, not the absorption heart count
-     * Use {@link #setBeeStingerCount(int)} instead
-     * @param value The number of bee stingers for this entity to have
-     */
-    @Deprecated
-    public void setHealthAddedByAbsorption(int value) {
-        super.metadata.setIndex(OFFSET + 5, Metadata.VarInt(value));
-    }
-
     /**
      * Gets the amount of bee stingers in this entity
      * @return The amount of bee stingers
@@ -122,5 +100,4 @@ public class LivingEntityMeta extends EntityMeta {
     public void setBedInWhichSleepingPosition(@Nullable Point value) {
         super.metadata.setIndex(OFFSET + 6, Metadata.OptPosition(value));
     }
-
 }
diff --git a/src/main/java/net/minestom/server/event/player/PlayerEatEvent.java b/src/main/java/net/minestom/server/event/player/PlayerEatEvent.java
index 54fd191ee02827e82946e2cf82533ee93ab94f59..df17858bfe3dbff2b21554d14ab5fcf8efc898b5 100644
--- a/src/main/java/net/minestom/server/event/player/PlayerEatEvent.java
+++ b/src/main/java/net/minestom/server/event/player/PlayerEatEvent.java
@@ -9,6 +9,7 @@ import org.jetbrains.annotations.NotNull;
 /**
  * Called when a player is finished eating.
  */
+// Microtus - remove deprecated methods
 public class PlayerEatEvent implements ItemEvent, PlayerInstanceEvent {
 
     private final Player player;
@@ -21,17 +22,6 @@ public class PlayerEatEvent implements ItemEvent, PlayerInstanceEvent {
         this.hand = hand;
     }
 
-    /**
-     * Gets the food item that has been eaten.
-     *
-     * @return the food item
-     * @deprecated use getItemStack() for the eaten item
-     */
-    @Deprecated(forRemoval = true, since = "1.1.3") //Microtus - Update deprecation status
-    public @NotNull ItemStack getFoodItem() {
-        return foodItem;
-    }
-
     public @NotNull Player.Hand getHand() {
         return hand;
     }
diff --git a/src/main/java/net/minestom/server/event/player/PlayerPreEatEvent.java b/src/main/java/net/minestom/server/event/player/PlayerPreEatEvent.java
index 5511e29017aa54958b62f2e6f1fcfc53a3757a92..3d6960753812ef9d01047ec3c5f91a3623542e99 100644
--- a/src/main/java/net/minestom/server/event/player/PlayerPreEatEvent.java
+++ b/src/main/java/net/minestom/server/event/player/PlayerPreEatEvent.java
@@ -12,6 +12,7 @@ import org.jetbrains.annotations.NotNull;
  * or to cancel its processing, cancelling the event means that the player will
  * continue the animation indefinitely.
  */
+// Microtus - remove deprecated methods
 public class PlayerPreEatEvent implements ItemEvent, PlayerInstanceEvent, CancellableEvent {
 
     private final Player player;
@@ -28,17 +29,6 @@ public class PlayerPreEatEvent implements ItemEvent, PlayerInstanceEvent, Cancel
         this.eatingTime = eatingTime;
     }
 
-    /**
-     * The food item which will be eaten.
-     *
-     * @return the food item
-     * @deprecated use getItemStack() for the eaten item
-     */
-    @Deprecated(forRemoval = true, since = "1.1.3") //Microtus - Update deprecation status
-    public @NotNull ItemStack getFoodItem() {
-        return foodItem;
-    }
-
     public @NotNull Player.Hand getHand() {
         return hand;
     }
diff --git a/src/main/java/net/minestom/server/instance/Instance.java b/src/main/java/net/minestom/server/instance/Instance.java
index d6dd0549db22ee58c4057012fcc46a3aba8f087b..b93d4b722d088ec3de4a3fd4874ff94b6bb22364 100644
--- a/src/main/java/net/minestom/server/instance/Instance.java
+++ b/src/main/java/net/minestom/server/instance/Instance.java
@@ -61,6 +61,7 @@ import java.util.stream.Collectors;
  * with {@link InstanceManager#registerInstance(Instance)}, and
  * you need to be sure to signal the {@link ThreadDispatcher} of every partition/element changes.
  */
+// Microtus - remove deprecated methods
 public abstract class Instance implements Block.Getter, Block.Setter,
         Tickable, Schedulable, Snapshotable, EventHandler<InstanceEvent>, Taggable, PacketGroupingAudience {
 
diff --git a/src/main/java/net/minestom/server/inventory/InventoryType.java b/src/main/java/net/minestom/server/inventory/InventoryType.java
index 4667b7c3a3fe0a49c3b4e00210c103155d85ea39..d467975f66e85606eb551b845cc0415773a2d517 100644
--- a/src/main/java/net/minestom/server/inventory/InventoryType.java
+++ b/src/main/java/net/minestom/server/inventory/InventoryType.java
@@ -3,6 +3,7 @@ package net.minestom.server.inventory;
 /**
  * Represents a type of {@link Inventory}
  */
+// Microtus - remove deprecated methods
 public enum InventoryType {
 
     CHEST_1_ROW(9),
@@ -43,13 +44,4 @@ public enum InventoryType {
     public int getSize() {
         return size;
     }
-
-    /**
-     * @deprecated use {@link #getSize()}
-     */
-    @Deprecated(forRemoval = true, since = "1.1.3") //Microtus - Update deprecation status
-    public int getAdditionalSlot() {
-        return size;
-    }
-
 }
diff --git a/src/main/java/net/minestom/server/item/ItemStack.java b/src/main/java/net/minestom/server/item/ItemStack.java
index 0e3f2a48d1a1c75368fc0c33905cc5b07cb27251..ce210a60496d042456f5dcf7342d6d1d48956c35 100644
--- a/src/main/java/net/minestom/server/item/ItemStack.java
+++ b/src/main/java/net/minestom/server/item/ItemStack.java
@@ -24,6 +24,7 @@ import java.util.function.UnaryOperator;
  * <p>
  * An item stack cannot be null, {@link ItemStack#AIR} should be used instead.
  */
+// Microtus - remove deprecated methods
 public sealed interface ItemStack extends TagReadable, HoverEventSource<HoverEvent.ShowItem>
         permits ItemStackImpl {
     /**
@@ -180,25 +181,6 @@ public sealed interface ItemStack extends TagReadable, HoverEventSource<HoverEve
     @ApiStatus.Experimental
     @NotNull NBTCompound toItemNBT();
 
-
-    @Deprecated(forRemoval = true, since = "1.1.3") //Microtus - Update deprecation status
-    @Contract(pure = true)
-    default @NotNull Material getMaterial() {
-        return material();
-    }
-
-    @Deprecated(forRemoval = true, since = "1.1.3") //Microtus - Update deprecation status
-    @Contract(pure = true)
-    default int getAmount() {
-        return amount();
-    }
-
-    @Deprecated(forRemoval = true, since = "1.1.3") //Microtus - Update deprecation status
-    @Contract(pure = true)
-    default @NotNull ItemMeta getMeta() {
-        return meta();
-    }
-
     sealed interface Builder extends TagWritable
             permits ItemStackImpl.Builder {
         @Contract(value = "_ -> this")
diff --git a/src/main/java/net/minestom/server/ping/ResponseData.java b/src/main/java/net/minestom/server/ping/ResponseData.java
index 693edb9b906f347c81bda3d769ac2f17ed9db072..35856fe0e16e47b252c4693ebca9057f35af9c9b 100644
--- a/src/main/java/net/minestom/server/ping/ResponseData.java
+++ b/src/main/java/net/minestom/server/ping/ResponseData.java
@@ -1,11 +1,7 @@
 package net.minestom.server.ping;
 
-import com.google.gson.JsonObject;
 import net.kyori.adventure.text.Component;
-import net.kyori.adventure.text.serializer.legacy.LegacyComponentSerializer;
-import net.kyori.adventure.text.serializer.plain.PlainComponentSerializer;
 import net.minestom.server.MinecraftServer;
-import net.minestom.server.entity.Player;
 import net.minestom.server.event.server.ServerListPingEvent;
 import net.minestom.server.utils.identity.NamedAndIdentified;
 import org.jetbrains.annotations.NotNull;
@@ -17,6 +13,7 @@ import java.util.*;
  *
  * @see ServerListPingEvent
  */
+// Microtus - remove deprecated methods
 public class ResponseData {
     private static final Component DEFAULT_DESCRIPTION = Component.text("Minestom Server");
 
@@ -44,17 +41,6 @@ public class ResponseData {
         this.playersHidden = false;
     }
 
-    /**
-     * Sets the name for the response.
-     *
-     * @param name The name for the response data.
-     * @deprecated This is named incorrectly, use {@link #setVersion(String)} instead
-     */
-    @Deprecated(forRemoval = true, since = "1.1.3") //Microtus - Update deprecation status
-    public void setName(String name) {
-        this.setVersion(name);
-    }
-
     /**
      * Sets the version name for the response.
      *
@@ -127,89 +113,6 @@ public class ResponseData {
         return online;
     }
 
-    /**
-     * Adds some players to the response.
-     *
-     * @param players the players
-     * @deprecated Use {@link #addEntries(Collection)}}
-     */
-    @Deprecated(forRemoval = true, since = "1.1.3") //Microtus - Update deprecation status
-    public void addPlayer(Iterable<Player> players) {
-        for (Player player : players) {
-            this.addPlayer(player);
-        }
-    }
-
-    /**
-     * Adds a player to the response.
-     *
-     * @param player the player
-     * @deprecated Use {@link #addEntry(NamedAndIdentified)}
-     */
-    @Deprecated(forRemoval = true, since = "1.1.3") //Microtus - Update deprecation status
-    public void addPlayer(Player player) {
-        this.addEntry(player);
-    }
-
-    /**
-     * Adds a player to the response.
-     *
-     * @param name The name of the player.
-     * @param uuid The unique identifier of the player.
-     * @deprecated Use {@link #addEntry(NamedAndIdentified)} with {@link NamedAndIdentified#of(String, UUID)}
-     */
-    @Deprecated(forRemoval = true, since = "1.1.3") //Microtus - Update deprecation status
-    public void addPlayer(String name, UUID uuid) {
-        this.addEntry(NamedAndIdentified.of(name, uuid));
-    }
-
-    /**
-     * Adds a player to the response.
-     * <p>
-     * {@link UUID#randomUUID()} is used as the player's UUID.
-     *
-     * @param name The name of the player.
-     * @deprecated Use {@link #addEntry(NamedAndIdentified)} with {@link NamedAndIdentified#named(String)}
-     */
-    @Deprecated(forRemoval = true, since = "1.1.3") //Microtus - Update deprecation status
-    public void addPlayer(String name) {
-        this.addEntry(NamedAndIdentified.named(name));
-    }
-
-    /**
-     * Removes all of the ping players from this {@link #entries}. The {@link #entries} list
-     * will be empty this call returns.
-     *
-     * @deprecated Use {@link #clearEntries()}
-     */
-    @Deprecated(forRemoval = true, since = "1.1.3") //Microtus - Update deprecation status
-    public void clearPlayers() {
-        this.clearEntries();
-    }
-
-    /**
-     * Get the list of the response players.
-     *
-     * @return the list of the response players.
-     * @deprecated Use {@link #getEntries()}. This return value is now unmodifiable and this operation is incredibly costly.
-     */
-    @Deprecated(forRemoval = true) // to throw an error for people using it - this method is *horrible*
-    public List<PingPlayer> getPlayers() {
-        return this.entries.stream()
-                .map(entry -> PingPlayer.of(PlainComponentSerializer.plain().serialize(entry.getName()), entry.getUuid())).toList();
-    }
-
-    /**
-     * Sets the response description.
-     *
-     * @param description The description for the response data.
-     * @deprecated Use {@link #setDescription(Component)}
-     */
-    @Deprecated(forRemoval = true, since = "1.1.3") //Microtus - Update deprecation status
-    public void setDescription(String description) {
-        this.description = LegacyComponentSerializer.legacySection().deserialize(description);
-    }
-
     /**
      * Sets the response description.
      *
@@ -313,43 +216,4 @@ public class ResponseData {
         return playersHidden;
     }
 
-    /**
-     * Converts the response data into a {@link JsonObject}.
-     *
-     * @return The converted response data as a json tree.
-     * @deprecated Use {@link ServerListPingType#getPingResponse(ResponseData)}
-     */
-    @Deprecated
-    public @NotNull JsonObject build() {
-        return ServerListPingType.getModernPingResponse(this, true);
-    }
-
-    /**
-     * Represents a player line in the server list hover.
-     *
-     * @deprecated See {@link NamedAndIdentified}
-     */
-    @Deprecated
-    public static class PingPlayer {
-
-        private static @NotNull PingPlayer of(@NotNull String name, @NotNull UUID uuid) {
-            return new PingPlayer(name, uuid);
-        }
-
-        private final String name;
-        private final UUID uuid;
-
-        private PingPlayer(@NotNull String name, @NotNull UUID uuid) {
-            this.name = name;
-            this.uuid = uuid;
-        }
-
-        public @NotNull String getName() {
-            return name;
-        }
-
-        public @NotNull UUID getUuid() {
-            return uuid;
-        }
-    }
 }
diff --git a/src/main/java/net/minestom/server/scoreboard/BelowNameTag.java b/src/main/java/net/minestom/server/scoreboard/BelowNameTag.java
index b5c67427644748d5199ac42757c7360e5cb82d21..efc4545df49796557446f2edca16ad52775f0809 100644
--- a/src/main/java/net/minestom/server/scoreboard/BelowNameTag.java
+++ b/src/main/java/net/minestom/server/scoreboard/BelowNameTag.java
@@ -12,6 +12,7 @@ import java.util.concurrent.CopyOnWriteArraySet;
 /**
  * Represents a scoreboard which rendered a tag below the name.
  */
+// Microtus - remove deprecated methods
 public class BelowNameTag implements Scoreboard {
 
     /**
@@ -25,18 +26,6 @@ public class BelowNameTag implements Scoreboard {
 
     private final ScoreboardObjectivePacket scoreboardObjectivePacket;
 
-    /**
-     * Creates a new below name scoreboard.
-     *
-     * @param name  The objective name of the scoreboard
-     * @param value The value of the scoreboard
-     * @deprecated Use {@link #BelowNameTag(String, Component)}
-     */
-    @Deprecated(forRemoval = true, since = "1.1.3") //Microtus - Update deprecation status
-    public BelowNameTag(String name, String value) {
-        this(name, Component.text(value));
-    }
-
     /**
      * Creates a new below name scoreboard.
      *
diff --git a/src/main/java/net/minestom/server/scoreboard/Scoreboard.java b/src/main/java/net/minestom/server/scoreboard/Scoreboard.java
index 9a1ad120480b8cf11106915185ef63cf8d445a3b..ac1def5cf1771ae79f83f2603192d822b2741fdb 100644
--- a/src/main/java/net/minestom/server/scoreboard/Scoreboard.java
+++ b/src/main/java/net/minestom/server/scoreboard/Scoreboard.java
@@ -14,22 +14,9 @@ import java.util.Collection;
 /**
  * This interface represents all scoreboard of Minecraft.
  */
+// Microtus - remove deprecated methods
 public interface Scoreboard extends Viewable, PacketGroupingAudience {
 
-    /**
-     * Creates a creation objective packet.
-     *
-     * @param value The value for the objective
-     * @param type  The type for the objective
-     * @return the creation objective packet
-     * @deprecated Use {@link #getCreationObjectivePacket(Component, ScoreboardObjectivePacket.Type)}
-     */
-    @Deprecated(forRemoval = true, since = "1.1.3") //Microtus - Update deprecation status
-    @NotNull
-    default ScoreboardObjectivePacket getCreationObjectivePacket(String value, ScoreboardObjectivePacket.Type type) {
-        return this.getCreationObjectivePacket(Component.text(value), type);
-    }
-
     /**
      * Creates a creation objective packet.
      *
diff --git a/src/main/java/net/minestom/server/scoreboard/Sidebar.java b/src/main/java/net/minestom/server/scoreboard/Sidebar.java
index efc5627faddcc6b688437df7febb6e2593348e93..f7f1749dd151728337bd4b8d67765c6ad93ceac4 100644
--- a/src/main/java/net/minestom/server/scoreboard/Sidebar.java
+++ b/src/main/java/net/minestom/server/scoreboard/Sidebar.java
@@ -21,13 +21,14 @@ import java.util.concurrent.atomic.AtomicInteger;
 /**
  * Represents a sidebar which can contain up to 16 {@link ScoreboardLine}.
  * <p>
- * In order to use it you need to create a new instance using the constructor {@link #Sidebar(String)} and create new lines
+ * In order to use it you need to create a new instance using the constructor {@link #Sidebar(Component)} and create new lines
  * with {@link #createLine(ScoreboardLine)}. You can then add a {@link Player} to the viewing list using {@link #addViewer(Player)}
  * and remove him later with {@link #removeViewer(Player)}.
  * <p>
  * Lines can be modified using their respective identifier using
  * {@link #updateLineContent(String, Component)} and {@link #updateLineScore(String, int)}.
  */
+// Microtus - remove deprecated methods
 public class Sidebar implements Scoreboard {
 
     private static final AtomicInteger COUNTER = new AtomicInteger();
@@ -52,17 +53,6 @@ public class Sidebar implements Scoreboard {
 
     private Component title;
 
-    /**
-     * Creates a new sidebar
-     *
-     * @param title The title of the sidebar
-     * @deprecated Use {@link #Sidebar(Component)}
-     */
-    @Deprecated(forRemoval = true, since = "1.1.3") //Microtus - Update deprecation status
-    public Sidebar(@NotNull String title) {
-        this(Component.text(title));
-    }
-
     /**
      * Creates a new sidebar
      *
@@ -79,17 +69,6 @@ public class Sidebar implements Scoreboard {
         }
     }
 
-    /**
-     * Changes the {@link Sidebar} title
-     *
-     * @param title The new sidebar title
-     * @deprecated Use {@link #setTitle(Component)}
-     */
-    @Deprecated(forRemoval = true, since = "1.1.3") //Microtus - Update deprecation status
-    public void setTitle(@NotNull String title) {
-        this.setTitle(Component.text(title));
-    }
-
     /**
      * Changes the {@link Sidebar} title
      *
diff --git a/src/main/java/net/minestom/server/scoreboard/TabList.java b/src/main/java/net/minestom/server/scoreboard/TabList.java
index 7127b875ce50b8637ea63921519b780ea6b6bc4a..35cc1584d9e319264cb385d63958b7c9c237a8f0 100644
--- a/src/main/java/net/minestom/server/scoreboard/TabList.java
+++ b/src/main/java/net/minestom/server/scoreboard/TabList.java
@@ -12,6 +12,7 @@ import java.util.concurrent.CopyOnWriteArraySet;
 /**
  * Represents the {@link Player} tab list as a {@link Scoreboard}.
  */
+// Microtus - remove deprecated methods
 public class TabList implements Scoreboard {
 
     /**
@@ -68,14 +69,13 @@ public class TabList implements Scoreboard {
         return result;
     }
 
-    @NotNull
     @Override
-    public Set<Player> getViewers() {
+    public @NotNull Set<Player> getViewers() {
         return unmodifiableViewers;
     }
 
     @Override
-    public String getObjectiveName() {
+    public @NotNull String getObjectiveName() {
         return this.objectiveName;
     }
 }
diff --git a/src/main/java/net/minestom/server/utils/NamespaceID.java b/src/main/java/net/minestom/server/utils/NamespaceID.java
index 988eb3f876c4720243876bdbea45e4ff5e4a6629..3dff7d06972dfb2aeab589cb1032e5eee5cc1fb1 100644
--- a/src/main/java/net/minestom/server/utils/NamespaceID.java
+++ b/src/main/java/net/minestom/server/utils/NamespaceID.java
@@ -12,6 +12,7 @@ import java.util.Objects;
  * Represents a namespaced ID
  * https://minecraft.gamepedia.com/Namespaced_ID
  */
+// Microtus - remove deprecated methods
 public final class NamespaceID implements CharSequence, Key {
     private static final String legalLetters = "[0123456789abcdefghijklmnopqrstuvwxyz_-]+";
     private static final String legalPathLetters = "[0123456789abcdefghijklmnopqrstuvwxyz./_-]+";
@@ -67,14 +68,12 @@ public final class NamespaceID implements CharSequence, Key {
     public boolean equals(Object o) {
         if (this == o) return true;
         if (!(o instanceof final Key that)) return false;
-        return Objects.equals(this.domain, that.namespace()) && Objects.equals(this.path, that.value()); // Microtus - Fix namespace hashcode and equals
+        return Objects.equals(this.full, that.asString());
     }
 
     @Override
     public int hashCode() {
-        int result = this.domain.hashCode(); // Microtus start - Fix namespace hashcode and equals
-        result = (31 * result) + this.path.hashCode();
-        return result; // Microtus end - Fix namespace hashcode and equals
+        return full.hashCode();
     }
 
     @Override
@@ -112,14 +111,4 @@ public final class NamespaceID implements CharSequence, Key {
     public @NotNull String asString() {
         return this.full;
     }
-
-    @Deprecated(forRemoval = true, since = "1.1.3") //Microtus - Update deprecation status
-    public String getDomain() {
-        return domain();
-    }
-
-    @Deprecated(forRemoval = true, since = "1.1.3") //Microtus - Update deprecation status
-    public String getPath() {
-        return path();
-    }
 }
diff --git a/src/main/java/net/minestom/server/utils/time/TimeUnit.java b/src/main/java/net/minestom/server/utils/time/TimeUnit.java
index 7e181a47c91f8f925b20eb9f4705dabdb24a0c2f..43e25a3f283ea6af40d32deb142c274117e0d91c 100644
--- a/src/main/java/net/minestom/server/utils/time/TimeUnit.java
+++ b/src/main/java/net/minestom/server/utils/time/TimeUnit.java
@@ -11,11 +11,6 @@ public final class TimeUnit {
     public static final TemporalUnit MILLISECOND = ChronoUnit.MILLIS;
     public static final TemporalUnit SERVER_TICK = Tick.SERVER_TICKS;
     public static final TemporalUnit CLIENT_TICK = Tick.CLIENT_TICKS;
-    /**
-     * @deprecated Please use either {@link #SERVER_TICK} or {@link #CLIENT_TICK}
-     */
-    @Deprecated(forRemoval = true)
-    public static final TemporalUnit TICK = CLIENT_TICK;
 
     private TimeUnit() {
     }
