From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: theEvilReaper <steffenwx@gmail.com>
Date: Sat, 5 Aug 2023 23:04:05 +0200
Subject: [PATCH] Improve string pattern usage


diff --git a/src/main/java/net/minestom/server/utils/UniqueIdUtils.java b/src/main/java/net/minestom/server/utils/UniqueIdUtils.java
index 242a6a1b55102507c15ffa37ff26abcac38f6823..f894c867452c9bd1849d9d01c98a71880c1e450b 100644
--- a/src/main/java/net/minestom/server/utils/UniqueIdUtils.java
+++ b/src/main/java/net/minestom/server/utils/UniqueIdUtils.java
@@ -1,6 +1,7 @@
 package net.minestom.server.utils;
 
 import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.NotNull;
 
 import java.util.UUID;
 import java.util.regex.Pattern;
@@ -12,13 +13,16 @@ import java.util.regex.Pattern;
 public final class UniqueIdUtils {
     public static final Pattern UNIQUE_ID_PATTERN = Pattern.compile("\\b[0-9a-f]{8}\\b-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-\\b[0-9a-f]{12}\\b");
 
+    // Microtus - improve string pattern usage + add private constructor
+    private UniqueIdUtils() {}
+
     /**
      * Checks whether the {@code input} string is an {@link UUID}.
      *
      * @param input The input string to be checked
      * @return {@code true} if the input an unique identifier, otherwise {@code false}
      */
-    public static boolean isUniqueId(String input) {
-        return input.matches(UNIQUE_ID_PATTERN.pattern());
+    public static boolean isUniqueId(@NotNull String input) {
+        return !input.trim().isEmpty() && UNIQUE_ID_PATTERN.matcher(input).matches(); // Microtus - improve string pattern usage
     }
 }
diff --git a/src/test/java/net/minestom/server/utils/UniqueIdUtilsTest.java b/src/test/java/net/minestom/server/utils/UniqueIdUtilsTest.java
new file mode 100644
index 0000000000000000000000000000000000000000..f7bc989b8f89f37798e2645a54d71127fa5a767d
--- /dev/null
+++ b/src/test/java/net/minestom/server/utils/UniqueIdUtilsTest.java
@@ -0,0 +1,18 @@
+package net.minestom.server.utils;
+
+import org.junit.jupiter.api.Test;
+
+import java.util.UUID;
+
+import static net.minestom.server.utils.UniqueIdUtils.*;
+import static org.junit.jupiter.api.Assertions.*;
+
+// Microtus - improve string pattern usage + test
+class UniqueIdUtilsTest {
+
+    @Test
+    void testUniqueIDCheck() {
+        assertFalse(isUniqueId(""));
+        assertTrue(isUniqueId(UUID.randomUUID().toString()));
+    }
+}
