From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: TheMeinerLP <p.glanz@madfix.me>
Date: Wed, 14 Jun 2023 12:23:54 +0200
Subject: [PATCH] Bump dependencies


diff --git a/gradle/libs.versions.toml b/gradle/libs.versions.toml
index 01c3f8d80bc23a627af5e27b48ffba052cfe1190..b661e4949293c31d49b2c74f9b39f802ff71c7fd 100644
--- a/gradle/libs.versions.toml
+++ b/gradle/libs.versions.toml
@@ -3,30 +3,29 @@ metadata.format.version = "1.1"
 [versions]
 
 # Important dependencies
-adventure = "4.12.0"
-kotlin = "1.7.22"
+adventure = "4.14.0"
+kotlin = "1.8.22"
 hydrazine = "1.7.2"
 dependencyGetter = "v1.0.1"
 data = "1.20-rv2" # Microtous - 1.20 temp change to ce version
-#minestomData = "e6ebd0fb9b" # Microtus - 1.19.4 update
 hephaistos = "2.6.1"
-jetbrainsAnnotations = "23.0.0"
+jetbrainsAnnotations = "24.0.1"
 
 # Terminal / Logging
 tinylog = "2.5.0"
-jline = "3.21.0"
+jline = "3.23.0"
 
 # Performance / Data Structures
-caffeine = "3.1.2"
-fastutil = "8.5.9"
+caffeine = "3.1.6"
+fastutil = "8.5.12"
 flare = "2.0.1"
-gson = "2.9.1"
+gson = "2.10.1"
 jcTools = "4.0.1"
 
 # Test
-junit-jupiter = "5.8.2"
+junit-jupiter = "5.9.3"
 junit-platform = "1.8.2"
-mockito = "4.2.0"
+mockito = "5.3.1"
 
 # Code Generation
 javaPoet = "1.13.0"
@@ -35,10 +34,10 @@ javaPoet = "1.13.0"
 jNoise = "b93008e35e"
 
 # JMH
-jmh = "1.35"
+jmh = "1.36"
 
 # JCStress
-jcstress = "0.8"
+jcstress = "0.16"
 
 # Gradle plugins
 blossom = "2.1.0"
