From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Phillipp Glanz <6745190+TheMeinerLP@users.noreply.github.com>
Date: Mon, 8 Jan 2024 13:42:41 +0100
Subject: [PATCH] Add better biome system


diff --git a/code-generators/src/main/java/net/minestom/codegen/Generators.java b/code-generators/src/main/java/net/minestom/codegen/Generators.java
index 2f143b9d196670b7f62d128222674624efe7c38f..07dfb812a5dc4e10944545d4e19e00a125490d31 100644
--- a/code-generators/src/main/java/net/minestom/codegen/Generators.java
+++ b/code-generators/src/main/java/net/minestom/codegen/Generators.java
@@ -1,5 +1,6 @@
 package net.minestom.codegen;
 
+import net.minestom.codegen.biome.BiomeGenerator;
 import net.minestom.codegen.color.DyeColorGenerator;
 import net.minestom.codegen.fluid.FluidGenerator;
 import org.slf4j.Logger;
@@ -21,6 +22,7 @@ public class Generators {
 
         // Generate DyeColors
         new DyeColorGenerator(resource("dye_colors.json"), outputFolder).generate();
+        new BiomeGenerator(resource("biomes.json"), outputFolder).generate();
 
 
         var generator = new CodeGenerator(outputFolder);
diff --git a/code-generators/src/main/java/net/minestom/codegen/MinestomCodeGenerator.java b/code-generators/src/main/java/net/minestom/codegen/MinestomCodeGenerator.java
index a56248de382badce98656ce4006452beb8990cc8..69eb7e67dd5253bca7e7948335a8f2f2acd849f1 100644
--- a/code-generators/src/main/java/net/minestom/codegen/MinestomCodeGenerator.java
+++ b/code-generators/src/main/java/net/minestom/codegen/MinestomCodeGenerator.java
@@ -3,6 +3,7 @@ package net.minestom.codegen;
 import com.google.gson.Gson;
 import com.google.gson.GsonBuilder;
 import com.squareup.javapoet.JavaFile;
+import java.util.regex.Pattern;
 import org.jetbrains.annotations.NotNull;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
@@ -15,6 +16,7 @@ import java.util.Locale;
 public abstract class MinestomCodeGenerator {
     private static final Logger LOGGER = LoggerFactory.getLogger(MinestomCodeGenerator.class);
     protected static final Gson GSON = new GsonBuilder().setPrettyPrinting().disableHtmlEscaping().create();
+    protected static final Pattern minecraftNamespaceRegex = Pattern.compile("minecraft:");
 
     public abstract void generate();
 
@@ -28,7 +30,8 @@ public abstract class MinestomCodeGenerator {
         }
     }
 
+
     protected static String toConstant(String namespace) {
-        return namespace.replace("minecraft:", "").toUpperCase(Locale.ROOT);
+        return minecraftNamespaceRegex.matcher(namespace).replaceAll("").toUpperCase(Locale.ROOT);
     }
 }
diff --git a/code-generators/src/main/java/net/minestom/codegen/biome/BiomeGenerator.java b/code-generators/src/main/java/net/minestom/codegen/biome/BiomeGenerator.java
new file mode 100644
index 0000000000000000000000000000000000000000..3c125bfa40ac413f79d8d17318de1236ebddc7ae
--- /dev/null
+++ b/code-generators/src/main/java/net/minestom/codegen/biome/BiomeGenerator.java
@@ -0,0 +1,153 @@
+package net.minestom.codegen.biome;
+
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+
+import com.squareup.javapoet.ArrayTypeName;
+import com.squareup.javapoet.ClassName;
+
+import com.squareup.javapoet.FieldSpec;
+import com.squareup.javapoet.TypeSpec;
+import com.squareup.javapoet.MethodSpec;
+import com.squareup.javapoet.ParameterSpec;
+import com.squareup.javapoet.AnnotationSpec;
+import com.squareup.javapoet.TypeName;
+import com.squareup.javapoet.CodeBlock;
+import com.squareup.javapoet.JavaFile;
+
+import java.awt.Color;
+
+import net.minestom.codegen.MinestomCodeGenerator;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.lang.model.element.Modifier;
+import java.io.File;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.util.List;
+
+public class BiomeGenerator extends MinestomCodeGenerator {
+
+    private static final String BIOME_FILE_PACKAGE = "net.minestom.server.world.biomes"; // Microtus - Biomes
+    private static final String CLASS_NAME = "Biomes"; // Microtus - Biomes
+    private static final Logger LOGGER = LoggerFactory.getLogger(BiomeGenerator.class);
+    private final InputStream biomeFile;
+    private final File outputFolder;
+
+    public BiomeGenerator(@Nullable InputStream biomeFile, @NotNull File outputFolder) {
+        this.biomeFile = biomeFile;
+        this.outputFolder = outputFolder;
+    }
+
+    @Override
+    public void generate() {
+        if (biomeFile == null) {
+            LOGGER.error("Failed to find biome.json.");
+            LOGGER.error("Stopped code generation for biomes.");
+            return;
+        }
+        if (!outputFolder.exists() && !outputFolder.mkdirs()) {
+            LOGGER.error("Output folder for code generation does not exist and could not be created.");
+            return;
+        }
+
+
+        JsonElement biomes = GSON.fromJson(new InputStreamReader(biomeFile), JsonElement.class);
+        ClassName biomeCN = ClassName.get(BIOME_FILE_PACKAGE, CLASS_NAME);
+        ClassName biomeCn = ClassName.get(BIOME_FILE_PACKAGE, "Biome");
+        ClassName categoryCn = biomeCn.nestedClass("Category");
+        ClassName minecraftServerCn = ClassName.get("net.minestom.server", "MinecraftServer");
+        ClassName namespaceIDCn = ClassName.get("net.minestom.server.utils", "NamespaceID");
+        ClassName biomeEffectsCn = ClassName.get(BIOME_FILE_PACKAGE, "BiomeEffects");
+        // Dye Color Enum
+        TypeSpec.Builder biomesEnum = TypeSpec.enumBuilder(biomeCN)
+                .addModifiers(Modifier.PUBLIC).addJavadoc("AUTOGENERATED by " + getClass().getSimpleName());
+
+        // Fields
+        biomesEnum.addFields(
+                List.of(
+                        FieldSpec.builder(biomeCn, "biome", Modifier.PRIVATE, Modifier.FINAL).build(),
+                        FieldSpec.builder(ArrayTypeName.of(biomeCN), "VALUES", Modifier.PRIVATE, Modifier.STATIC, Modifier.FINAL).initializer(CLASS_NAME + ".values()").build()  // Microtus - Banner and shield meta
+                )
+        );
+
+        // Methods
+        biomesEnum.addMethods(
+                List.of(
+                        // Constructor
+                        MethodSpec.constructorBuilder()
+                                .addParameters(
+                                        List.of(
+                                                ParameterSpec.builder(biomeCn, "biome").addAnnotation(NotNull.class).build()
+                                        )
+                                )
+                                .addStatement("this.biome = biome")
+                                .build(),
+                        MethodSpec.methodBuilder("biome")
+                                .addModifiers(Modifier.PUBLIC)
+                                .returns(biomeCn.annotated(AnnotationSpec.builder(NotNull.class).build()))
+                                .addStatement("return this.biome")
+                                .build(),
+                        MethodSpec.methodBuilder("getValue")
+                                .addModifiers(Modifier.PUBLIC, Modifier.STATIC)
+                                .addParameter(ParameterSpec.builder(TypeName.INT, "id").build())
+                                .addAnnotation(Nullable.class)
+                                .returns(ClassName.get(BIOME_FILE_PACKAGE, CLASS_NAME))
+                                .addCode("return VALUES[$L];", "id")
+                                .build(),
+                        MethodSpec.methodBuilder("registerBiomes")
+                                .addModifiers(Modifier.PUBLIC, Modifier.STATIC)
+                                .beginControlFlow("for (var biome : VALUES)")
+                                .addStatement("$T.getBiomeManager().addBiome(biome.biome())", minecraftServerCn)
+                                .endControlFlow()
+                                .build()
+                )
+        );
+        final JsonObject biomesAsJsonObject = biomes.getAsJsonObject();
+        biomesAsJsonObject.keySet().forEach(biome -> {
+
+
+            var biomeObject = biomesAsJsonObject.get(biome).getAsJsonObject();
+            var tempature = biomeObject.get("temperature").getAsFloat();
+            var downfall = biomeObject.get("downfall").getAsFloat();
+            var hasPrecipitation = biomeObject.get("has_precipitation").getAsBoolean();
+            var fogColor = Integer.toHexString(new Color(biomeObject.get("fogColor").getAsInt()).getRGB()).substring(2);
+            var waterColor = Integer.toHexString(new Color(biomeObject.get("waterColor").getAsInt()).getRGB()).substring(2);
+            var waterFogColor = Integer.toHexString(new Color(biomeObject.get("waterFogColor").getAsInt()).getRGB()).substring(2);
+            var skyColor = Integer.toHexString(new Color(biomeObject.get("skyColor").getAsInt()).getRGB()).substring(2);
+            var effectCodeBlock = CodeBlock.builder();
+
+            effectCodeBlock = effectCodeBlock.add("$T.builder().fogColor(0x$L).waterColor(0x$L).waterFogColor(0x$L).skyColor(0x$L)", biomeEffectsCn, fogColor, waterColor, waterFogColor, skyColor);
+            if (biomeObject.has("grassColor")) {
+                var grassColor = Integer.toHexString(new Color(biomeObject.get("grassColor").getAsInt()).getRGB()).substring(2);
+                effectCodeBlock = effectCodeBlock.add(".grassColor(0x$L)", grassColor);
+            }
+            if (biomeObject.has("foliageColor")) {
+                var foliageColor = Integer.toHexString(new Color(biomeObject.get("foliageColor").getAsInt()).getRGB()).substring(2);
+                ;
+                effectCodeBlock = effectCodeBlock.add(".foliageColor(0x$L)", foliageColor);
+            }
+            effectCodeBlock = effectCodeBlock.add(".build()");
+            var namespacedCode = CodeBlock.of("$T.from($S)", namespaceIDCn, biome.replaceFirst("minecraft:", ""));
+            biomesEnum.addEnumConstant(toConstant(biome), TypeSpec.anonymousClassBuilder(
+                            "$T.builder().name($L).category($T.NONE).depth(0.125F).scale(0.05F).temperature($Lf).downfall($Lf).showPrecipitation($L).effects($L).build()",
+                            biomeCn, namespacedCode, categoryCn, tempature, downfall, hasPrecipitation, effectCodeBlock.build()
+                    ).build()
+            );
+        });
+
+        // Write files to outputFolder
+        writeFiles(
+                List.of(
+                        JavaFile.builder(BIOME_FILE_PACKAGE, biomesEnum.build())
+                                .indent("    ")
+                                .skipJavaLangImports(true)
+                                .build()
+                ),
+                outputFolder
+        );
+    }
+}
diff --git a/demo/src/main/java/net/minestom/demo/Main.java b/demo/src/main/java/net/minestom/demo/Main.java
index 7a5cd6295fbdf87b24b346ddfaeeaab9dc91fdcf..977d1435323be6c0abb12d904f733496c31088ee 100644
--- a/demo/src/main/java/net/minestom/demo/Main.java
+++ b/demo/src/main/java/net/minestom/demo/Main.java
@@ -11,7 +11,6 @@ import net.minestom.server.command.CommandManager;
 import net.minestom.server.event.server.ServerListPingEvent;
 import net.minestom.server.extras.lan.OpenToLAN;
 import net.minestom.server.extras.lan.OpenToLANConfig;
-import net.minestom.server.extras.optifine.OptifineSupport;
 import net.minestom.server.instance.block.BlockManager;
 import net.minestom.server.instance.block.rule.vanilla.RedstonePlacementRule;
 import net.minestom.server.ping.ResponseData;
@@ -107,6 +106,44 @@ public class Main {
         OpenToLAN.open(new OpenToLANConfig().eventCallDelay(Duration.of(1, TimeUnit.DAY)));
 
         minecraftServer.start("0.0.0.0", 25565);
+        /* var oldBiome = MinecraftServer.getBiomeManager().getByName(NamespaceID.from("minecraft:crimson_forest"));
+        MinecraftServer.getBiomeManager().removeBiome(oldBiome);
+        var newBiome = Biome.builder().name(NamespaceID.from("minecraft:crimson_forest"))
+                .category(Biome.Category.NONE).depth(0.125F).scale(0.05F).temperature(2.0f).downfall(0.0f)
+                .showPrecipitation(false)
+                .effects(BiomeEffects.builder().fogColor(0x330303).waterColor(0x3f76e4).waterFogColor(0x050533).skyColor(0x6eb1ff)
+                        .biomeParticle(new BiomeParticle(0.025f, new DefaultOption(Particle.CRIMSON_SPORE)))
+                        .build()).build();
+        MinecraftServer.getBiomeManager().addBiome(newBiome);
+
+        oldBiome = MinecraftServer.getBiomeManager().getByName(NamespaceID.from("minecraft:warped_forest"));
+        MinecraftServer.getBiomeManager().removeBiome(oldBiome);
+        newBiome = Biome.builder().name(NamespaceID.from("minecraft:warped_forest")).category(Biome.Category.NONE).depth(0.125F).scale(0.05F).temperature(2.0f).downfall(0.0f).showPrecipitation(false).effects(BiomeEffects.builder().fogColor(0x1a051a).waterColor(0x3f76e4).waterFogColor(0x050533).skyColor(0x6eb1ff)
+                        .biomeParticle(new BiomeParticle(0.01428f, new DefaultOption(Particle.WARPED_SPORE)))
+                        .build()).build();
+        MinecraftServer.getBiomeManager().addBiome(newBiome);
+
+        oldBiome = MinecraftServer.getBiomeManager().getByName(NamespaceID.from("minecraft:basalt_deltas"));
+        MinecraftServer.getBiomeManager().removeBiome(oldBiome);
+        newBiome = Biome.builder().name(NamespaceID.from("minecraft:basalt_deltas")).category(Biome.Category.NONE).depth(0.125F).scale(0.05F).temperature(2.0f).downfall(0.0f).showPrecipitation(false).effects(BiomeEffects.builder().fogColor(0x685f70).waterColor(0x3f76e4).waterFogColor(0x050533).skyColor(0x6eb1ff)
+                .biomeParticle(new BiomeParticle(0.118093334f, new DefaultOption(Particle.WHITE_ASH)))
+                .build()).build();
+        MinecraftServer.getBiomeManager().addBiome(newBiome);
+
+        oldBiome = MinecraftServer.getBiomeManager().getByName(NamespaceID.from("minecraft:soul_sand_valley"));
+        MinecraftServer.getBiomeManager().removeBiome(oldBiome);
+        newBiome = Biome.builder().name(NamespaceID.from("minecraft:soul_sand_valley")).category(Biome.Category.NONE).depth(0.125F).scale(0.05F).temperature(2.0f).downfall(0.0f).showPrecipitation(false).effects(BiomeEffects.builder().fogColor(0x1b4745).waterColor(0x3f76e4).waterFogColor(0x050533).skyColor(0x6eb1ff)
+                .biomeParticle(new BiomeParticle(0.00625f, new DefaultOption(Particle.ASH)))
+                .build()).build();
+        MinecraftServer.getBiomeManager().addBiome(newBiome);
+
+        oldBiome = MinecraftServer.getBiomeManager().getByName(NamespaceID.from("minecraft:river"));
+        MinecraftServer.getBiomeManager().removeBiome(oldBiome);
+        newBiome = Biome.builder().name(NamespaceID.from("minecraft:river")).category(Biome.Category.NONE).depth(0.125F).scale(0.05F).temperature(0.5f).downfall(0.5f).showPrecipitation(true).effects(BiomeEffects.builder().fogColor(0xc0d8ff).waterColor(0x3f76e4).waterFogColor(0x050533).skyColor(0x7ba4ff)
+                // .biomeParticle(new BiomeParticle(0.118093334f, new DefaultOption(Particle.DRAGON_BREATH)))
+                .biomeParticle(new BiomeParticle(1f, new DustOption(1f,0f,0f, 1f)))
+                .build()).build();
+        MinecraftServer.getBiomeManager().addBiome(newBiome);*/
 //        minecraftServer.start(java.net.UnixDomainSocketAddress.of("minestom-demo.sock"));
         //Runtime.getRuntime().addShutdownHook(new Thread(MinecraftServer::stopCleanly));
     }
diff --git a/demo/src/main/resources/extension.json b/demo/src/main/resources/extension.json
deleted file mode 100644
index efb46608f1c0ea0ab08486a554040c71e5341391..0000000000000000000000000000000000000000
--- a/demo/src/main/resources/extension.json
+++ /dev/null
@@ -1,13 +0,0 @@
-{
-  "entrypoint": "net.minestom.demo.extension.TestExtension",
-  "name": "TestExtension",
-  "version": "1.0.0",
-  "externalDependencies": {
-    "repositories": [
-      { "name": "Central", "url": "https://repo1.maven.org/maven2/" }
-    ],
-    "artifacts": [
-      "mysql:mysql-connector-java:8.0.26"
-    ]
-  }
-}
\ No newline at end of file
diff --git a/src/autogenerated/java/net/minestom/server/world/biomes/Biomes.java b/src/autogenerated/java/net/minestom/server/world/biomes/Biomes.java
new file mode 100644
index 0000000000000000000000000000000000000000..a11ba6e607e5c2e224632c10a258a647231f9722
--- /dev/null
+++ b/src/autogenerated/java/net/minestom/server/world/biomes/Biomes.java
@@ -0,0 +1,162 @@
+package net.minestom.server.world.biomes;
+
+import net.minestom.server.MinecraftServer;
+import net.minestom.server.utils.NamespaceID;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * AUTOGENERATED by BiomeGenerator
+ */
+public enum Biomes {
+    SNOWY_SLOPES(Biome.builder().name(NamespaceID.from("minecraft:snowy_slopes")).category(Biome.Category.NONE).depth(0.125F).scale(0.05F).temperature(-0.3f).downfall(0.9f).showPrecipitation(true).effects(BiomeEffects.builder().fogColor(0xc0d8ff).waterColor(0x3f76e4).waterFogColor(0x050533).skyColor(0x829fff).build()).build()),
+
+    OLD_GROWTH_PINE_TAIGA(Biome.builder().name(NamespaceID.from("minecraft:old_growth_pine_taiga")).category(Biome.Category.NONE).depth(0.125F).scale(0.05F).temperature(0.3f).downfall(0.8f).showPrecipitation(true).effects(BiomeEffects.builder().fogColor(0xc0d8ff).waterColor(0x3f76e4).waterFogColor(0x050533).skyColor(0x7ca3ff).build()).build()),
+
+    MUSHROOM_FIELDS(Biome.builder().name(NamespaceID.from("minecraft:mushroom_fields")).category(Biome.Category.NONE).depth(0.125F).scale(0.05F).temperature(0.9f).downfall(1.0f).showPrecipitation(true).effects(BiomeEffects.builder().fogColor(0xc0d8ff).waterColor(0x3f76e4).waterFogColor(0x050533).skyColor(0x77a8ff).build()).build()),
+
+    TAIGA(Biome.builder().name(NamespaceID.from("minecraft:taiga")).category(Biome.Category.NONE).depth(0.125F).scale(0.05F).temperature(0.25f).downfall(0.8f).showPrecipitation(true).effects(BiomeEffects.builder().fogColor(0xc0d8ff).waterColor(0x3f76e4).waterFogColor(0x050533).skyColor(0x7da3ff).build()).build()),
+
+    DEEP_OCEAN(Biome.builder().name(NamespaceID.from("minecraft:deep_ocean")).category(Biome.Category.NONE).depth(0.125F).scale(0.05F).temperature(0.5f).downfall(0.5f).showPrecipitation(true).effects(BiomeEffects.builder().fogColor(0xc0d8ff).waterColor(0x3f76e4).waterFogColor(0x050533).skyColor(0x7ba4ff).build()).build()),
+
+    ERODED_BADLANDS(Biome.builder().name(NamespaceID.from("minecraft:eroded_badlands")).category(Biome.Category.NONE).depth(0.125F).scale(0.05F).temperature(2.0f).downfall(0.0f).showPrecipitation(false).effects(BiomeEffects.builder().fogColor(0xc0d8ff).waterColor(0x3f76e4).waterFogColor(0x050533).skyColor(0x6eb1ff).grassColor(0x90814d).foliageColor(0x9e814d).build()).build()),
+
+    FROZEN_RIVER(Biome.builder().name(NamespaceID.from("minecraft:frozen_river")).category(Biome.Category.NONE).depth(0.125F).scale(0.05F).temperature(0.0f).downfall(0.5f).showPrecipitation(true).effects(BiomeEffects.builder().fogColor(0xc0d8ff).waterColor(0x3938c9).waterFogColor(0x050533).skyColor(0x7fa1ff).build()).build()),
+
+    END_HIGHLANDS(Biome.builder().name(NamespaceID.from("minecraft:end_highlands")).category(Biome.Category.NONE).depth(0.125F).scale(0.05F).temperature(0.5f).downfall(0.5f).showPrecipitation(false).effects(BiomeEffects.builder().fogColor(0xa080a0).waterColor(0x3f76e4).waterFogColor(0x050533).skyColor(0x000000).build()).build()),
+
+    CHERRY_GROVE(Biome.builder().name(NamespaceID.from("minecraft:cherry_grove")).category(Biome.Category.NONE).depth(0.125F).scale(0.05F).temperature(0.5f).downfall(0.8f).showPrecipitation(true).effects(BiomeEffects.builder().fogColor(0xc0d8ff).waterColor(0x5db7ef).waterFogColor(0x5db7ef).skyColor(0x7ba4ff).grassColor(0xb6db61).foliageColor(0xb6db61).build()).build()),
+
+    SUNFLOWER_PLAINS(Biome.builder().name(NamespaceID.from("minecraft:sunflower_plains")).category(Biome.Category.NONE).depth(0.125F).scale(0.05F).temperature(0.8f).downfall(0.4f).showPrecipitation(true).effects(BiomeEffects.builder().fogColor(0xc0d8ff).waterColor(0x3f76e4).waterFogColor(0x050533).skyColor(0x78a7ff).build()).build()),
+
+    BIRCH_FOREST(Biome.builder().name(NamespaceID.from("minecraft:birch_forest")).category(Biome.Category.NONE).depth(0.125F).scale(0.05F).temperature(0.6f).downfall(0.6f).showPrecipitation(true).effects(BiomeEffects.builder().fogColor(0xc0d8ff).waterColor(0x3f76e4).waterFogColor(0x050533).skyColor(0x7aa5ff).build()).build()),
+
+    WINDSWEPT_HILLS(Biome.builder().name(NamespaceID.from("minecraft:windswept_hills")).category(Biome.Category.NONE).depth(0.125F).scale(0.05F).temperature(0.2f).downfall(0.3f).showPrecipitation(true).effects(BiomeEffects.builder().fogColor(0xc0d8ff).waterColor(0x3f76e4).waterFogColor(0x050533).skyColor(0x7da2ff).build()).build()),
+
+    BAMBOO_JUNGLE(Biome.builder().name(NamespaceID.from("minecraft:bamboo_jungle")).category(Biome.Category.NONE).depth(0.125F).scale(0.05F).temperature(0.95f).downfall(0.9f).showPrecipitation(true).effects(BiomeEffects.builder().fogColor(0xc0d8ff).waterColor(0x3f76e4).waterFogColor(0x050533).skyColor(0x77a8ff).build()).build()),
+
+    WOODED_BADLANDS(Biome.builder().name(NamespaceID.from("minecraft:wooded_badlands")).category(Biome.Category.NONE).depth(0.125F).scale(0.05F).temperature(2.0f).downfall(0.0f).showPrecipitation(false).effects(BiomeEffects.builder().fogColor(0xc0d8ff).waterColor(0x3f76e4).waterFogColor(0x050533).skyColor(0x6eb1ff).grassColor(0x90814d).foliageColor(0x9e814d).build()).build()),
+
+    BADLANDS(Biome.builder().name(NamespaceID.from("minecraft:badlands")).category(Biome.Category.NONE).depth(0.125F).scale(0.05F).temperature(2.0f).downfall(0.0f).showPrecipitation(false).effects(BiomeEffects.builder().fogColor(0xc0d8ff).waterColor(0x3f76e4).waterFogColor(0x050533).skyColor(0x6eb1ff).grassColor(0x90814d).foliageColor(0x9e814d).build()).build()),
+
+    SAVANNA_PLATEAU(Biome.builder().name(NamespaceID.from("minecraft:savanna_plateau")).category(Biome.Category.NONE).depth(0.125F).scale(0.05F).temperature(2.0f).downfall(0.0f).showPrecipitation(false).effects(BiomeEffects.builder().fogColor(0xc0d8ff).waterColor(0x3f76e4).waterFogColor(0x050533).skyColor(0x6eb1ff).build()).build()),
+
+    BEACH(Biome.builder().name(NamespaceID.from("minecraft:beach")).category(Biome.Category.NONE).depth(0.125F).scale(0.05F).temperature(0.8f).downfall(0.4f).showPrecipitation(true).effects(BiomeEffects.builder().fogColor(0xc0d8ff).waterColor(0x3f76e4).waterFogColor(0x050533).skyColor(0x78a7ff).build()).build()),
+
+    DARK_FOREST(Biome.builder().name(NamespaceID.from("minecraft:dark_forest")).category(Biome.Category.NONE).depth(0.125F).scale(0.05F).temperature(0.7f).downfall(0.8f).showPrecipitation(true).effects(BiomeEffects.builder().fogColor(0xc0d8ff).waterColor(0x3f76e4).waterFogColor(0x050533).skyColor(0x79a6ff).build()).build()),
+
+    STONY_PEAKS(Biome.builder().name(NamespaceID.from("minecraft:stony_peaks")).category(Biome.Category.NONE).depth(0.125F).scale(0.05F).temperature(1.0f).downfall(0.3f).showPrecipitation(true).effects(BiomeEffects.builder().fogColor(0xc0d8ff).waterColor(0x3f76e4).waterFogColor(0x050533).skyColor(0x76a8ff).build()).build()),
+
+    MANGROVE_SWAMP(Biome.builder().name(NamespaceID.from("minecraft:mangrove_swamp")).category(Biome.Category.NONE).depth(0.125F).scale(0.05F).temperature(0.8f).downfall(0.9f).showPrecipitation(true).effects(BiomeEffects.builder().fogColor(0xc0d8ff).waterColor(0x3a7a6a).waterFogColor(0x4d7a60).skyColor(0x78a7ff).foliageColor(0x8db127).build()).build()),
+
+    SPARSE_JUNGLE(Biome.builder().name(NamespaceID.from("minecraft:sparse_jungle")).category(Biome.Category.NONE).depth(0.125F).scale(0.05F).temperature(0.95f).downfall(0.8f).showPrecipitation(true).effects(BiomeEffects.builder().fogColor(0xc0d8ff).waterColor(0x3f76e4).waterFogColor(0x050533).skyColor(0x77a8ff).build()).build()),
+
+    LUKEWARM_OCEAN(Biome.builder().name(NamespaceID.from("minecraft:lukewarm_ocean")).category(Biome.Category.NONE).depth(0.125F).scale(0.05F).temperature(0.5f).downfall(0.5f).showPrecipitation(true).effects(BiomeEffects.builder().fogColor(0xc0d8ff).waterColor(0x45adf2).waterFogColor(0x041633).skyColor(0x7ba4ff).build()).build()),
+
+    RIVER(Biome.builder().name(NamespaceID.from("minecraft:river")).category(Biome.Category.NONE).depth(0.125F).scale(0.05F).temperature(0.5f).downfall(0.5f).showPrecipitation(true).effects(BiomeEffects.builder().fogColor(0xc0d8ff).waterColor(0x3f76e4).waterFogColor(0x050533).skyColor(0x7ba4ff).build()).build()),
+
+    STONY_SHORE(Biome.builder().name(NamespaceID.from("minecraft:stony_shore")).category(Biome.Category.NONE).depth(0.125F).scale(0.05F).temperature(0.2f).downfall(0.3f).showPrecipitation(true).effects(BiomeEffects.builder().fogColor(0xc0d8ff).waterColor(0x3f76e4).waterFogColor(0x050533).skyColor(0x7da2ff).build()).build()),
+
+    WARPED_FOREST(Biome.builder().name(NamespaceID.from("minecraft:warped_forest")).category(Biome.Category.NONE).depth(0.125F).scale(0.05F).temperature(2.0f).downfall(0.0f).showPrecipitation(false).effects(BiomeEffects.builder().fogColor(0x1a051a).waterColor(0x3f76e4).waterFogColor(0x050533).skyColor(0x6eb1ff).build()).build()),
+
+    SNOWY_PLAINS(Biome.builder().name(NamespaceID.from("minecraft:snowy_plains")).category(Biome.Category.NONE).depth(0.125F).scale(0.05F).temperature(0.0f).downfall(0.5f).showPrecipitation(true).effects(BiomeEffects.builder().fogColor(0xc0d8ff).waterColor(0x3f76e4).waterFogColor(0x050533).skyColor(0x7fa1ff).build()).build()),
+
+    DRIPSTONE_CAVES(Biome.builder().name(NamespaceID.from("minecraft:dripstone_caves")).category(Biome.Category.NONE).depth(0.125F).scale(0.05F).temperature(0.8f).downfall(0.4f).showPrecipitation(true).effects(BiomeEffects.builder().fogColor(0xc0d8ff).waterColor(0x3f76e4).waterFogColor(0x050533).skyColor(0x78a7ff).build()).build()),
+
+    SNOWY_TAIGA(Biome.builder().name(NamespaceID.from("minecraft:snowy_taiga")).category(Biome.Category.NONE).depth(0.125F).scale(0.05F).temperature(-0.5f).downfall(0.4f).showPrecipitation(true).effects(BiomeEffects.builder().fogColor(0xc0d8ff).waterColor(0x3d57d6).waterFogColor(0x050533).skyColor(0x839eff).build()).build()),
+
+    GROVE(Biome.builder().name(NamespaceID.from("minecraft:grove")).category(Biome.Category.NONE).depth(0.125F).scale(0.05F).temperature(-0.2f).downfall(0.8f).showPrecipitation(true).effects(BiomeEffects.builder().fogColor(0xc0d8ff).waterColor(0x3f76e4).waterFogColor(0x050533).skyColor(0x81a0ff).build()).build()),
+
+    SWAMP(Biome.builder().name(NamespaceID.from("minecraft:swamp")).category(Biome.Category.NONE).depth(0.125F).scale(0.05F).temperature(0.8f).downfall(0.9f).showPrecipitation(true).effects(BiomeEffects.builder().fogColor(0xc0d8ff).waterColor(0x617b64).waterFogColor(0x232317).skyColor(0x78a7ff).foliageColor(0x6a7039).build()).build()),
+
+    JAGGED_PEAKS(Biome.builder().name(NamespaceID.from("minecraft:jagged_peaks")).category(Biome.Category.NONE).depth(0.125F).scale(0.05F).temperature(-0.7f).downfall(0.9f).showPrecipitation(true).effects(BiomeEffects.builder().fogColor(0xc0d8ff).waterColor(0x3f76e4).waterFogColor(0x050533).skyColor(0x859dff).build()).build()),
+
+    COLD_OCEAN(Biome.builder().name(NamespaceID.from("minecraft:cold_ocean")).category(Biome.Category.NONE).depth(0.125F).scale(0.05F).temperature(0.5f).downfall(0.5f).showPrecipitation(true).effects(BiomeEffects.builder().fogColor(0xc0d8ff).waterColor(0x3d57d6).waterFogColor(0x050533).skyColor(0x7ba4ff).build()).build()),
+
+    FOREST(Biome.builder().name(NamespaceID.from("minecraft:forest")).category(Biome.Category.NONE).depth(0.125F).scale(0.05F).temperature(0.7f).downfall(0.8f).showPrecipitation(true).effects(BiomeEffects.builder().fogColor(0xc0d8ff).waterColor(0x3f76e4).waterFogColor(0x050533).skyColor(0x79a6ff).build()).build()),
+
+    LUSH_CAVES(Biome.builder().name(NamespaceID.from("minecraft:lush_caves")).category(Biome.Category.NONE).depth(0.125F).scale(0.05F).temperature(0.5f).downfall(0.5f).showPrecipitation(true).effects(BiomeEffects.builder().fogColor(0xc0d8ff).waterColor(0x3f76e4).waterFogColor(0x050533).skyColor(0x7ba4ff).build()).build()),
+
+    BASALT_DELTAS(Biome.builder().name(NamespaceID.from("minecraft:basalt_deltas")).category(Biome.Category.NONE).depth(0.125F).scale(0.05F).temperature(2.0f).downfall(0.0f).showPrecipitation(false).effects(BiomeEffects.builder().fogColor(0x685f70).waterColor(0x3f76e4).waterFogColor(0x050533).skyColor(0x6eb1ff).build()).build()),
+
+    DEEP_COLD_OCEAN(Biome.builder().name(NamespaceID.from("minecraft:deep_cold_ocean")).category(Biome.Category.NONE).depth(0.125F).scale(0.05F).temperature(0.5f).downfall(0.5f).showPrecipitation(true).effects(BiomeEffects.builder().fogColor(0xc0d8ff).waterColor(0x3d57d6).waterFogColor(0x050533).skyColor(0x7ba4ff).build()).build()),
+
+    ICE_SPIKES(Biome.builder().name(NamespaceID.from("minecraft:ice_spikes")).category(Biome.Category.NONE).depth(0.125F).scale(0.05F).temperature(0.0f).downfall(0.5f).showPrecipitation(true).effects(BiomeEffects.builder().fogColor(0xc0d8ff).waterColor(0x3f76e4).waterFogColor(0x050533).skyColor(0x7fa1ff).build()).build()),
+
+    END_MIDLANDS(Biome.builder().name(NamespaceID.from("minecraft:end_midlands")).category(Biome.Category.NONE).depth(0.125F).scale(0.05F).temperature(0.5f).downfall(0.5f).showPrecipitation(false).effects(BiomeEffects.builder().fogColor(0xa080a0).waterColor(0x3f76e4).waterFogColor(0x050533).skyColor(0x000000).build()).build()),
+
+    FROZEN_OCEAN(Biome.builder().name(NamespaceID.from("minecraft:frozen_ocean")).category(Biome.Category.NONE).depth(0.125F).scale(0.05F).temperature(0.0f).downfall(0.5f).showPrecipitation(true).effects(BiomeEffects.builder().fogColor(0xc0d8ff).waterColor(0x3938c9).waterFogColor(0x050533).skyColor(0x7fa1ff).build()).build()),
+
+    DESERT(Biome.builder().name(NamespaceID.from("minecraft:desert")).category(Biome.Category.NONE).depth(0.125F).scale(0.05F).temperature(2.0f).downfall(0.0f).showPrecipitation(false).effects(BiomeEffects.builder().fogColor(0xc0d8ff).waterColor(0x3f76e4).waterFogColor(0x050533).skyColor(0x6eb1ff).build()).build()),
+
+    DEEP_FROZEN_OCEAN(Biome.builder().name(NamespaceID.from("minecraft:deep_frozen_ocean")).category(Biome.Category.NONE).depth(0.125F).scale(0.05F).temperature(0.5f).downfall(0.5f).showPrecipitation(true).effects(BiomeEffects.builder().fogColor(0xc0d8ff).waterColor(0x3938c9).waterFogColor(0x050533).skyColor(0x7ba4ff).build()).build()),
+
+    WINDSWEPT_FOREST(Biome.builder().name(NamespaceID.from("minecraft:windswept_forest")).category(Biome.Category.NONE).depth(0.125F).scale(0.05F).temperature(0.2f).downfall(0.3f).showPrecipitation(true).effects(BiomeEffects.builder().fogColor(0xc0d8ff).waterColor(0x3f76e4).waterFogColor(0x050533).skyColor(0x7da2ff).build()).build()),
+
+    JUNGLE(Biome.builder().name(NamespaceID.from("minecraft:jungle")).category(Biome.Category.NONE).depth(0.125F).scale(0.05F).temperature(0.95f).downfall(0.9f).showPrecipitation(true).effects(BiomeEffects.builder().fogColor(0xc0d8ff).waterColor(0x3f76e4).waterFogColor(0x050533).skyColor(0x77a8ff).build()).build()),
+
+    OCEAN(Biome.builder().name(NamespaceID.from("minecraft:ocean")).category(Biome.Category.NONE).depth(0.125F).scale(0.05F).temperature(0.5f).downfall(0.5f).showPrecipitation(true).effects(BiomeEffects.builder().fogColor(0xc0d8ff).waterColor(0x3f76e4).waterFogColor(0x050533).skyColor(0x7ba4ff).build()).build()),
+
+    OLD_GROWTH_SPRUCE_TAIGA(Biome.builder().name(NamespaceID.from("minecraft:old_growth_spruce_taiga")).category(Biome.Category.NONE).depth(0.125F).scale(0.05F).temperature(0.25f).downfall(0.8f).showPrecipitation(true).effects(BiomeEffects.builder().fogColor(0xc0d8ff).waterColor(0x3f76e4).waterFogColor(0x050533).skyColor(0x7da3ff).build()).build()),
+
+    SNOWY_BEACH(Biome.builder().name(NamespaceID.from("minecraft:snowy_beach")).category(Biome.Category.NONE).depth(0.125F).scale(0.05F).temperature(0.05f).downfall(0.3f).showPrecipitation(true).effects(BiomeEffects.builder().fogColor(0xc0d8ff).waterColor(0x3d57d6).waterFogColor(0x050533).skyColor(0x7fa1ff).build()).build()),
+
+    WINDSWEPT_SAVANNA(Biome.builder().name(NamespaceID.from("minecraft:windswept_savanna")).category(Biome.Category.NONE).depth(0.125F).scale(0.05F).temperature(2.0f).downfall(0.0f).showPrecipitation(false).effects(BiomeEffects.builder().fogColor(0xc0d8ff).waterColor(0x3f76e4).waterFogColor(0x050533).skyColor(0x6eb1ff).build()).build()),
+
+    END_BARRENS(Biome.builder().name(NamespaceID.from("minecraft:end_barrens")).category(Biome.Category.NONE).depth(0.125F).scale(0.05F).temperature(0.5f).downfall(0.5f).showPrecipitation(false).effects(BiomeEffects.builder().fogColor(0xa080a0).waterColor(0x3f76e4).waterFogColor(0x050533).skyColor(0x000000).build()).build()),
+
+    WARM_OCEAN(Biome.builder().name(NamespaceID.from("minecraft:warm_ocean")).category(Biome.Category.NONE).depth(0.125F).scale(0.05F).temperature(0.5f).downfall(0.5f).showPrecipitation(true).effects(BiomeEffects.builder().fogColor(0xc0d8ff).waterColor(0x43d5ee).waterFogColor(0x041f33).skyColor(0x7ba4ff).build()).build()),
+
+    DEEP_LUKEWARM_OCEAN(Biome.builder().name(NamespaceID.from("minecraft:deep_lukewarm_ocean")).category(Biome.Category.NONE).depth(0.125F).scale(0.05F).temperature(0.5f).downfall(0.5f).showPrecipitation(true).effects(BiomeEffects.builder().fogColor(0xc0d8ff).waterColor(0x45adf2).waterFogColor(0x041633).skyColor(0x7ba4ff).build()).build()),
+
+    FLOWER_FOREST(Biome.builder().name(NamespaceID.from("minecraft:flower_forest")).category(Biome.Category.NONE).depth(0.125F).scale(0.05F).temperature(0.7f).downfall(0.8f).showPrecipitation(true).effects(BiomeEffects.builder().fogColor(0xc0d8ff).waterColor(0x3f76e4).waterFogColor(0x050533).skyColor(0x79a6ff).build()).build()),
+
+    SOUL_SAND_VALLEY(Biome.builder().name(NamespaceID.from("minecraft:soul_sand_valley")).category(Biome.Category.NONE).depth(0.125F).scale(0.05F).temperature(2.0f).downfall(0.0f).showPrecipitation(false).effects(BiomeEffects.builder().fogColor(0x1b4745).waterColor(0x3f76e4).waterFogColor(0x050533).skyColor(0x6eb1ff).build()).build()),
+
+    NETHER_WASTES(Biome.builder().name(NamespaceID.from("minecraft:nether_wastes")).category(Biome.Category.NONE).depth(0.125F).scale(0.05F).temperature(2.0f).downfall(0.0f).showPrecipitation(false).effects(BiomeEffects.builder().fogColor(0x330808).waterColor(0x3f76e4).waterFogColor(0x050533).skyColor(0x6eb1ff).build()).build()),
+
+    FROZEN_PEAKS(Biome.builder().name(NamespaceID.from("minecraft:frozen_peaks")).category(Biome.Category.NONE).depth(0.125F).scale(0.05F).temperature(-0.7f).downfall(0.9f).showPrecipitation(true).effects(BiomeEffects.builder().fogColor(0xc0d8ff).waterColor(0x3f76e4).waterFogColor(0x050533).skyColor(0x859dff).build()).build()),
+
+    THE_END(Biome.builder().name(NamespaceID.from("minecraft:the_end")).category(Biome.Category.NONE).depth(0.125F).scale(0.05F).temperature(0.5f).downfall(0.5f).showPrecipitation(false).effects(BiomeEffects.builder().fogColor(0xa080a0).waterColor(0x3f76e4).waterFogColor(0x050533).skyColor(0x000000).build()).build()),
+
+    SMALL_END_ISLANDS(Biome.builder().name(NamespaceID.from("minecraft:small_end_islands")).category(Biome.Category.NONE).depth(0.125F).scale(0.05F).temperature(0.5f).downfall(0.5f).showPrecipitation(false).effects(BiomeEffects.builder().fogColor(0xa080a0).waterColor(0x3f76e4).waterFogColor(0x050533).skyColor(0x000000).build()).build()),
+
+    OLD_GROWTH_BIRCH_FOREST(Biome.builder().name(NamespaceID.from("minecraft:old_growth_birch_forest")).category(Biome.Category.NONE).depth(0.125F).scale(0.05F).temperature(0.6f).downfall(0.6f).showPrecipitation(true).effects(BiomeEffects.builder().fogColor(0xc0d8ff).waterColor(0x3f76e4).waterFogColor(0x050533).skyColor(0x7aa5ff).build()).build()),
+
+    CRIMSON_FOREST(Biome.builder().name(NamespaceID.from("minecraft:crimson_forest")).category(Biome.Category.NONE).depth(0.125F).scale(0.05F).temperature(2.0f).downfall(0.0f).showPrecipitation(false).effects(BiomeEffects.builder().fogColor(0x330303).waterColor(0x3f76e4).waterFogColor(0x050533).skyColor(0x6eb1ff).build()).build()),
+
+    THE_VOID(Biome.builder().name(NamespaceID.from("minecraft:the_void")).category(Biome.Category.NONE).depth(0.125F).scale(0.05F).temperature(0.5f).downfall(0.5f).showPrecipitation(false).effects(BiomeEffects.builder().fogColor(0xc0d8ff).waterColor(0x3f76e4).waterFogColor(0x050533).skyColor(0x7ba4ff).build()).build()),
+
+    DEEP_DARK(Biome.builder().name(NamespaceID.from("minecraft:deep_dark")).category(Biome.Category.NONE).depth(0.125F).scale(0.05F).temperature(0.8f).downfall(0.4f).showPrecipitation(true).effects(BiomeEffects.builder().fogColor(0xc0d8ff).waterColor(0x3f76e4).waterFogColor(0x050533).skyColor(0x78a7ff).build()).build()),
+
+    MEADOW(Biome.builder().name(NamespaceID.from("minecraft:meadow")).category(Biome.Category.NONE).depth(0.125F).scale(0.05F).temperature(0.5f).downfall(0.8f).showPrecipitation(true).effects(BiomeEffects.builder().fogColor(0xc0d8ff).waterColor(0x0e4ecf).waterFogColor(0x050533).skyColor(0x7ba4ff).build()).build()),
+
+    WINDSWEPT_GRAVELLY_HILLS(Biome.builder().name(NamespaceID.from("minecraft:windswept_gravelly_hills")).category(Biome.Category.NONE).depth(0.125F).scale(0.05F).temperature(0.2f).downfall(0.3f).showPrecipitation(true).effects(BiomeEffects.builder().fogColor(0xc0d8ff).waterColor(0x3f76e4).waterFogColor(0x050533).skyColor(0x7da2ff).build()).build()),
+
+    SAVANNA(Biome.builder().name(NamespaceID.from("minecraft:savanna")).category(Biome.Category.NONE).depth(0.125F).scale(0.05F).temperature(2.0f).downfall(0.0f).showPrecipitation(false).effects(BiomeEffects.builder().fogColor(0xc0d8ff).waterColor(0x3f76e4).waterFogColor(0x050533).skyColor(0x6eb1ff).build()).build()),
+
+    PLAINS(Biome.builder().name(NamespaceID.from("minecraft:plains")).category(Biome.Category.NONE).depth(0.125F).scale(0.05F).temperature(0.8f).downfall(0.4f).showPrecipitation(true).effects(BiomeEffects.builder().fogColor(0xc0d8ff).waterColor(0x3f76e4).waterFogColor(0x050533).skyColor(0x78a7ff).build()).build());
+
+    private static final Biomes[] VALUES = Biomes.values();
+
+    private final Biome biome;
+
+    Biomes(@NotNull Biome biome) {
+        this.biome = biome;
+    }
+
+    public @NotNull Biome biome() {
+        return this.biome;
+    }
+
+    @Nullable
+    public static Biomes getValue(int id) {
+        return VALUES[id];
+    }
+
+    public static void registerBiomes() {
+        for (var biome : VALUES) {
+            MinecraftServer.getBiomeManager().addBiome(biome.biome());
+        }
+    }
+}
diff --git a/src/main/java/net/minestom/server/ServerProcessImpl.java b/src/main/java/net/minestom/server/ServerProcessImpl.java
index 308c99d4233cc68a0915800b3b8d3cead123f2a1..2d917a7d8185c2cdb8605bf2b01a843e2decfe73 100644
--- a/src/main/java/net/minestom/server/ServerProcessImpl.java
+++ b/src/main/java/net/minestom/server/ServerProcessImpl.java
@@ -37,6 +37,7 @@ import net.minestom.server.timer.SchedulerManager;
 import net.minestom.server.utils.collection.MappedCollection;
 import net.minestom.server.world.DimensionTypeManager;
 import net.minestom.server.world.biomes.BiomeManager;
+import net.minestom.server.world.biomes.Biomes;
 import org.jetbrains.annotations.NotNull;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
@@ -231,6 +232,9 @@ final class ServerProcessImpl implements ServerProcess {
 
         LOGGER.info("{} server started successfully.", MinecraftServer.getBrandName());
 
+        Biomes.registerBiomes();
+        LOGGER.info("Register Biomes({})", biome.unmodifiableCollection().size());
+
         if (MinecraftServer.isTerminalEnabled()) {
             MinestomTerminal.start();
         }
diff --git a/src/main/java/net/minestom/server/extras/optifine/OptifineSupport.java b/src/main/java/net/minestom/server/extras/optifine/OptifineSupport.java
deleted file mode 100644
index 9f2a9584a32025b937d8b27187c1ed3b1688e973..0000000000000000000000000000000000000000
--- a/src/main/java/net/minestom/server/extras/optifine/OptifineSupport.java
+++ /dev/null
@@ -1,36 +0,0 @@
-package net.minestom.server.extras.optifine;
-
-import net.minestom.server.MinecraftServer;
-import net.minestom.server.utils.NamespaceID;
-import net.minestom.server.utils.validate.Check;
-import net.minestom.server.world.biomes.Biome;
-import net.minestom.server.world.biomes.BiomeManager;
-
-/**
- * Hacky class for Optifine because of an issue making the client crash if biomes 'swamp' and 'swamp_hills'
- * are not registered.
- * <p>
- * Can be removed anytime, hope that it will be fixed.
- */
-public final class OptifineSupport {
-
-    private static volatile boolean enabled;
-
-    /**
-     * Enables optifine support by registering the required biomes.
-     *
-     * @throws IllegalStateException if optifine support is already enabled
-     */
-    public static void enable() {
-        Check.stateCondition(enabled, "Optifine support is already enabled!");
-        OptifineSupport.enabled = true;
-
-        BiomeManager biomeManager = MinecraftServer.getBiomeManager();
-        biomeManager.addBiome(Biome.builder().name(NamespaceID.from("minecraft:swamp")).build());
-        biomeManager.addBiome(Biome.builder().name(NamespaceID.from("minecraft:swamp_hills")).build());
-    }
-
-    public static boolean isEnabled() {
-        return enabled;
-    }
-}
diff --git a/src/main/java/net/minestom/server/instance/AnvilLoader.java b/src/main/java/net/minestom/server/instance/AnvilLoader.java
index 9111770d834c98d3c31a60207d57462381f61177..6873897aef672fdd78e8f154427465809bf5c81c 100644
--- a/src/main/java/net/minestom/server/instance/AnvilLoader.java
+++ b/src/main/java/net/minestom/server/instance/AnvilLoader.java
@@ -9,6 +9,7 @@ import net.minestom.server.instance.block.BlockHandler;
 import net.minestom.server.utils.NamespaceID;
 import net.minestom.server.utils.async.AsyncUtils;
 import net.minestom.server.world.biomes.Biome;
+import net.minestom.server.world.biomes.Biomes;
 import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
 import org.jglrxavpok.hephaistos.mca.*;
@@ -34,7 +35,7 @@ import java.util.concurrent.ConcurrentHashMap;
 
 public class AnvilLoader implements IChunkLoader {
     private final static Logger LOGGER = LoggerFactory.getLogger(AnvilLoader.class);
-    private static final Biome BIOME = Biome.PLAINS;
+    private static final Biome BIOME = Biomes.PLAINS.biome();
 
     private final Map<String, RegionFile> alreadyLoaded = new ConcurrentHashMap<>();
     private final Path path;
diff --git a/src/main/java/net/minestom/server/world/biomes/Biome.java b/src/main/java/net/minestom/server/world/biomes/Biome.java
index b58c200c1fb4787c6c383782426cf5f97b1d0943..8ceadbf3b891f883f19a0f68c4ce22e4f6174498 100644
--- a/src/main/java/net/minestom/server/world/biomes/Biome.java
+++ b/src/main/java/net/minestom/server/world/biomes/Biome.java
@@ -1,5 +1,6 @@
 package net.minestom.server.world.biomes;
 
+import java.util.Locale;
 import net.minestom.server.coordinate.Point;
 import net.minestom.server.utils.NamespaceID;
 import net.minestom.server.utils.validate.Check;
@@ -7,11 +8,8 @@ import org.jetbrains.annotations.NotNull;
 import org.jglrxavpok.hephaistos.nbt.NBT;
 import org.jglrxavpok.hephaistos.nbt.NBTCompound;
 
-import java.util.Locale;
-import java.util.concurrent.atomic.AtomicInteger;
-
 public final class Biome {
-    public static final AtomicInteger ID_COUNTER = new AtomicInteger(0);
+
     private static final BiomeEffects DEFAULT_EFFECTS = BiomeEffects.builder()
             .fogColor(0xC0D8FF)
             .skyColor(0x78A7FF)
@@ -19,41 +17,35 @@ public final class Biome {
             .waterFogColor(0x50533)
             .build();
 
-    //A plains biome has to be registered or else minecraft will crash
-    public static final Biome PLAINS = Biome.builder()
-            .category(Category.NONE)
-            .name(NamespaceID.from("minecraft:plains"))
-            .temperature(0.8F)
-            .downfall(0.4F)
-            .depth(0.125F)
-            .scale(0.05F)
-            .effects(DEFAULT_EFFECTS)
-            .build();
-
-    private final int id = ID_COUNTER.getAndIncrement();
-
+    private int id;
     private final NamespaceID name;
     private final float depth;
     private final float temperature;
     private final float scale;
     private final float downfall;
+    private final boolean hasPrecipitation;
     private final Category category;
     private final BiomeEffects effects;
     private final Precipitation precipitation;
     private final TemperatureModifier temperatureModifier;
 
-    Biome(NamespaceID name, float depth, float temperature, float scale, float downfall, Category category, BiomeEffects effects, Precipitation precipitation, TemperatureModifier temperatureModifier) {
+    Biome(NamespaceID name, float depth, float temperature, float scale, float downfall, boolean hasPrecipitation, Category category, BiomeEffects effects, Precipitation precipitation, TemperatureModifier temperatureModifier) {
         this.name = name;
         this.depth = depth;
         this.temperature = temperature;
         this.scale = scale;
         this.downfall = downfall;
+        this.hasPrecipitation = hasPrecipitation;
         this.category = category;
         this.effects = effects;
         this.precipitation = precipitation;
         this.temperatureModifier = temperatureModifier;
     }
 
+    void setId(final int id) {
+        this.id = id;
+    }
+
     public static Builder builder() {
         return new Builder();
     }
@@ -72,8 +64,11 @@ public final class Biome {
                 element.setFloat("scale", scale);
                 element.setFloat("downfall", downfall);
                 element.setString("category", category.name().toLowerCase(Locale.ROOT));
-                // Todo: Bad fix
-                element.setByte("has_precipitation", (byte) 1); // Microtus - 1.19.4 update
+                if (isHasPrecipitation()) {
+                    element.setByte("has_precipitation", (byte) 1); // Microtus - 1.19.4 update
+                } else {
+                    element.setByte("has_precipitation", (byte) 0); // Microtus - 1.19.4 update
+                }
                 element.setString("precipitation", precipitation.name().toLowerCase(Locale.ROOT));
                 if (temperatureModifier != TemperatureModifier.NONE)
                     element.setString("temperature_modifier", temperatureModifier.name().toLowerCase(Locale.ROOT));
@@ -90,6 +85,10 @@ public final class Biome {
         return this.name;
     }
 
+    public boolean isHasPrecipitation() {
+        return this.hasPrecipitation;
+    }
+
     public float depth() {
         return this.depth;
     }
@@ -143,6 +142,7 @@ public final class Biome {
         private float temperature = 0.25f;
         private float scale = 0.2f;
         private float downfall = 0.8f;
+        private boolean hasPrecipitation = false;
         private Category category = Category.NONE;
         private BiomeEffects effects = DEFAULT_EFFECTS;
         private Precipitation precipitation = Precipitation.RAIN;
@@ -156,6 +156,11 @@ public final class Biome {
             return this;
         }
 
+        public Builder showPrecipitation(boolean hasPrecipitation) {
+            this.hasPrecipitation = hasPrecipitation;
+            return this;
+        }
+
         public Builder depth(float depth) {
             this.depth = depth;
             return this;
@@ -197,7 +202,8 @@ public final class Biome {
         }
 
         public Biome build() {
-            return new Biome(name, depth, temperature, scale, downfall, category, effects, precipitation, temperatureModifier);
+            return new Biome(name, depth, temperature, scale, downfall, hasPrecipitation , category, effects, precipitation,
+                    temperatureModifier);
         }
     }
 
diff --git a/src/main/java/net/minestom/server/world/biomes/BiomeManager.java b/src/main/java/net/minestom/server/world/biomes/BiomeManager.java
index 61c63ecf1235b8ae94d8e0c0700bd9beebdda681..8532902956806455dbbb378ba56e08ffb43ecf2d 100644
--- a/src/main/java/net/minestom/server/world/biomes/BiomeManager.java
+++ b/src/main/java/net/minestom/server/world/biomes/BiomeManager.java
@@ -1,27 +1,30 @@
 package net.minestom.server.world.biomes;
 
-import it.unimi.dsi.fastutil.ints.Int2ObjectMap;
-import it.unimi.dsi.fastutil.ints.Int2ObjectOpenHashMap;
 import net.minestom.server.utils.NamespaceID;
 import org.jglrxavpok.hephaistos.nbt.NBT;
 import org.jglrxavpok.hephaistos.nbt.NBTCompound;
 import org.jglrxavpok.hephaistos.nbt.NBTType;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 import java.util.Collection;
 import java.util.Collections;
 import java.util.Map;
+import java.util.List;
+import java.util.ArrayList;
+import java.util.stream.IntStream;
 
 
 /**
  * Allows servers to register custom dimensions. Also used during player joining to send the list of all existing dimensions.
  * <p>
- * Contains {@link Biome#PLAINS} by default but can be removed.
+ * Contains {@link Biomes#PLAINS} by default but can be removed.
  */
 public final class BiomeManager {
-    private final Int2ObjectMap<Biome> biomes = new Int2ObjectOpenHashMap<>();
+    private final List<Biome> biomes = new ArrayList<>();
+    private final static Logger LOGGER = LoggerFactory.getLogger(BiomeManager.class);
 
     public BiomeManager() {
-        addBiome(Biome.PLAINS);
     }
 
     /**
@@ -30,7 +33,11 @@ public final class BiomeManager {
      * @param biome the biome to add
      */
     public synchronized void addBiome(Biome biome) {
-        this.biomes.put(biome.id(), biome);
+        if (biomes.stream().anyMatch(listBiome -> listBiome.name().equals(biome.name()))) {
+            LOGGER.error("You cannot register the biome because it is already registered. Please remove the biome to overwrite it");
+            return;
+        }
+        this.biomes.add(biome);
     }
 
     /**
@@ -39,7 +46,7 @@ public final class BiomeManager {
      * @param biome the biome to remove
      */
     public synchronized void removeBiome(Biome biome) {
-        this.biomes.remove(biome.id());
+        this.biomes.remove(biome);
     }
 
     /**
@@ -48,7 +55,7 @@ public final class BiomeManager {
      * @return an immutable copy of the biomes already registered
      */
     public synchronized Collection<Biome> unmodifiableCollection() {
-        return Collections.unmodifiableCollection(biomes.values());
+        return Collections.unmodifiableCollection(biomes);
     }
 
     /**
@@ -63,7 +70,7 @@ public final class BiomeManager {
 
     public synchronized Biome getByName(NamespaceID namespaceID) {
         Biome biome = null;
-        for (final Biome biomeT : biomes.values()) {
+        for (final Biome biomeT : biomes) {
             if (biomeT.name().equals(namespaceID)) {
                 biome = biomeT;
                 break;
@@ -73,8 +80,16 @@ public final class BiomeManager {
     }
 
     public synchronized NBTCompound toNBT() {
+        var biomes = new ArrayList<>(unmodifiableCollection());
+        List<NBTCompound> nbtCompounds = IntStream
+                .range(0, biomes.size())
+                .mapToObj(i -> {
+                    var biome = biomes.get(i);
+                    biome.setId(i);
+                    return biome;
+                }).map(Biome::toNbt).toList();
         return NBT.Compound(Map.of(
                 "type", NBT.String("minecraft:worldgen/biome"),
-                "value", NBT.List(NBTType.TAG_Compound, biomes.values().stream().map(Biome::toNbt).toList())));
+                "value", NBT.List(NBTType.TAG_Compound, nbtCompounds)));
     }
 }
diff --git a/src/main/java/net/minestom/server/world/biomes/BiomeParticle.java b/src/main/java/net/minestom/server/world/biomes/BiomeParticle.java
index f2f3edeb37600c2b603c292a3d7179f61848f0c2..ce20735425d344574ad17001f410bd41db1dbda3 100644
--- a/src/main/java/net/minestom/server/world/biomes/BiomeParticle.java
+++ b/src/main/java/net/minestom/server/world/biomes/BiomeParticle.java
@@ -1,70 +1,15 @@
 package net.minestom.server.world.biomes;
 
-import net.minestom.server.instance.block.Block;
-import net.minestom.server.item.ItemStack;
-import net.minestom.server.utils.NamespaceID;
+import net.minestom.server.world.biomes.particle.BiomeOption;
 import org.jglrxavpok.hephaistos.nbt.NBT;
 import org.jglrxavpok.hephaistos.nbt.NBTCompound;
 
 import java.util.Map;
 
-public record BiomeParticle(float probability, Option option) {
+public record BiomeParticle(float probability, BiomeOption option) {
     public NBTCompound toNbt() {
         return NBT.Compound(Map.of(
                 "probability", NBT.Float(probability),
                 "options", option.toNbt()));
     }
-
-    public interface Option {
-        NBTCompound toNbt();
-    }
-
-    public record BlockOption(Block block) implements Option {
-        //TODO also can be falling_dust
-        private static final String type = "block";
-
-        @Override
-        public NBTCompound toNbt() {
-            return NBT.Compound(nbtCompound -> {
-                nbtCompound.setString("type", type);
-                nbtCompound.setString("Name", block.name());
-                Map<String, String> propertiesMap = block.properties();
-                if (propertiesMap.size() != 0) {
-                    nbtCompound.set("Properties", NBT.Compound(p -> propertiesMap.forEach(p::setString)));
-                }
-            });
-        }
-    }
-
-    public record DustOption(float red, float green, float blue, float scale) implements Option {
-        private static final String type = "dust";
-
-        @Override
-        public NBTCompound toNbt() {
-            return NBT.Compound(Map.of(
-                    "type", NBT.String(type),
-                    "r", NBT.Float(red),
-                    "g", NBT.Float(green),
-                    "b", NBT.Float(blue),
-                    "scale", NBT.Float(scale)));
-        }
-    }
-
-    public record ItemOption(ItemStack item) implements Option {
-        private static final String type = "item";
-
-        @Override
-        public NBTCompound toNbt() {
-            //todo test count might be wrong type
-            NBTCompound nbtCompound = item.meta().toNBT();
-            return nbtCompound.modify(n -> n.setString("type", type));
-        }
-    }
-
-    public record NormalOption(NamespaceID type) implements Option {
-        @Override
-        public NBTCompound toNbt() {
-            return NBT.Compound(Map.of("type", NBT.String(type.toString())));
-        }
-    }
 }
diff --git a/src/main/java/net/minestom/server/world/biomes/particle/BiomeOption.java b/src/main/java/net/minestom/server/world/biomes/particle/BiomeOption.java
new file mode 100644
index 0000000000000000000000000000000000000000..07ffda2ba5d3b7664db1d1ee9f850a0504397636
--- /dev/null
+++ b/src/main/java/net/minestom/server/world/biomes/particle/BiomeOption.java
@@ -0,0 +1,16 @@
+package net.minestom.server.world.biomes.particle;
+
+import net.minestom.server.particle.Particle;
+import org.jetbrains.annotations.NotNull;
+import org.jglrxavpok.hephaistos.nbt.NBTCompound;
+
+/**
+ * Just a generic interface to hold data about biome particles
+ * @author TheMeinerLP
+ */
+public interface BiomeOption {
+    @NotNull
+    NBTCompound toNbt();
+    @NotNull
+    Particle getType();
+}
diff --git a/src/main/java/net/minestom/server/world/biomes/particle/BlockOption.java b/src/main/java/net/minestom/server/world/biomes/particle/BlockOption.java
new file mode 100644
index 0000000000000000000000000000000000000000..117d9a1ec79de53bf3e2df2a48c8a8a0010d3672
--- /dev/null
+++ b/src/main/java/net/minestom/server/world/biomes/particle/BlockOption.java
@@ -0,0 +1,32 @@
+package net.minestom.server.world.biomes.particle;
+
+import net.minestom.server.instance.block.Block;
+import net.minestom.server.particle.Particle;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.NotNull;
+import org.jglrxavpok.hephaistos.nbt.NBT;
+import org.jglrxavpok.hephaistos.nbt.NBTCompound;
+
+import java.util.Map;
+
+@ApiStatus.Experimental
+public record BlockOption(@NotNull Block block) implements BiomeOption {
+
+    @Override
+    public @NotNull NBTCompound toNbt() {
+        return NBT.Compound(nbtCompound -> {
+            nbtCompound.setString("type", getType().namespace().toString());
+            nbtCompound.setString("Name", block.name());
+            Map<String, String> propertiesMap = block.properties();
+            if (!propertiesMap.isEmpty()) {
+                nbtCompound.set("Properties", NBT.Compound(p -> propertiesMap.forEach(p::setString)));
+            }
+        });
+    }
+
+    @Override
+    public @NotNull Particle getType() {
+        return Particle.BLOCK;
+    }
+
+}
diff --git a/src/main/java/net/minestom/server/world/biomes/particle/DefaultOption.java b/src/main/java/net/minestom/server/world/biomes/particle/DefaultOption.java
new file mode 100644
index 0000000000000000000000000000000000000000..1ddad4d7ea400653786608669689a84ed3078143
--- /dev/null
+++ b/src/main/java/net/minestom/server/world/biomes/particle/DefaultOption.java
@@ -0,0 +1,20 @@
+package net.minestom.server.world.biomes.particle;
+
+import net.minestom.server.particle.Particle;
+import org.jetbrains.annotations.NotNull;
+import org.jglrxavpok.hephaistos.nbt.NBT;
+import org.jglrxavpok.hephaistos.nbt.NBTCompound;
+
+import java.util.Map;
+
+public record DefaultOption(@NotNull Particle particle) implements BiomeOption {
+    @Override
+    public @NotNull NBTCompound toNbt() {
+        return NBT.Compound(Map.of("type", NBT.String(getType().namespace().toString())));
+    }
+
+    @Override
+    public @NotNull Particle getType() {
+        return particle;
+    }
+}
diff --git a/src/main/java/net/minestom/server/world/biomes/particle/DustOption.java b/src/main/java/net/minestom/server/world/biomes/particle/DustOption.java
new file mode 100644
index 0000000000000000000000000000000000000000..0a84f495820afa685c08fd97bb454865ba976564
--- /dev/null
+++ b/src/main/java/net/minestom/server/world/biomes/particle/DustOption.java
@@ -0,0 +1,34 @@
+package net.minestom.server.world.biomes.particle;
+
+import net.minestom.server.particle.Particle;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.NotNull;
+import org.jglrxavpok.hephaistos.nbt.NBT;
+import org.jglrxavpok.hephaistos.nbt.NBTCompound;
+
+import java.util.Map;
+
+/**
+ * Particle effect which used based on the color values.
+ * @param red can between 0 and 1 and is divided by 255
+ * @param green can between 0 and 1 and is divided by 255
+ * @param blue can between 0 and 1 and is divided by 255
+ * @param scale can between 0.01 to 4
+ */
+@ApiStatus.Experimental
+public record DustOption(float red, float green, float blue, float scale) implements BiomeOption {
+    @Override
+    public @NotNull NBTCompound toNbt() {
+        return NBT.Compound(Map.of(
+                "type", NBT.String(getType().namespace().toString()),
+                "Red", NBT.Float(red),
+                "Green", NBT.Float(green),
+                "Blue", NBT.Float(blue),
+                "Scale", NBT.Float(scale)));
+    }
+
+    @Override
+    public @NotNull Particle getType() {
+        return Particle.DUST;
+    }
+}
diff --git a/src/main/java/net/minestom/server/world/biomes/particle/ItemOption.java b/src/main/java/net/minestom/server/world/biomes/particle/ItemOption.java
new file mode 100644
index 0000000000000000000000000000000000000000..a977cd052c10fd1b540b7934251d62e550361ca1
--- /dev/null
+++ b/src/main/java/net/minestom/server/world/biomes/particle/ItemOption.java
@@ -0,0 +1,27 @@
+package net.minestom.server.world.biomes.particle;
+
+import net.minestom.server.item.ItemStack;
+import net.minestom.server.particle.Particle;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.NotNull;
+import org.jglrxavpok.hephaistos.nbt.NBTCompound;
+
+/**
+ * Particle effect which used based on an item inside the biome.
+ * @param item to bes used
+ */
+@ApiStatus.Experimental
+public record ItemOption(@NotNull ItemStack item) implements BiomeOption {
+
+    @Override
+    public @NotNull NBTCompound toNbt() {
+        NBTCompound nbtCompound = item.meta().toNBT();
+        return nbtCompound.modify(n -> n.setString("type", getType().namespace().toString()));
+    }
+
+    @Override
+    public @NotNull Particle getType() {
+        return Particle.ITEM;
+    }
+
+}
diff --git a/src/test/java/net/minestom/server/instance/AnvilLoaderIntegrationTest.java b/src/test/java/net/minestom/server/instance/AnvilLoaderIntegrationTest.java
index 7fcb6ffa93f15baf5dc1d77f21a01068461f8115..a48b14e279ba57203e86315043aa6d913c1a85f5 100644
--- a/src/test/java/net/minestom/server/instance/AnvilLoaderIntegrationTest.java
+++ b/src/test/java/net/minestom/server/instance/AnvilLoaderIntegrationTest.java
@@ -1,9 +1,11 @@
 package net.minestom.server.instance;
 
+import net.minestom.server.MinecraftServer;
 import net.minestom.server.instance.block.Block;
 import net.minestom.server.network.NetworkBuffer;
 import net.minestom.server.utils.NamespaceID;
 import net.minestom.server.world.biomes.Biome;
+import net.minestom.server.world.biomes.Biomes;
 import net.minestom.testing.Env;
 import net.minestom.testing.EnvTest;
 import org.junit.jupiter.api.AfterAll;
@@ -49,6 +51,7 @@ class AnvilLoaderIntegrationTest {
 
     @Test
     void loadHouse(Env env) {
+        MinecraftServer.getBiomeManager().addBiome(Biomes.PLAINS.biome());
         // load a world that contains only a basic house and make sure it is loaded properly
 
         AnvilLoader chunkLoader = new AnvilLoader(worldFolder) {
@@ -146,6 +149,7 @@ class AnvilLoaderIntegrationTest {
 
     @Test
     void loadAndSaveChunk(Env env) throws InterruptedException {
+        MinecraftServer.getBiomeManager().addBiome(Biomes.PLAINS.biome());
         Instance instance = env.createFlatInstance(new AnvilLoader(worldFolder) {
             // Force loads inside current thread
             @Override
diff --git a/src/test/java/net/minestom/server/instance/GeneratorForkIntegrationTest.java b/src/test/java/net/minestom/server/instance/GeneratorForkIntegrationTest.java
index 3580264ac43c7268834e53315d968afe480e707c..0a00ce3ccc7fc54e4321c4c5a202328e950f0635 100644
--- a/src/test/java/net/minestom/server/instance/GeneratorForkIntegrationTest.java
+++ b/src/test/java/net/minestom/server/instance/GeneratorForkIntegrationTest.java
@@ -1,5 +1,6 @@
 package net.minestom.server.instance;
 
+import net.minestom.server.world.biomes.Biomes;
 import net.minestom.testing.Env;
 import net.minestom.testing.EnvTest;
 import net.minestom.server.coordinate.Point;
@@ -100,8 +101,8 @@ class GeneratorForkIntegrationTest {
         var instance = manager.createInstanceContainer();
         instance.setGenerator(unit -> {
             var u = unit.fork(unit.absoluteStart(), unit.absoluteEnd().add(16, 0, 16));
-            assertThrows(IllegalStateException.class, () -> u.modifier().setBiome(16, 0, 0, Biome.PLAINS));
-            assertThrows(IllegalStateException.class, () -> u.modifier().fillBiome(Biome.PLAINS));
+            assertThrows(IllegalStateException.class, () -> u.modifier().setBiome(16, 0, 0, Biomes.PLAINS.biome()));
+            assertThrows(IllegalStateException.class, () -> u.modifier().fillBiome(Biomes.PLAINS.biome()));
         });
         instance.loadChunk(0, 0).join();
     }
diff --git a/src/test/java/net/minestom/server/instance/GeneratorTest.java b/src/test/java/net/minestom/server/instance/GeneratorTest.java
index ad3ba534f6a91b90a713518aba46ac260fbcdd72..6ef0b29c6e535c856d8911e33315f2344d55cc73 100644
--- a/src/test/java/net/minestom/server/instance/GeneratorTest.java
+++ b/src/test/java/net/minestom/server/instance/GeneratorTest.java
@@ -8,6 +8,7 @@ import net.minestom.server.instance.generator.Generator;
 import net.minestom.server.utils.MathUtils;
 import net.minestom.server.utils.chunk.ChunkUtils;
 import net.minestom.server.world.biomes.Biome;
+import net.minestom.server.world.biomes.Biomes;
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.params.ParameterizedTest;
 import org.junit.jupiter.params.provider.Arguments;
@@ -253,13 +254,13 @@ class GeneratorTest {
         var chunkUnits = GeneratorImpl.chunk(minSection, maxSection, List.of(sections), chunkX, chunkZ);
         Generator generator = chunk -> {
             var modifier = chunk.modifier();
-            modifier.setBiome(48, 0, -32, Biome.PLAINS);
-            modifier.setBiome(48 + 8, 0, -32, Biome.PLAINS);
+            modifier.setBiome(48, 0, -32, Biomes.PLAINS.biome());
+            modifier.setBiome(48 + 8, 0, -32, Biomes.PLAINS.biome());
         };
         generator.generate(chunkUnits);
-        assertEquals(Biome.PLAINS.id(), sections[0].biomePalette().get(0, 0, 0));
+        assertEquals(Biomes.PLAINS.biome().id(), sections[0].biomePalette().get(0, 0, 0));
         assertEquals(0, sections[0].biomePalette().get(1, 0, 0));
-        assertEquals(Biome.PLAINS.id(), sections[0].biomePalette().get(2, 0, 0));
+        assertEquals(Biomes.PLAINS.biome().id(), sections[0].biomePalette().get(2, 0, 0));
     }
 
     @Test
@@ -274,12 +275,12 @@ class GeneratorTest {
         var chunkUnits = GeneratorImpl.chunk(minSection, maxSection, List.of(sections), chunkX, chunkZ);
         Generator generator = chunk -> {
             var modifier = chunk.modifier();
-            modifier.fillBiome(Biome.PLAINS);
+            modifier.fillBiome(Biomes.PLAINS.biome());
         };
         generator.generate(chunkUnits);
         for (var section : sections) {
             section.biomePalette().getAll((x, y, z, value) ->
-                    assertEquals(Biome.PLAINS.id(), value));
+                    assertEquals(Biomes.PLAINS.biome().id(), value));
         }
     }
 
