From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Phillipp Glanz <6745190+TheMeinerLP@users.noreply.github.com>
Date: Wed, 10 Jan 2024 21:48:06 +0100
Subject: [PATCH] Fix switching of entity metadata


diff --git a/src/main/java/net/minestom/server/entity/Entity.java b/src/main/java/net/minestom/server/entity/Entity.java
index 3241e9577c38e301096cebf96cb72816d52bdcff..ded65f2fe7b70f461bd091d34f53c3cfcdebb506 100644
--- a/src/main/java/net/minestom/server/entity/Entity.java
+++ b/src/main/java/net/minestom/server/entity/Entity.java
@@ -1232,7 +1232,7 @@ public class Entity implements Viewable, Tickable, Schedulable, Snapshotable, Ev
             setPose(Pose.FALL_FLYING);
         } else if (entityMeta.isSwimming()) {
             setPose(Pose.SWIMMING);
-        } else if (this instanceof LivingEntity && ((LivingEntityMeta) entityMeta).isInRiptideSpinAttack()) {
+        } else if (this instanceof LivingEntity && entityMeta instanceof LivingEntityMeta && ((LivingEntityMeta) entityMeta).isInRiptideSpinAttack()) {
             setPose(Pose.SPIN_ATTACK);
         } else if (entityMeta.isSneaking()) {
             setPose(Pose.SNEAKING);
diff --git a/src/main/java/net/minestom/server/entity/Player.java b/src/main/java/net/minestom/server/entity/Player.java
index cf72f8524ad5b549d3354b3dbaeab2d430f8161d..9400205088980c4f3c04a23a4039dab4e10ac55a 100644
--- a/src/main/java/net/minestom/server/entity/Player.java
+++ b/src/main/java/net/minestom/server/entity/Player.java
@@ -27,6 +27,7 @@ import net.minestom.server.coordinate.Vec;
 import net.minestom.server.effects.Effects;
 import net.minestom.server.entity.damage.DamageType;
 import net.minestom.server.entity.fakeplayer.FakePlayer;
+import net.minestom.server.entity.metadata.EntityMeta;
 import net.minestom.server.entity.metadata.PlayerMeta;
 import net.minestom.server.entity.vehicle.PlayerVehicleInformation;
 import net.minestom.server.event.EventDispatcher;
@@ -843,8 +844,7 @@ public class Player extends LivingEntity implements CommandSender, Localizable,
         sendPacket(new UpdateHealthPacket(health, food, foodSaturation));
     }
 
-    @Override
-    public @NotNull PlayerMeta getEntityMeta() {
+    public @NotNull PlayerMeta getUnsafeEntityMeta() {
         return (PlayerMeta) super.getEntityMeta();
     }
 
@@ -854,7 +854,7 @@ public class Player extends LivingEntity implements CommandSender, Localizable,
      * @return the player additional hearts
      */
     public float getAdditionalHearts() {
-        return getEntityMeta().getAdditionalHearts();
+        return getUnsafeEntityMeta().getAdditionalHearts();
     }
 
     /**
@@ -863,7 +863,7 @@ public class Player extends LivingEntity implements CommandSender, Localizable,
      * @param additionalHearts the count of additional hearts
      */
     public void setAdditionalHearts(float additionalHearts) {
-        getEntityMeta().setAdditionalHearts(additionalHearts);
+        getUnsafeEntityMeta().setAdditionalHearts(additionalHearts);
     }
 
     /**
diff --git a/src/main/java/net/minestom/server/listener/PlayerDiggingListener.java b/src/main/java/net/minestom/server/listener/PlayerDiggingListener.java
index fbdfe21c2cec8a5085aed299c6dc91fa04840f7c..d5bf810e291f3f2ac48fe1463f8f0bf8a4a64674 100644
--- a/src/main/java/net/minestom/server/listener/PlayerDiggingListener.java
+++ b/src/main/java/net/minestom/server/listener/PlayerDiggingListener.java
@@ -134,7 +134,7 @@ public final class PlayerDiggingListener {
     }
 
     private static void updateItemState(Player player) {
-        PlayerMeta meta = player.getEntityMeta();
+        PlayerMeta meta = player.getUnsafeEntityMeta();
         if (!meta.isHandActive()) return;
         Player.Hand hand = meta.getActiveHand();
 
