From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: theEvilReaper <steffenwx@gmail.com>
Date: Mon, 8 Jan 2024 11:01:34 +0100
Subject: [PATCH] Rethink generator structure and generation


diff --git a/code-generators/src/main/java/net/minestom/codegen/CodeExporter.java b/code-generators/src/main/java/net/minestom/codegen/CodeExporter.java
new file mode 100644
index 0000000000000000000000000000000000000000..8fe779599d52fa0cf15391738098ec8186c64625
--- /dev/null
+++ b/code-generators/src/main/java/net/minestom/codegen/CodeExporter.java
@@ -0,0 +1,46 @@
+package net.minestom.codegen;
+
+import com.squareup.javapoet.JavaFile;
+import org.jetbrains.annotations.NotNull;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.List;
+
+/**
+ * The class contains method which allows to write {@link JavaFile} to a given folder which is given as {@link java.nio.file.Path} reference.
+ * This class exists to reduce the amount of code duplication in the code generators.
+ * @version 1.0.0
+ * @since 1.1.4
+ */
+public interface CodeExporter {
+
+    Logger LOGGER = LoggerFactory.getLogger(CodeExporter.class);
+
+    /**
+     * Write a list of {@link JavaFile} objects to a given folder.
+     * @param fileList the list of files to write
+     * @param outputFolder the folder to write to
+     */
+    default void writeFiles(@NotNull List<JavaFile> fileList, File outputFolder) {
+        if (fileList.isEmpty()) return;
+        for (JavaFile javaFile : fileList) {
+            writeFile(javaFile, outputFolder);
+        }
+    }
+
+    /**
+     * Write a single {@link JavaFile} to a given folder.
+     * @param javaFile the file to write
+     * @param outputFolder the folder to write to
+     */
+    default void writeFile(@NotNull JavaFile javaFile, @NotNull File outputFolder) {
+        try {
+            javaFile.writeTo(outputFolder);
+        } catch (IOException e) {
+            LOGGER.error("An error occurred while writing source code to the file system.", e);
+        }
+    }
+}
diff --git a/code-generators/src/main/java/net/minestom/codegen/CodeGenerator.java b/code-generators/src/main/java/net/minestom/codegen/CodeGenerator.java
index ae427644f16ed3c6d99d07ba071cb0c1ebe776d4..e1d2d6277a74c731e3366a77ef9d95b2434836d2 100644
--- a/code-generators/src/main/java/net/minestom/codegen/CodeGenerator.java
+++ b/code-generators/src/main/java/net/minestom/codegen/CodeGenerator.java
@@ -3,39 +3,45 @@ package net.minestom.codegen;
 import com.google.gson.Gson;
 import com.google.gson.GsonBuilder;
 import com.google.gson.JsonObject;
-import com.squareup.javapoet.*;
+import com.squareup.javapoet.AnnotationSpec;
+import com.squareup.javapoet.ClassName;
+import com.squareup.javapoet.FieldSpec;
+import com.squareup.javapoet.JavaFile;
+import com.squareup.javapoet.TypeSpec;
+import org.jetbrains.annotations.ApiStatus;
 import org.jetbrains.annotations.NotNull;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 import javax.lang.model.element.Modifier;
 import java.io.File;
-import java.io.IOException;
 import java.io.InputStream;
 import java.io.InputStreamReader;
-import java.util.List;
-import java.util.Locale;
+import java.util.HashMap;
+import java.util.Map;
 
-public class CodeGenerator {
+import static net.minestom.codegen.MinestomCodeGenerator.DEFAULT_INDENT;
+import static net.minestom.codegen.MinestomCodeGenerator.extractNamespaces;
+
+@ApiStatus.Internal
+public class CodeGenerator implements CodeExporter {
     protected static final Gson GSON = new GsonBuilder().setPrettyPrinting().disableHtmlEscaping().create();
     private static final Logger LOGGER = LoggerFactory.getLogger(CodeGenerator.class);
-
     private final File outputFolder;
 
-    public CodeGenerator(File outputFolder) {
+    public CodeGenerator(@NotNull File outputFolder) {
         this.outputFolder = outputFolder;
     }
 
     public void generate(InputStream resourceFile, String packageName, String typeName, String loaderName, String generatedName) {
         if (resourceFile == null) {
-            LOGGER.error("Failed to find resource file for " + typeName);
+            LOGGER.error("Failed to find resource file for {}", typeName);
             return;
         }
         ClassName typeClass = ClassName.get(packageName, typeName);
         ClassName loaderClass = ClassName.get(packageName, loaderName);
 
-        JsonObject json;
-        json = GSON.fromJson(new InputStreamReader(resourceFile), JsonObject.class);
+        JsonObject json = GSON.fromJson(new InputStreamReader(resourceFile), JsonObject.class);
         ClassName materialsCN = ClassName.get(packageName, generatedName);
         // BlockConstants class
         TypeSpec.Builder blockConstantsClass = TypeSpec.interfaceBuilder(materialsCN)
@@ -43,17 +49,16 @@ public class CodeGenerator {
                 .addAnnotation(AnnotationSpec.builder(SuppressWarnings.class).addMember("value", "$S", "unused").build())
                 .addJavadoc("Code autogenerated, do not edit!");
 
+        Map<String, String> replacementOptions = new HashMap<>();
+        replacementOptions.put("minecraft:", "");
+        replacementOptions.put(".", "_");
         // Use data
         json.keySet().forEach(namespace -> {
-            final String constantName = namespace
-                    .replace("minecraft:", "")
-                    .replace(".", "_")
-                    .toUpperCase(Locale.ROOT);
+            final String constantName = extractNamespaces(namespace, replacementOptions);
             blockConstantsClass.addField(
                     FieldSpec.builder(typeClass, constantName)
                             .addModifiers(Modifier.PUBLIC, Modifier.STATIC, Modifier.FINAL)
                             .initializer(
-                                    // TypeClass.STONE = MaterialLoader.fromNamespaceId("minecraft:stone")
                                     "$T.get($S)",
                                     loaderClass,
                                     namespace
@@ -61,21 +66,12 @@ public class CodeGenerator {
                             .build()
             );
         });
-        writeFiles(
-                List.of(JavaFile.builder(packageName, blockConstantsClass.build())
-                        .indent("    ")
+        writeFile(
+                JavaFile.builder(packageName, blockConstantsClass.build())
+                        .indent(DEFAULT_INDENT)
                         .skipJavaLangImports(true)
-                        .build()),
-                outputFolder);
-    }
-
-    private void writeFiles(@NotNull List<JavaFile> fileList, File outputFolder) {
-        for (JavaFile javaFile : fileList) {
-            try {
-                javaFile.writeTo(outputFolder);
-            } catch (IOException e) {
-                LOGGER.error("An error occured while writing source code to the file system.", e);
-            }
-        }
+                        .build(),
+                outputFolder
+        );
     }
 }
diff --git a/code-generators/src/main/java/net/minestom/codegen/Generators.java b/code-generators/src/main/java/net/minestom/codegen/Generators.java
index 07dfb812a5dc4e10944545d4e19e00a125490d31..607fec9a560670c728ef3208d3987901b28a5dbf 100644
--- a/code-generators/src/main/java/net/minestom/codegen/Generators.java
+++ b/code-generators/src/main/java/net/minestom/codegen/Generators.java
@@ -44,7 +44,7 @@ public class Generators {
 //        new AttributeGenerator(
 //                new File(inputFolder, targetVersion + "_attributes.json"),
 //                outputFolder
-//        ).generate();
+//        2).generate();
         // TODO: Generate villager professions
 //        new VillagerProfessionGenerator(
 //                new File(inputFolder, targetVersion + "_villager_professions.json"),
diff --git a/code-generators/src/main/java/net/minestom/codegen/MinestomCodeGenerator.java b/code-generators/src/main/java/net/minestom/codegen/MinestomCodeGenerator.java
index 69eb7e67dd5253bca7e7948335a8f2f2acd849f1..9085f2fba74caacf6709b7d630e99ee72082787b 100644
--- a/code-generators/src/main/java/net/minestom/codegen/MinestomCodeGenerator.java
+++ b/code-generators/src/main/java/net/minestom/codegen/MinestomCodeGenerator.java
@@ -2,36 +2,68 @@ package net.minestom.codegen;
 
 import com.google.gson.Gson;
 import com.google.gson.GsonBuilder;
-import com.squareup.javapoet.JavaFile;
-import java.util.regex.Pattern;
+import com.squareup.javapoet.ClassName;
+import org.jetbrains.annotations.ApiStatus;
 import org.jetbrains.annotations.NotNull;
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
 
-import java.io.File;
-import java.io.IOException;
-import java.util.List;
+import javax.lang.model.element.Modifier;
 import java.util.Locale;
+import java.util.Map;
 
-public abstract class MinestomCodeGenerator {
-    private static final Logger LOGGER = LoggerFactory.getLogger(MinestomCodeGenerator.class);
+@ApiStatus.Internal
+public abstract class MinestomCodeGenerator implements CodeExporter {
+    public static final ClassName NAMESPACE_ID_CLASS =
+            ClassName.get("net.minestom.server.utils", "NamespaceID");
     protected static final Gson GSON = new GsonBuilder().setPrettyPrinting().disableHtmlEscaping().create();
-    protected static final Pattern minecraftNamespaceRegex = Pattern.compile("minecraft:");
+    protected static final ClassName REGISTRY_CLASS =
+            ClassName.get("net.minestom.server.registry", "Registries");
+    protected static final ClassName KEYORI_ADVENTURE_KEY =
+            ClassName.get("net.kyori.adventure.key", "Keyed");
+    protected static final Modifier[] CONSTANT_MODIFIERS = {Modifier.PRIVATE, Modifier.STATIC, Modifier.FINAL};
+    protected static final Modifier[] PRIVATE_FINAL_MODIFIERS = {Modifier.PRIVATE, Modifier.FINAL};
+    protected static final String DEFAULT_INDENT = "    ";
+    protected String packageName;
+
+    /**
+     * Creates a new code generator.
+     * @param packageName the package name of the generated class
+     */
+    protected MinestomCodeGenerator(@NotNull String packageName) {
+        if (packageName.trim().isEmpty()) {
+            throw new IllegalArgumentException("Package name cannot be empty");
+        }
+        this.packageName = packageName;
+    }
 
     public abstract void generate();
 
-    protected void writeFiles(@NotNull List<JavaFile> fileList, File outputFolder) {
-        for (JavaFile javaFile : fileList) {
-            try {
-                javaFile.writeTo(outputFolder);
-            } catch (IOException e) {
-                LOGGER.error("An error occured while writing source code to the file system.", e);
-            }
-        }
+    protected static @NotNull String extractNamespace(@NotNull String namespace) {
+        return namespace.replace("minecraft:", "").toUpperCase(Locale.ROOT);
     }
 
+    protected static @NotNull String extractNamespaces(@NotNull String namespace, @NotNull Map<String, String> arguments) {
+        if (arguments.isEmpty()) return extractNamespace(namespace);
 
-    protected static String toConstant(String namespace) {
-        return minecraftNamespaceRegex.matcher(namespace).replaceAll("").toUpperCase(Locale.ROOT);
+        for (Map.Entry<String, String> entry : arguments.entrySet()) {
+            namespace = namespace.replace(entry.getKey(), entry.getValue());
+        }
+        return namespace.toUpperCase(Locale.ROOT);
     }
+
+    // When the method above consumes too much memory, use this one instead
+    /*
+    protected static @NotNull String toConstant(@NotNull String namespace, @NotNull Map<String, String> arguments) {
+        if (arguments.isEmpty()) return toConstant(namespace);
+
+        StringBuilder builder = new StringBuilder(namespace);
+
+        for (Map.Entry<String, String> entry : arguments.entrySet()) {
+            int startIndex = builder.indexOf(entry.getKey());
+            while (startIndex != -1) {
+                builder.replace(startIndex, startIndex + entry.getKey().length(), entry.getValue());
+                startIndex = builder.indexOf(entry.getKey(), startIndex + entry.getValue().length());
+            }
+        }
+        return builder.toString().toUpperCase(Locale.ROOT);
+    }*/
 }
diff --git a/code-generators/src/main/java/net/minestom/codegen/attribute/AttributeGenerator.java b/code-generators/src/main/java/net/minestom/codegen/attribute/AttributeGenerator.java
index 11ee9eef838aeda0f5e0c4fb3d9a21940563d9f9..618040b6ff868f2327c641ed3e5c205c8a00344f 100644
--- a/code-generators/src/main/java/net/minestom/codegen/attribute/AttributeGenerator.java
+++ b/code-generators/src/main/java/net/minestom/codegen/attribute/AttributeGenerator.java
@@ -3,12 +3,22 @@ package net.minestom.codegen.attribute;
 import com.google.gson.JsonArray;
 import com.google.gson.JsonElement;
 import com.google.gson.JsonObject;
-import com.squareup.javapoet.*;
+import com.squareup.javapoet.CodeBlock;
+import com.squareup.javapoet.ClassName;
+import com.squareup.javapoet.FieldSpec;
+import com.squareup.javapoet.JavaFile;
+import com.squareup.javapoet.TypeSpec;
+import com.squareup.javapoet.TypeName;
+import com.squareup.javapoet.ParameterizedTypeName;
+import com.squareup.javapoet.MethodSpec;
+import com.squareup.javapoet.ParameterSpec;
 import net.minestom.codegen.MinestomCodeGenerator;
+import org.jetbrains.annotations.ApiStatus;
 import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
+import net.minestom.codegen.util.GenerationHelper;
 
 import javax.lang.model.element.Modifier;
 import java.io.File;
@@ -17,12 +27,22 @@ import java.io.InputStreamReader;
 import java.util.ArrayList;
 import java.util.List;
 
+import static net.minestom.codegen.util.GenerationHelper.VARIABLE_SETTER;
+
+@ApiStatus.NonExtendable
+@ApiStatus.Internal
 public final class AttributeGenerator extends MinestomCodeGenerator {
     private static final Logger LOGGER = LoggerFactory.getLogger(AttributeGenerator.class);
+    private static final String MIN_VALUE = "minValue";
+    private static final String MAX_VALUE = "maxValue";
+    private static final String DEFAULT_VALUE = "defaultValue";
+    private static final String CLIENT_SYNCABLE = "clientSyncable";
+    private static final String LITERAL_RETURN = "return this.$L";
     private final InputStream attributesFile;
     private final File outputFolder;
 
     public AttributeGenerator(@Nullable InputStream attributesFile, @NotNull File outputFolder) {
+        super("net.minestom.server.attribute");
         this.attributesFile = attributesFile;
         this.outputFolder = outputFolder;
     }
@@ -38,66 +58,47 @@ public final class AttributeGenerator extends MinestomCodeGenerator {
             LOGGER.error("Output folder for code generation does not exist and could not be created.");
             return;
         }
-        // Important classes we use alot
-        ClassName namespaceIDClassName = ClassName.get("net.minestom.server.utils", "NamespaceID");
-        ClassName registryClassName = ClassName.get("net.minestom.server.registry", "Registry");
 
         JsonArray attributes = GSON.fromJson(new InputStreamReader(attributesFile), JsonArray.class);
         List<JavaFile> filesToWrite = new ArrayList<>();
 
-        ClassName attributeClassName = ClassName.get("net.minestom.server.attribute", "Attribute");
-
+        ClassName attributeClassName = ClassName.get(packageName, "Attribute");
         // Attribute
         TypeSpec.Builder attributeClass = TypeSpec.classBuilder(attributeClassName)
-                .addSuperinterface(ClassName.get("net.kyori.adventure.key", "Keyed"))
+                .addSuperinterface(KEYORI_ADVENTURE_KEY)
                 .addModifiers(Modifier.PUBLIC).addJavadoc("AUTOGENERATED by " + getClass().getSimpleName());
         attributeClass.addField(
-                FieldSpec.builder(namespaceIDClassName, "id")
-                        .addModifiers(Modifier.PRIVATE, Modifier.FINAL).addAnnotation(NotNull.class).build()
+                FieldSpec.builder(NAMESPACE_ID_CLASS, "id")
+                        .addModifiers(PRIVATE_FINAL_MODIFIERS).addAnnotation(NotNull.class).build()
         );
         attributeClass.addField(
-                FieldSpec.builder(TypeName.DOUBLE, "defaultValue")
-                        .addModifiers(Modifier.PRIVATE, Modifier.FINAL).build()
+                FieldSpec.builder(TypeName.DOUBLE, DEFAULT_VALUE)
+                        .addModifiers(PRIVATE_FINAL_MODIFIERS).build()
         );
         attributeClass.addField(
-                FieldSpec.builder(TypeName.BOOLEAN, "clientSyncable")
-                        .addModifiers(Modifier.PRIVATE, Modifier.FINAL).build()
+                FieldSpec.builder(TypeName.BOOLEAN, CLIENT_SYNCABLE)
+                        .addModifiers(PRIVATE_FINAL_MODIFIERS).build()
         );
         attributeClass.addMethod(
                 MethodSpec.constructorBuilder()
-                        .addParameter(ParameterSpec.builder(namespaceIDClassName, "id").addAnnotation(NotNull.class).build())
-                        .addParameter(ParameterSpec.builder(TypeName.BOOLEAN, "clientSyncable").build())
-                        .addParameter(ParameterSpec.builder(TypeName.DOUBLE, "defaultValue").build())
-                        .addStatement("this.id = id")
-                        .addStatement("this.clientSyncable = clientSyncable")
-                        .addStatement("this.defaultValue = defaultValue")
+                        .addParameter(ParameterSpec.builder(NAMESPACE_ID_CLASS, "id").addAnnotation(NotNull.class).build())
+                        .addParameter(ParameterSpec.builder(TypeName.BOOLEAN, CLIENT_SYNCABLE).build())
+                        .addParameter(ParameterSpec.builder(TypeName.DOUBLE, DEFAULT_VALUE).build())
+                        .addStatement(VARIABLE_SETTER, "id")
+                        .addStatement(VARIABLE_SETTER, CLIENT_SYNCABLE)
+                        .addStatement(VARIABLE_SETTER, DEFAULT_VALUE)
                         .addModifiers(Modifier.PROTECTED)
                         .build()
         );
         // Override key method (adventure)
-        attributeClass.addMethod(
-                MethodSpec.methodBuilder("key")
-                        .returns(ClassName.get("net.kyori.adventure.key", "Key"))
-                        .addAnnotation(Override.class)
-                        .addAnnotation(NotNull.class)
-                        .addStatement("return this.id")
-                        .addModifiers(Modifier.PUBLIC)
-                        .build()
-        );
+        attributeClass.addMethod(GenerationHelper.ADVENTURE_KEY_METHOD);
         // getId method
-        attributeClass.addMethod(
-                MethodSpec.methodBuilder("getId")
-                        .returns(namespaceIDClassName)
-                        .addAnnotation(NotNull.class)
-                        .addStatement("return this.id")
-                        .addModifiers(Modifier.PUBLIC)
-                        .build()
-        );
+        attributeClass.addMethod(GenerationHelper.ID_GETTER);
         // getDefaultValue
         attributeClass.addMethod(
                 MethodSpec.methodBuilder("getDefaultValue")
                         .returns(TypeName.DOUBLE)
-                        .addStatement("return this.defaultValue")
+                        .addStatement("return this.%L", DEFAULT_VALUE)
                         .addModifiers(Modifier.PUBLIC)
                         .build()
         );
@@ -105,7 +106,7 @@ public final class AttributeGenerator extends MinestomCodeGenerator {
         attributeClass.addMethod(
                 MethodSpec.methodBuilder("isClientSyncable")
                         .returns(TypeName.BOOLEAN)
-                        .addStatement("return this.clientSyncable")
+                        .addStatement(LITERAL_RETURN, CLIENT_SYNCABLE)
                         .addModifiers(Modifier.PUBLIC)
                         .build()
         );
@@ -114,7 +115,7 @@ public final class AttributeGenerator extends MinestomCodeGenerator {
                 MethodSpec.methodBuilder("isShared")
                         .addAnnotation(Deprecated.class)
                         .returns(TypeName.BOOLEAN)
-                        .addStatement("return this.clientSyncable")
+                        .addStatement(LITERAL_RETURN, CLIENT_SYNCABLE)
                         .addModifiers(Modifier.PUBLIC)
                         .build()
         );
@@ -123,64 +124,15 @@ public final class AttributeGenerator extends MinestomCodeGenerator {
                 MethodSpec.methodBuilder("values")
                         .addAnnotation(NotNull.class)
                         .returns(ParameterizedTypeName.get(ClassName.get(List.class), attributeClassName))
-                        .addStatement("return $T.ATTRIBUTE_REGISTRY.values()", registryClassName)
+                        .addStatement("return $T.ATTRIBUTE_REGISTRY.values()", REGISTRY_CLASS)
                         .addModifiers(Modifier.PUBLIC, Modifier.STATIC)
                         .build()
         );
         // toString method
-        attributeClass.addMethod(
-                MethodSpec.methodBuilder("toString")
-                        .addAnnotation(NotNull.class)
-                        .addAnnotation(Override.class)
-                        .returns(String.class)
-                        // this resolves to [Namespace]
-                        .addStatement("return \"[\" + this.id + \"]\"")
-                        .addModifiers(Modifier.PUBLIC)
-                        .build()
-        );
-        // Creating ClampedAttribute
-        ClassName clampedAttributeClassName = ClassName.get("net.minestom.server.attribute", "ClampedAttribute");
+        attributeClass.addMethod(GenerationHelper.TO_STRING);
 
-        TypeSpec.Builder clampedAttributeClass = TypeSpec.classBuilder(clampedAttributeClassName)
-                .superclass(attributeClassName)
-                .addModifiers(Modifier.PUBLIC).addJavadoc("AUTOGENERATED by " + getClass().getSimpleName());
-        clampedAttributeClass.addField(
-                FieldSpec.builder(TypeName.DOUBLE, "minValue")
-                        .addModifiers(Modifier.PRIVATE, Modifier.FINAL).build()
-        );
-        clampedAttributeClass.addField(
-                FieldSpec.builder(TypeName.DOUBLE, "maxValue")
-                        .addModifiers(Modifier.PRIVATE, Modifier.FINAL).build()
-        );
-        clampedAttributeClass.addMethod(
-                MethodSpec.constructorBuilder()
-                        .addParameter(ParameterSpec.builder(namespaceIDClassName, "id").addAnnotation(NotNull.class).build())
-                        .addParameter(ParameterSpec.builder(TypeName.BOOLEAN, "clientSyncable").build())
-                        .addParameter(ParameterSpec.builder(TypeName.DOUBLE, "defaultValue").build())
-                        .addParameter(ParameterSpec.builder(TypeName.DOUBLE, "minValue").build())
-                        .addParameter(ParameterSpec.builder(TypeName.DOUBLE, "maxValue").build())
-                        .addStatement("super(id, clientSyncable, defaultValue)")
-                        .addStatement("this.minValue = minValue")
-                        .addStatement("this.maxValue = maxValue")
-                        .addModifiers(Modifier.PROTECTED)
-                        .build()
-        );
-        // getMinValue
-        clampedAttributeClass.addMethod(
-                MethodSpec.methodBuilder("getMinValue")
-                        .returns(TypeName.DOUBLE)
-                        .addStatement("return this.minValue")
-                        .addModifiers(Modifier.PUBLIC)
-                        .build()
-        );
-        // getMaxValue
-        clampedAttributeClass.addMethod(
-                MethodSpec.methodBuilder("getMaxValue")
-                        .returns(TypeName.DOUBLE)
-                        .addStatement("return this.maxValue")
-                        .addModifiers(Modifier.PUBLIC)
-                        .build()
-        );
+        // Creating ClampedAttribute
+        TypeSpec clampedAttributeClass = generateClampedAttributeClass(attributeClassName);
 
         CodeBlock.Builder staticBlock = CodeBlock.builder();
         // Use data
@@ -198,11 +150,11 @@ public final class AttributeGenerator extends MinestomCodeGenerator {
                         ).initializer(
                                 "new $T($T.from($S), $L, $L)",
                                 attributeClassName,
-                                namespaceIDClassName,
+                                NAMESPACE_ID_CLASS,
                                 attribute.get("id").getAsString(),
                                 attribute.get("clientSync").getAsBoolean(),
-                                attribute.get("defaultValue").getAsDouble()
-                        ).addModifiers(Modifier.PUBLIC, Modifier.STATIC, Modifier.FINAL).build()
+                                attribute.get(DEFAULT_VALUE).getAsDouble()
+                        ).addModifiers(CONSTANT_MODIFIERS).build()
                 );
             } else {
                 // ClampedAttribute
@@ -212,30 +164,30 @@ public final class AttributeGenerator extends MinestomCodeGenerator {
                                 attributeName
                         ).initializer(
                                 "new $T($T.from($S), $L, $L, $L, $L)",
-                                clampedAttributeClassName,
-                                namespaceIDClassName,
+                                clampedAttributeClass,
+                                NAMESPACE_ID_CLASS,
                                 attribute.get("id").getAsString(),
                                 attribute.get("clientSync").getAsBoolean(),
-                                attribute.get("defaultValue").getAsDouble(),
-                                range.get("minValue").getAsDouble(),
-                                range.get("maxValue").getAsDouble()
-                        ).addModifiers(Modifier.PUBLIC, Modifier.STATIC, Modifier.FINAL).build()
+                                attribute.get(DEFAULT_VALUE).getAsDouble(),
+                                range.get(MIN_VALUE).getAsDouble(),
+                                range.get(MAX_VALUE).getAsDouble()
+                        ).addModifiers(CONSTANT_MODIFIERS).build()
                 );
             }
             // Add to static init.
-            staticBlock.addStatement("$T.ATTRIBUTE_REGISTRY.register($N)", registryClassName, attributeName);
+            staticBlock.addStatement("$T.ATTRIBUTE_REGISTRY.register($N)", REGISTRY_CLASS, attributeName);
         }
         attributeClass.addStaticBlock(staticBlock.build());
 
         filesToWrite.add(
-                JavaFile.builder("net.minestom.server.attribute", attributeClass.build())
-                        .indent("    ")
+                JavaFile.builder(packageName, attributeClass.build())
+                        .indent(DEFAULT_INDENT)
                         .skipJavaLangImports(true)
                         .build()
         );
         filesToWrite.add(
-                JavaFile.builder("net.minestom.server.attribute", clampedAttributeClass.build())
-                        .indent("    ")
+                JavaFile.builder(packageName, clampedAttributeClass)
+                        .indent(DEFAULT_VALUE)
                         .skipJavaLangImports(true)
                         .build()
         );
@@ -246,4 +198,49 @@ public final class AttributeGenerator extends MinestomCodeGenerator {
                 outputFolder
         );
     }
+
+    private @NotNull TypeSpec generateClampedAttributeClass(@NotNull ClassName attributeClassName) {
+        ClassName clampedAttributeClassName = ClassName.get(packageName, "ClampedAttribute");
+        TypeSpec.Builder clampedAttributeClass = TypeSpec.classBuilder(clampedAttributeClassName)
+                .superclass(attributeClassName)
+                .addModifiers(Modifier.PUBLIC).addJavadoc("AUTOGENERATED by " + getClass().getSimpleName());
+        clampedAttributeClass.addField(
+                FieldSpec.builder(TypeName.DOUBLE, MIN_VALUE)
+                        .addModifiers(Modifier.PRIVATE, Modifier.FINAL).build()
+        );
+        clampedAttributeClass.addField(
+                FieldSpec.builder(TypeName.DOUBLE, MAX_VALUE)
+                        .addModifiers(Modifier.PRIVATE, Modifier.FINAL).build()
+        );
+        clampedAttributeClass.addMethod(
+                MethodSpec.constructorBuilder()
+                        .addParameter(ParameterSpec.builder(NAMESPACE_ID_CLASS, "id").addAnnotation(NotNull.class).build())
+                        .addParameter(ParameterSpec.builder(TypeName.BOOLEAN, CLIENT_SYNCABLE).build())
+                        .addParameter(ParameterSpec.builder(TypeName.DOUBLE, DEFAULT_VALUE).build())
+                        .addParameter(ParameterSpec.builder(TypeName.DOUBLE, MIN_VALUE).build())
+                        .addParameter(ParameterSpec.builder(TypeName.DOUBLE, MAX_VALUE).build())
+                        .addStatement("super(id, $L, $L)", CLIENT_SYNCABLE, DEFAULT_VALUE)
+                        .addStatement(VARIABLE_SETTER, MIN_VALUE)
+                        .addStatement(VARIABLE_SETTER, MAX_VALUE)
+                        .addModifiers(Modifier.PROTECTED)
+                        .build()
+        );
+        // getMinValue
+        clampedAttributeClass.addMethod(
+                MethodSpec.methodBuilder("getMinValue")
+                        .returns(TypeName.DOUBLE)
+                        .addStatement(LITERAL_RETURN, MIN_VALUE)
+                        .addModifiers(Modifier.PUBLIC)
+                        .build()
+        );
+        // getMaxValue
+        clampedAttributeClass.addMethod(
+                MethodSpec.methodBuilder("getMaxValue")
+                        .returns(TypeName.DOUBLE)
+                        .addStatement(LITERAL_RETURN, MAX_VALUE)
+                        .addModifiers(Modifier.PUBLIC)
+                        .build()
+        );
+        return clampedAttributeClass.build();
+    }
 }
diff --git a/code-generators/src/main/java/net/minestom/codegen/biome/BiomeGenerator.java b/code-generators/src/main/java/net/minestom/codegen/biome/BiomeGenerator.java
index 13c12efedb65bc5880cd8c4d186427d36d847df8..b0095d0cd6d906bf6a68a586e4eb4c61dae6a3dd 100644
--- a/code-generators/src/main/java/net/minestom/codegen/biome/BiomeGenerator.java
+++ b/code-generators/src/main/java/net/minestom/codegen/biome/BiomeGenerator.java
@@ -18,11 +18,14 @@ import java.io.InputStreamReader;
 import java.util.List;
 import javax.lang.model.element.Modifier;
 import net.minestom.codegen.MinestomCodeGenerator;
+import org.jetbrains.annotations.ApiStatus;
 import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+@ApiStatus.Internal
+@ApiStatus.NonExtendable
 public final class BiomeGenerator extends MinestomCodeGenerator {
 
     private static final String BIOME_KEY = "biome";
@@ -33,6 +36,7 @@ public final class BiomeGenerator extends MinestomCodeGenerator {
     private final File outputFolder;
 
     public BiomeGenerator(@Nullable InputStream biomeFile, @NotNull File outputFolder) {
+        super(BIOME_FILE_PACKAGE);
         this.biomeFile = biomeFile;
         this.outputFolder = outputFolder;
     }
@@ -120,10 +124,10 @@ public final class BiomeGenerator extends MinestomCodeGenerator {
                 effectCodeBlock.add(".foliageColor(0x$L)", foliageColor);
             }
             effectCodeBlock.add(".build()");
-            var namespacedCode = CodeBlock.of("$T.from($S)",namespaceIDCn, biome.replaceFirst("minecraft:", ""));
-            biomesEnum.addEnumConstant(toConstant(biome), TypeSpec.anonymousClassBuilder(
+            var namespacedCode = CodeBlock.of("$T.from($S)", namespaceIDCn, biome.replaceFirst("minecraft:", ""));
+            biomesEnum.addEnumConstant(extractNamespace(biome), TypeSpec.anonymousClassBuilder(
                             "$T.builder().name($L).category($T.NONE).depth(0.125F).scale(0.05F).temperature($Lf).downfall($Lf).showPrecipitation($L).effects($L).build()",
-                            biomeCn, namespacedCode,categoryCn, temperature, downfall, hasPrecipitation, effectCodeBlock.build()
+                            biomeCn, namespacedCode, categoryCn, temperature, downfall, hasPrecipitation, effectCodeBlock.build()
                     ).build()
             );
         });
@@ -132,7 +136,7 @@ public final class BiomeGenerator extends MinestomCodeGenerator {
         writeFiles(
                 List.of(
                         JavaFile.builder(BIOME_FILE_PACKAGE, biomesEnum.build())
-                                .indent("    ")
+                                .indent(DEFAULT_INDENT)
                                 .skipJavaLangImports(true)
                                 .build()
                 ),
diff --git a/code-generators/src/main/java/net/minestom/codegen/color/DyeColorGenerator.java b/code-generators/src/main/java/net/minestom/codegen/color/DyeColorGenerator.java
index cbb759818ed4a7d61769bafaf4bc3fb24776ecbc..c27b85e155bd974c449c38adb062cff93ed46e9f 100644
--- a/code-generators/src/main/java/net/minestom/codegen/color/DyeColorGenerator.java
+++ b/code-generators/src/main/java/net/minestom/codegen/color/DyeColorGenerator.java
@@ -5,6 +5,8 @@ import com.google.gson.JsonElement;
 import com.google.gson.JsonObject;
 import com.squareup.javapoet.*;
 import net.minestom.codegen.MinestomCodeGenerator;
+import net.minestom.codegen.util.GenerationHelper;
+import org.jetbrains.annotations.ApiStatus;
 import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
 import org.slf4j.Logger;
@@ -18,15 +20,21 @@ import java.util.Comparator;
 import java.util.List;
 import java.util.stream.StreamSupport;
 
-public class DyeColorGenerator extends MinestomCodeGenerator {
+@ApiStatus.NonExtendable
+@ApiStatus.Internal
+public final class DyeColorGenerator extends MinestomCodeGenerator {
 
-    private static final String COLOR_FILE_PACKAGE = "net.minestom.server.color"; // Microtus - Banner and shield meta
     private static final String CLASS_NAME = "DyeColor"; // Microtus - Banner and shield meta
+    private static final String TEXTURE_DIFFUSE_COLOR = "textureDiffuseColor";
+    private static final String TEXT_COLOR = "textColor";
+    private static final String FIREWORK_COLOR = "fireworkColor";
+    private static final String MAP_COLOR_ID = "mapColorId";
     private static final Logger LOGGER = LoggerFactory.getLogger(DyeColorGenerator.class);
     private final InputStream dyeColorsFile;
     private final File outputFolder;
 
     public DyeColorGenerator(@Nullable InputStream dyeColorsFile, @NotNull File outputFolder) {
+        super("net.minestom.server.color");
         this.dyeColorsFile = dyeColorsFile;
         this.outputFolder = outputFolder;
     }
@@ -43,10 +51,10 @@ public class DyeColorGenerator extends MinestomCodeGenerator {
             return;
         }
         // Important classes we use a lot
-        ClassName colorCN = ClassName.get(COLOR_FILE_PACKAGE, "Color");
+        ClassName colorCN = ClassName.get(packageName, "Color");
 
         JsonArray dyeColors = GSON.fromJson(new InputStreamReader(dyeColorsFile), JsonArray.class);
-        ClassName dyeColorCN = ClassName.get(COLOR_FILE_PACKAGE, CLASS_NAME);
+        ClassName dyeColorCN = ClassName.get(packageName, CLASS_NAME);
         // Dye Color Enum
         TypeSpec.Builder dyeColorEnum = TypeSpec.enumBuilder(dyeColorCN)
                 .addSuperinterface(ClassName.get("net.kyori.adventure.util", "RGBLike"))
@@ -55,11 +63,11 @@ public class DyeColorGenerator extends MinestomCodeGenerator {
         // Fields
         dyeColorEnum.addFields(
                 List.of(
-                        FieldSpec.builder(colorCN, "textureDiffuseColor", Modifier.PRIVATE, Modifier.FINAL).build(),
-                        FieldSpec.builder(colorCN, "textColor", Modifier.PRIVATE, Modifier.FINAL).build(),
-                        FieldSpec.builder(colorCN, "fireworkColor", Modifier.PRIVATE, Modifier.FINAL).build(),
-                        FieldSpec.builder(TypeName.INT, "mapColorId", Modifier.PRIVATE, Modifier.FINAL).build(),
-                        FieldSpec.builder(ArrayTypeName.of(dyeColorCN), "VALUES", Modifier.PRIVATE, Modifier.STATIC, Modifier.FINAL).initializer(CLASS_NAME + ".values()").build()  // Microtus - Banner and shield meta
+                        FieldSpec.builder(colorCN, TEXTURE_DIFFUSE_COLOR, PRIVATE_FINAL_MODIFIERS).build(),
+                        FieldSpec.builder(colorCN, TEXT_COLOR, PRIVATE_FINAL_MODIFIERS).build(),
+                        FieldSpec.builder(colorCN, FIREWORK_COLOR, PRIVATE_FINAL_MODIFIERS).build(),
+                        FieldSpec.builder(TypeName.INT, MAP_COLOR_ID, PRIVATE_FINAL_MODIFIERS).build(),
+                        FieldSpec.builder(ArrayTypeName.of(dyeColorCN), "VALUES", CONSTANT_MODIFIERS).initializer(CLASS_NAME + ".values()").build()  // Microtus - Banner and shield meta
                 )
         );
 
@@ -70,28 +78,28 @@ public class DyeColorGenerator extends MinestomCodeGenerator {
                         MethodSpec.constructorBuilder()
                                 .addParameters(
                                         List.of(
-                                                ParameterSpec.builder(colorCN, "textureDiffuseColor").addAnnotation(NotNull.class).build(),
-                                                ParameterSpec.builder(colorCN, "textColor").addAnnotation(NotNull.class).build(),
-                                                ParameterSpec.builder(colorCN, "fireworkColor").addAnnotation(NotNull.class).build(),
-                                                ParameterSpec.builder(TypeName.INT, "mapColorId").build()
+                                                ParameterSpec.builder(colorCN, TEXTURE_DIFFUSE_COLOR).addAnnotation(NotNull.class).build(),
+                                                ParameterSpec.builder(colorCN, TEXT_COLOR).addAnnotation(NotNull.class).build(),
+                                                ParameterSpec.builder(colorCN, FIREWORK_COLOR).addAnnotation(NotNull.class).build(),
+                                                ParameterSpec.builder(TypeName.INT, MAP_COLOR_ID).build()
                                         )
                                 )
-                                .addStatement("this.textureDiffuseColor = textureDiffuseColor")
-                                .addStatement("this.textColor = textColor")
-                                .addStatement("this.fireworkColor = fireworkColor")
-                                .addStatement("this.mapColorId = mapColorId")
+                                .addStatement(GenerationHelper.VARIABLE_SETTER, TEXTURE_DIFFUSE_COLOR)
+                                .addStatement(GenerationHelper.VARIABLE_SETTER, TEXT_COLOR)
+                                .addStatement(GenerationHelper.VARIABLE_SETTER, FIREWORK_COLOR)
+                                .addStatement(GenerationHelper.VARIABLE_SETTER, MAP_COLOR_ID)
                                 .build(),
                         MethodSpec.methodBuilder("color")
                                 .addModifiers(Modifier.PUBLIC)
                                 .returns(colorCN.annotated(AnnotationSpec.builder(NotNull.class).build()))
                                 .addStatement("return this.textureDiffuseColor")
                                 .build(),
-                        MethodSpec.methodBuilder("textColor")
+                        MethodSpec.methodBuilder(TEXT_COLOR)
                                 .addModifiers(Modifier.PUBLIC)
                                 .returns(colorCN.annotated(AnnotationSpec.builder(NotNull.class).build()))
                                 .addStatement("return this.textColor")
                                 .build(),
-                        MethodSpec.methodBuilder("fireworkColor")
+                        MethodSpec.methodBuilder(FIREWORK_COLOR)
                                 .addModifiers(Modifier.PUBLIC)
                                 .returns(colorCN.annotated(AnnotationSpec.builder(NotNull.class).build()))
                                 .addStatement("return this.fireworkColor")
@@ -114,7 +122,7 @@ public class DyeColorGenerator extends MinestomCodeGenerator {
                                 .addAnnotation(Override.class)
                                 .addStatement("return this.textureDiffuseColor.blue()")
                                 .build(),
-                        MethodSpec.methodBuilder("mapColorId")
+                        MethodSpec.methodBuilder(MAP_COLOR_ID)
                                 .addModifiers(Modifier.PUBLIC)
                                 .returns(TypeName.INT)
                                 .addStatement("return this.mapColorId")
@@ -123,7 +131,7 @@ public class DyeColorGenerator extends MinestomCodeGenerator {
                                 .addModifiers(Modifier.PUBLIC, Modifier.STATIC)
                                 .addParameter(ParameterSpec.builder(TypeName.INT, "id").build())
                                 .addAnnotation(Nullable.class)
-                                .returns(ClassName.get(COLOR_FILE_PACKAGE, CLASS_NAME))
+                                .returns(ClassName.get(packageName, CLASS_NAME))
                                 .addCode("return VALUES[$L];", "id")
                                 .build() // Microtus end - Banner and shield meta
                 )
@@ -132,12 +140,12 @@ public class DyeColorGenerator extends MinestomCodeGenerator {
         // Use data
         for (JsonObject dyeColorObject : StreamSupport.stream(dyeColors.spliterator(), true).map(JsonElement::getAsJsonObject).sorted(Comparator.comparingInt(o -> o.get("id").getAsInt())).toList()) {
             String dyeColorName = dyeColorObject.get("name").getAsString();
-            dyeColorEnum.addEnumConstant(toConstant(dyeColorName), TypeSpec.anonymousClassBuilder(
+            dyeColorEnum.addEnumConstant(extractNamespace(dyeColorName), TypeSpec.anonymousClassBuilder(
                             "new $T(0x$L), new $T(0x$L), new $T(0x$L), $L",
-                            colorCN, Integer.toString(dyeColorObject.get("textureDiffuseColor").getAsInt(), 16),
-                            colorCN, Integer.toString(dyeColorObject.get("textColor").getAsInt(), 16),
-                            colorCN, Integer.toString(dyeColorObject.get("fireworkColor").getAsInt(), 16),
-                            dyeColorObject.get("mapColorId").getAsInt()
+                            colorCN, Integer.toString(dyeColorObject.get(TEXTURE_DIFFUSE_COLOR).getAsInt(), 16),
+                            colorCN, Integer.toString(dyeColorObject.get(TEXT_COLOR).getAsInt(), 16),
+                            colorCN, Integer.toString(dyeColorObject.get(FIREWORK_COLOR).getAsInt(), 16),
+                            dyeColorObject.get(MAP_COLOR_ID).getAsInt()
                     ).build()
             );
         }
@@ -145,8 +153,8 @@ public class DyeColorGenerator extends MinestomCodeGenerator {
         // Write files to outputFolder
         writeFiles(
                 List.of(
-                        JavaFile.builder(COLOR_FILE_PACKAGE, dyeColorEnum.build())
-                                .indent("    ")
+                        JavaFile.builder(packageName, dyeColorEnum.build())
+                                .indent(DEFAULT_INDENT)
                                 .skipJavaLangImports(true)
                                 .build()
                 ),
diff --git a/code-generators/src/main/java/net/minestom/codegen/entity/VillagerProfessionGenerator.java b/code-generators/src/main/java/net/minestom/codegen/entity/VillagerProfessionGenerator.java
index 54eaae59378b7ad634ff0168b6c6a788512476cf..d25441a83db0795e92b22c20adf4bac22cb8173d 100644
--- a/code-generators/src/main/java/net/minestom/codegen/entity/VillagerProfessionGenerator.java
+++ b/code-generators/src/main/java/net/minestom/codegen/entity/VillagerProfessionGenerator.java
@@ -5,10 +5,12 @@ import com.google.gson.JsonElement;
 import com.google.gson.JsonObject;
 import com.squareup.javapoet.*;
 import net.minestom.codegen.MinestomCodeGenerator;
+import org.jetbrains.annotations.ApiStatus;
 import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
+import net.minestom.codegen.util.GenerationHelper;
 
 import javax.lang.model.element.Modifier;
 import java.io.File;
@@ -16,13 +18,20 @@ import java.io.InputStream;
 import java.io.InputStreamReader;
 import java.util.List;
 
+import static net.minestom.codegen.util.GenerationHelper.*;
+
+@ApiStatus.NonExtendable
+@ApiStatus.Internal
 public final class VillagerProfessionGenerator extends MinestomCodeGenerator {
+
+    private static final String VILLAGER_PROFESSION_DATE = "villagerProfessionData";
     private static final Logger LOGGER = LoggerFactory.getLogger(VillagerProfessionGenerator.class);
     private final InputStream villagerProfessionsFile;
     private final File outputFolder;
 
 
     public VillagerProfessionGenerator(@Nullable InputStream villagerProfessionsFile, @NotNull File outputFolder) {
+        super("");
         this.villagerProfessionsFile = villagerProfessionsFile;
         this.outputFolder = outputFolder;
     }
@@ -39,55 +48,38 @@ public final class VillagerProfessionGenerator extends MinestomCodeGenerator {
             return;
         }
         // Important classes we use alot
-        ClassName namespaceIDClassName = ClassName.get("net.minestom.server.utils", "NamespaceID");
         ClassName rawVillagerProfessionDataClassName = ClassName.get("net.minestom.server.raw_data", "RawVillagerProfessionData");
-        ClassName registryClassName = ClassName.get("net.minestom.server.registry", "Registry");
 
         JsonArray villagerProfessions = GSON.fromJson(new InputStreamReader(villagerProfessionsFile), JsonArray.class);
         ClassName villagerProfessionClassName = ClassName.get("net.minestom.server.entity.metadata.villager", "VillagerProfession");
 
         // Particle
         TypeSpec.Builder villagerProfessionClass = TypeSpec.classBuilder(villagerProfessionClassName)
-                .addSuperinterface(ClassName.get("net.kyori.adventure.key", "Keyed"))
+                .addSuperinterface(KEYORI_ADVENTURE_KEY)
                 .addModifiers(Modifier.PUBLIC).addJavadoc("AUTOGENERATED by " + getClass().getSimpleName());
         villagerProfessionClass.addField(
-                FieldSpec.builder(namespaceIDClassName, "id")
+                FieldSpec.builder(NAMESPACE_ID_CLASS, "id")
                         .addModifiers(Modifier.PRIVATE, Modifier.FINAL).addAnnotation(NotNull.class).build()
         );
         villagerProfessionClass.addField(
-                FieldSpec.builder(rawVillagerProfessionDataClassName, "villagerProfessionData")
+                FieldSpec.builder(rawVillagerProfessionDataClassName, VILLAGER_PROFESSION_DATE)
                         .addModifiers(Modifier.PRIVATE, Modifier.VOLATILE)
                         .addAnnotation(NotNull.class)
                         .build()
         );
         villagerProfessionClass.addMethod(
                 MethodSpec.constructorBuilder()
-                        .addParameter(ParameterSpec.builder(namespaceIDClassName, "id").addAnnotation(NotNull.class).build())
-                        .addParameter(ParameterSpec.builder(rawVillagerProfessionDataClassName, "villagerProfessionData").addAnnotation(NotNull.class).build())
-                        .addStatement("this.id = id")
-                        .addStatement("this.villagerProfessionData = villagerProfessionData")
+                        .addParameter(ParameterSpec.builder(NAMESPACE_ID_CLASS, "id").addAnnotation(NotNull.class).build())
+                        .addParameter(ParameterSpec.builder(rawVillagerProfessionDataClassName, VILLAGER_PROFESSION_DATE).addAnnotation(NotNull.class).build())
+                        .addStatement(VARIABLE_SETTER, "id")
+                        .addStatement(VARIABLE_SETTER, VILLAGER_PROFESSION_DATE)
                         .addModifiers(Modifier.PROTECTED)
                         .build()
         );
         // Override key method (adventure)
-        villagerProfessionClass.addMethod(
-                MethodSpec.methodBuilder("key")
-                        .returns(ClassName.get("net.kyori.adventure.key", "Key"))
-                        .addAnnotation(Override.class)
-                        .addAnnotation(NotNull.class)
-                        .addStatement("return this.id")
-                        .addModifiers(Modifier.PUBLIC)
-                        .build()
-        );
+        villagerProfessionClass.addMethod(GenerationHelper.ADVENTURE_KEY_METHOD);
         // getId method
-        villagerProfessionClass.addMethod(
-                MethodSpec.methodBuilder("getId")
-                        .returns(namespaceIDClassName)
-                        .addAnnotation(NotNull.class)
-                        .addStatement("return this.id")
-                        .addModifiers(Modifier.PUBLIC)
-                        .build()
-        );
+        villagerProfessionClass.addMethod(GenerationHelper.ID_GETTER);
         // getVillagerProfessionData method
         villagerProfessionClass.addMethod(
                 MethodSpec.methodBuilder("getVillagerProfessionData")
@@ -100,8 +92,8 @@ public final class VillagerProfessionGenerator extends MinestomCodeGenerator {
         // setVillagerProfessionData method
         villagerProfessionClass.addMethod(
                 MethodSpec.methodBuilder("setVillagerProfessionData")
-                        .addParameter(ParameterSpec.builder(rawVillagerProfessionDataClassName, "villagerProfessionData").addAnnotation(NotNull.class).build())
-                        .addStatement("this.villagerProfessionData = villagerProfessionData")
+                        .addParameter(ParameterSpec.builder(rawVillagerProfessionDataClassName, VILLAGER_PROFESSION_DATE).addAnnotation(NotNull.class).build())
+                        .addStatement("this.$L1 = $L1", VILLAGER_PROFESSION_DATE)
                         .addModifiers(Modifier.PUBLIC, Modifier.FINAL)
                         .build()
         );
@@ -109,7 +101,7 @@ public final class VillagerProfessionGenerator extends MinestomCodeGenerator {
         villagerProfessionClass.addMethod(
                 MethodSpec.methodBuilder("getNumericalId")
                         .returns(TypeName.INT)
-                        .addStatement("return $T.VILLAGER_PROFESSION_REGISTRY.getId(this)", registryClassName)
+                        .addStatement("return $T.VILLAGER_PROFESSION_REGISTRY.getId(this)", REGISTRY_CLASS)
                         .addModifiers(Modifier.PUBLIC)
                         .build()
         );
@@ -119,7 +111,7 @@ public final class VillagerProfessionGenerator extends MinestomCodeGenerator {
                         .returns(villagerProfessionClassName)
                         .addAnnotation(Nullable.class)
                         .addParameter(TypeName.INT, "id")
-                        .addStatement("return $T.VILLAGER_PROFESSION_REGISTRY.get((short) id)", registryClassName)
+                        .addStatement("return $T.VILLAGER_PROFESSION_REGISTRY.get((short) id)", REGISTRY_CLASS)
                         .addModifiers(Modifier.PUBLIC, Modifier.STATIC)
                         .build()
         );
@@ -128,28 +120,19 @@ public final class VillagerProfessionGenerator extends MinestomCodeGenerator {
                 MethodSpec.methodBuilder("fromId")
                         .returns(villagerProfessionClassName)
                         .addAnnotation(NotNull.class)
-                        .addParameter(ClassName.get("net.kyori.adventure.key", "Key"), "id")
-                        .addStatement("return $T.VILLAGER_PROFESSION_REGISTRY.get(id)", registryClassName)
+                        .addParameter(ADVENTURE_KEY, "id")
+                        .addStatement("return $T.VILLAGER_PROFESSION_REGISTRY.get(id)", REGISTRY_CLASS)
                         .addModifiers(Modifier.PUBLIC, Modifier.STATIC)
                         .build()
         );
         // toString method
-        villagerProfessionClass.addMethod(
-                MethodSpec.methodBuilder("toString")
-                        .addAnnotation(NotNull.class)
-                        .addAnnotation(Override.class)
-                        .returns(String.class)
-                        // this resolves to [Namespace]
-                        .addStatement("return \"[\" + this.id + \"]\"")
-                        .addModifiers(Modifier.PUBLIC)
-                        .build()
-        );
+        villagerProfessionClass.addMethod(GenerationHelper.TO_STRING);
         // values method
         villagerProfessionClass.addMethod(
                 MethodSpec.methodBuilder("values")
                         .addAnnotation(NotNull.class)
                         .returns(ParameterizedTypeName.get(ClassName.get(List.class), villagerProfessionClassName))
-                        .addStatement("return $T.VILLAGER_PROFESSION_REGISTRY.values()", registryClassName)
+                        .addStatement("return $T.VILLAGER_PROFESSION_REGISTRY.values()", REGISTRY_CLASS)
                         .addModifiers(Modifier.PUBLIC, Modifier.STATIC)
                         .build()
         );
@@ -168,7 +151,7 @@ public final class VillagerProfessionGenerator extends MinestomCodeGenerator {
                         ).initializer(
                                 "new $T($T.from($S), new $T(() -> null))",
                                 villagerProfessionClassName,
-                                namespaceIDClassName,
+                                NAMESPACE_ID_CLASS,
                                 villagerProfession.get("id").getAsString(),
 
                                 rawVillagerProfessionDataClassName
@@ -182,18 +165,18 @@ public final class VillagerProfessionGenerator extends MinestomCodeGenerator {
                         ).initializer(
                                 "new $T($T.from($S), new $T(() -> $T.SOUND_EVENT_REGISTRY.get($S)))",
                                 villagerProfessionClassName,
-                                namespaceIDClassName,
+                                NAMESPACE_ID_CLASS,
                                 villagerProfession.get("id").getAsString(),
 
                                 rawVillagerProfessionDataClassName,
-                                registryClassName,
+                                REGISTRY_CLASS,
                                 workSound.getAsString()
                         ).addModifiers(Modifier.PUBLIC, Modifier.STATIC, Modifier.FINAL).build()
                 );
             }
 
             // Add to static init.
-            staticBlock.addStatement("$T.VILLAGER_PROFESSION_REGISTRY.register($N)", registryClassName, villagerProfessionName);
+            staticBlock.addStatement("$T.VILLAGER_PROFESSION_REGISTRY.register($N)", REGISTRY_CLASS, villagerProfessionName);
         }
 
         villagerProfessionClass.addStaticBlock(staticBlock.build());
@@ -202,7 +185,7 @@ public final class VillagerProfessionGenerator extends MinestomCodeGenerator {
         writeFiles(
                 List.of(
                         JavaFile.builder("net.minestom.server.entity.metadata.villager", villagerProfessionClass.build())
-                                .indent("    ")
+                                .indent(DEFAULT_INDENT)
                                 .skipJavaLangImports(true)
                                 .build()
                 ),
diff --git a/code-generators/src/main/java/net/minestom/codegen/entity/VillagerTypeGenerator.java b/code-generators/src/main/java/net/minestom/codegen/entity/VillagerTypeGenerator.java
index 8defb0243195fba6e0dee7a7b0aa22894bbf70a1..bc26041d017621effb5080d7e39a262bf1e7716a 100644
--- a/code-generators/src/main/java/net/minestom/codegen/entity/VillagerTypeGenerator.java
+++ b/code-generators/src/main/java/net/minestom/codegen/entity/VillagerTypeGenerator.java
@@ -5,6 +5,8 @@ import com.google.gson.JsonElement;
 import com.google.gson.JsonObject;
 import com.squareup.javapoet.*;
 import net.minestom.codegen.MinestomCodeGenerator;
+import net.minestom.codegen.util.GenerationHelper;
+import org.jetbrains.annotations.ApiStatus;
 import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
 import org.slf4j.Logger;
@@ -16,12 +18,18 @@ import java.io.InputStream;
 import java.io.InputStreamReader;
 import java.util.List;
 
+import static net.minestom.codegen.util.GenerationHelper.ADVENTURE_KEY;
+import static net.minestom.codegen.util.GenerationHelper.TO_STRING;
+
+@ApiStatus.NonExtendable
+@ApiStatus.Internal
 public final class VillagerTypeGenerator extends MinestomCodeGenerator {
     private static final Logger LOGGER = LoggerFactory.getLogger(VillagerTypeGenerator.class);
     private final InputStream villagerTypesFile;
     private final File outputFolder;
 
     public VillagerTypeGenerator(@Nullable InputStream villagerTypesFile, @NotNull File outputFolder) {
+        super("");
         this.villagerTypesFile = villagerTypesFile;
         this.outputFolder = outputFolder;
     }
@@ -38,41 +46,30 @@ public final class VillagerTypeGenerator extends MinestomCodeGenerator {
             return;
         }
         // Important classes we use alot
-        ClassName namespaceIDClassName = ClassName.get("net.minestom.server.utils", "NamespaceID");
-        ClassName registryClassName = ClassName.get("net.minestom.server.registry", "Registry");
-
         JsonArray villagerTypes = GSON.fromJson(new InputStreamReader(villagerTypesFile), JsonArray.class);
         ClassName villagerTypeClassName = ClassName.get("net.minestom.server.entity.metadata.villager", "VillagerType");
 
         // Particle
         TypeSpec.Builder villagerTypeClass = TypeSpec.classBuilder(villagerTypeClassName)
-                .addSuperinterface(ClassName.get("net.kyori.adventure.key", "Keyed"))
+                .addSuperinterface(KEYORI_ADVENTURE_KEY)
                 .addModifiers(Modifier.PUBLIC).addJavadoc("AUTOGENERATED by " + getClass().getSimpleName());
         villagerTypeClass.addField(
-                FieldSpec.builder(namespaceIDClassName, "id")
+                FieldSpec.builder(NAMESPACE_ID_CLASS, "id")
                         .addModifiers(Modifier.PRIVATE, Modifier.FINAL).addAnnotation(NotNull.class).build()
         );
         villagerTypeClass.addMethod(
                 MethodSpec.constructorBuilder()
-                        .addParameter(ParameterSpec.builder(namespaceIDClassName, "id").addAnnotation(NotNull.class).build())
+                        .addParameter(ParameterSpec.builder(NAMESPACE_ID_CLASS, "id").addAnnotation(NotNull.class).build())
                         .addStatement("this.id = id")
                         .addModifiers(Modifier.PROTECTED)
                         .build()
         );
         // Override key method (adventure)
-        villagerTypeClass.addMethod(
-                MethodSpec.methodBuilder("key")
-                        .returns(ClassName.get("net.kyori.adventure.key", "Key"))
-                        .addAnnotation(Override.class)
-                        .addAnnotation(NotNull.class)
-                        .addStatement("return this.id")
-                        .addModifiers(Modifier.PUBLIC)
-                        .build()
-        );
+        villagerTypeClass.addMethod(GenerationHelper.ID_GETTER);
         // getId method
         villagerTypeClass.addMethod(
                 MethodSpec.methodBuilder("getId")
-                        .returns(namespaceIDClassName)
+                        .returns(NAMESPACE_ID_CLASS)
                         .addAnnotation(NotNull.class)
                         .addStatement("return this.id")
                         .addModifiers(Modifier.PUBLIC)
@@ -82,7 +79,7 @@ public final class VillagerTypeGenerator extends MinestomCodeGenerator {
         villagerTypeClass.addMethod(
                 MethodSpec.methodBuilder("getNumericalId")
                         .returns(TypeName.INT)
-                        .addStatement("return $T.VILLAGER_TYPE_REGISTRY.getId(this)", registryClassName)
+                        .addStatement("return $T.VILLAGER_TYPE_REGISTRY.getId(this)", REGISTRY_CLASS)
                         .addModifiers(Modifier.PUBLIC)
                         .build()
         );
@@ -92,7 +89,7 @@ public final class VillagerTypeGenerator extends MinestomCodeGenerator {
                         .returns(villagerTypeClassName)
                         .addAnnotation(Nullable.class)
                         .addParameter(TypeName.INT, "id")
-                        .addStatement("return $T.VILLAGER_TYPE_REGISTRY.get((short) id)", registryClassName)
+                        .addStatement("return $T.VILLAGER_TYPE_REGISTRY.get((short) id)", REGISTRY_CLASS)
                         .addModifiers(Modifier.PUBLIC, Modifier.STATIC)
                         .build()
         );
@@ -101,28 +98,19 @@ public final class VillagerTypeGenerator extends MinestomCodeGenerator {
                 MethodSpec.methodBuilder("fromId")
                         .returns(villagerTypeClassName)
                         .addAnnotation(NotNull.class)
-                        .addParameter(ClassName.get("net.kyori.adventure.key", "Key"), "id")
-                        .addStatement("return $T.VILLAGER_TYPE_REGISTRY.get(id)", registryClassName)
+                        .addParameter(ADVENTURE_KEY, "id")
+                        .addStatement("return $T.VILLAGER_TYPE_REGISTRY.get(id)", REGISTRY_CLASS)
                         .addModifiers(Modifier.PUBLIC, Modifier.STATIC)
                         .build()
         );
         // toString method
-        villagerTypeClass.addMethod(
-                MethodSpec.methodBuilder("toString")
-                        .addAnnotation(NotNull.class)
-                        .addAnnotation(Override.class)
-                        .returns(String.class)
-                        // this resolves to [Namespace]
-                        .addStatement("return \"[\" + this.id + \"]\"")
-                        .addModifiers(Modifier.PUBLIC)
-                        .build()
-        );
+        villagerTypeClass.addMethod(TO_STRING);
         // values method
         villagerTypeClass.addMethod(
                 MethodSpec.methodBuilder("values")
                         .addAnnotation(NotNull.class)
                         .returns(ParameterizedTypeName.get(ClassName.get(List.class), villagerTypeClassName))
-                        .addStatement("return $T.VILLAGER_TYPE_REGISTRY.values()", registryClassName)
+                        .addStatement("return $T.VILLAGER_TYPE_REGISTRY.values()", REGISTRY_CLASS)
                         .addModifiers(Modifier.PUBLIC, Modifier.STATIC)
                         .build()
         );
@@ -140,12 +128,12 @@ public final class VillagerTypeGenerator extends MinestomCodeGenerator {
                     ).initializer(
                             "new $T($T.from($S))",
                             villagerTypeClassName,
-                            namespaceIDClassName,
+                            NAMESPACE_ID_CLASS,
                             villagerProfession.get("id").getAsString()
                     ).addModifiers(Modifier.PUBLIC, Modifier.STATIC, Modifier.FINAL).build()
             );
             // Add to static init.
-            staticBlock.addStatement("$T.VILLAGER_TYPE_REGISTRY.register($N)", registryClassName, villagerProfessionName);
+            staticBlock.addStatement("$T.VILLAGER_TYPE_REGISTRY.register($N)", REGISTRY_CLASS, villagerProfessionName);
         }
 
         villagerTypeClass.addStaticBlock(staticBlock.build());
@@ -154,7 +142,7 @@ public final class VillagerTypeGenerator extends MinestomCodeGenerator {
         writeFiles(
                 List.of(
                         JavaFile.builder("net.minestom.server.entity.metadata.villager", villagerTypeClass.build())
-                                .indent("    ")
+                                .indent(DEFAULT_INDENT)
                                 .skipJavaLangImports(true)
                                 .build()
                 ),
diff --git a/code-generators/src/main/java/net/minestom/codegen/fluid/FluidGenerator.java b/code-generators/src/main/java/net/minestom/codegen/fluid/FluidGenerator.java
index 29012054b8d8fbcc21c6a76966c74a6b1d4292e6..3968e9ec6eb833bf77f9fc3fcf662eb7b6e6110c 100644
--- a/code-generators/src/main/java/net/minestom/codegen/fluid/FluidGenerator.java
+++ b/code-generators/src/main/java/net/minestom/codegen/fluid/FluidGenerator.java
@@ -3,23 +3,27 @@ package net.minestom.codegen.fluid;
 import com.google.gson.JsonObject;
 import com.squareup.javapoet.*;
 import net.minestom.codegen.MinestomCodeGenerator;
+import org.jetbrains.annotations.ApiStatus;
 import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
+import net.minestom.codegen.util.GenerationHelper;
 
 import javax.lang.model.element.Modifier;
 import java.io.File;
 import java.io.InputStream;
 import java.io.InputStreamReader;
-import java.util.List;
 
+@ApiStatus.NonExtendable
+@ApiStatus.Internal
 public final class FluidGenerator extends MinestomCodeGenerator {
     private static final Logger LOGGER = LoggerFactory.getLogger(FluidGenerator.class);
     private final InputStream fluidsFile;
     private final File outputFolder;
 
     public FluidGenerator(@Nullable InputStream fluidsFile, @NotNull File outputFolder) {
+        super("net.minestom.server.fluid");
         this.fluidsFile = fluidsFile;
         this.outputFolder = outputFolder;
     }
@@ -35,47 +39,36 @@ public final class FluidGenerator extends MinestomCodeGenerator {
             LOGGER.error("Output folder for code generation does not exist and could not be created.");
             return;
         }
-        // Important classes we use alot
-        ClassName namespaceIDClassName = ClassName.get("net.minestom.server.utils", "NamespaceID");
-        ClassName registriesClassName = ClassName.get("net.minestom.server.registry", "Registries");
 
         JsonObject fluids = GSON.fromJson(new InputStreamReader(fluidsFile), JsonObject.class);
-        ClassName fluidClassName = ClassName.get("net.minestom.server.fluid", "Fluid");
+        ClassName fluidClassName = ClassName.get(packageName, "Fluid");
 
         // Particle
         TypeSpec.Builder fluidClass = TypeSpec.enumBuilder(fluidClassName)
-                .addSuperinterface(ClassName.get("net.kyori.adventure.key", "Keyed"))
+                .addSuperinterface(KEYORI_ADVENTURE_KEY)
                 .addModifiers(Modifier.PUBLIC).addJavadoc("AUTOGENERATED by " + getClass().getSimpleName());
 
         fluidClass.addField(
-                FieldSpec.builder(namespaceIDClassName, "id")
-                        .addModifiers(Modifier.PRIVATE, Modifier.FINAL).addAnnotation(NotNull.class).build()
+                FieldSpec.builder(NAMESPACE_ID_CLASS, "id")
+                        .addModifiers(PRIVATE_FINAL_MODIFIERS).addAnnotation(NotNull.class).build()
         );
         // static field
         fluidClass.addField(
                 FieldSpec.builder(ArrayTypeName.of(fluidClassName), "VALUES")
-                        .addModifiers(Modifier.PRIVATE, Modifier.STATIC, Modifier.FINAL)
+                        .addModifiers(CONSTANT_MODIFIERS)
                         .initializer("values()")
                         .build()
         );
 
         fluidClass.addMethod(
                 MethodSpec.constructorBuilder()
-                        .addParameter(ParameterSpec.builder(namespaceIDClassName, "id").addAnnotation(NotNull.class).build())
+                        .addParameter(ParameterSpec.builder(NAMESPACE_ID_CLASS, "id").addAnnotation(NotNull.class).build())
                         .addStatement("this.id = id")
-                        .addStatement("$T.fluids.put(id, this)", registriesClassName)
+                        .addStatement("$T.fluids.put(id, this)", REGISTRY_CLASS)
                         .build()
         );
         // Override key method (adventure)
-        fluidClass.addMethod(
-                MethodSpec.methodBuilder("key")
-                        .returns(ClassName.get("net.kyori.adventure.key", "Key"))
-                        .addAnnotation(Override.class)
-                        .addAnnotation(NotNull.class)
-                        .addStatement("return this.id")
-                        .addModifiers(Modifier.PUBLIC)
-                        .build()
-        );
+        fluidClass.addMethod(GenerationHelper.ADVENTURE_KEY_METHOD);
         // getId method
         fluidClass.addMethod(
                 MethodSpec.methodBuilder("getId")
@@ -87,23 +80,14 @@ public final class FluidGenerator extends MinestomCodeGenerator {
         // getNamespaceID method
         fluidClass.addMethod(
                 MethodSpec.methodBuilder("getNamespaceID")
-                        .returns(namespaceIDClassName)
+                        .returns(NAMESPACE_ID_CLASS)
                         .addAnnotation(NotNull.class)
                         .addStatement("return this.id")
                         .addModifiers(Modifier.PUBLIC)
                         .build()
         );
         // toString method
-        fluidClass.addMethod(
-                MethodSpec.methodBuilder("toString")
-                        .addAnnotation(NotNull.class)
-                        .addAnnotation(Override.class)
-                        .returns(String.class)
-                        // this resolves to [Namespace]
-                        .addStatement("return \"[\" + this.id + \"]\"")
-                        .addModifiers(Modifier.PUBLIC)
-                        .build()
-        );
+        fluidClass.addMethod(GenerationHelper.TO_STRING);
 
         // fromId Method
         fluidClass.addMethod(
@@ -122,22 +106,22 @@ public final class FluidGenerator extends MinestomCodeGenerator {
         // Use data
         fluids.entrySet().forEach(entry -> {
             final String fluidName = entry.getKey();
-            fluidClass.addEnumConstant(toConstant(fluidName), TypeSpec.anonymousClassBuilder(
+            fluidClass.addEnumConstant(
+                    extractNamespace(fluidName),
+                    TypeSpec.anonymousClassBuilder(
                             "$T.from($S)",
-                            namespaceIDClassName,
+                            NAMESPACE_ID_CLASS,
                             fluidName
                     ).build()
             );
         });
 
         // Write files to outputFolder
-        writeFiles(
-                List.of(
-                        JavaFile.builder("net.minestom.server.fluid", fluidClass.build())
-                                .indent("    ")
-                                .skipJavaLangImports(true)
-                                .build()
-                ),
+        writeFile(
+                JavaFile.builder(packageName, fluidClass.build())
+                        .indent(DEFAULT_INDENT)
+                        .skipJavaLangImports(true)
+                        .build(),
                 outputFolder
         );
     }
diff --git a/code-generators/src/main/java/net/minestom/codegen/util/GenerationHelper.java b/code-generators/src/main/java/net/minestom/codegen/util/GenerationHelper.java
new file mode 100644
index 0000000000000000000000000000000000000000..f56116180a47a65d4bde378e87cb09a09d807e80
--- /dev/null
+++ b/code-generators/src/main/java/net/minestom/codegen/util/GenerationHelper.java
@@ -0,0 +1,47 @@
+package net.minestom.codegen.util;
+
+import com.squareup.javapoet.ClassName;
+import com.squareup.javapoet.MethodSpec;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.NotNull;
+
+import javax.lang.model.element.Modifier;
+
+import static net.minestom.codegen.MinestomCodeGenerator.NAMESPACE_ID_CLASS;
+
+@ApiStatus.Internal
+@ApiStatus.NonExtendable
+public final class GenerationHelper {
+
+    public static final ClassName ADVENTURE_KEY = ClassName.get("net.kyori.adventure.key", "Key");
+    public static final MethodSpec ID_GETTER;
+    public static final MethodSpec ADVENTURE_KEY_METHOD;
+    public static final MethodSpec TO_STRING;
+    public static final String VARIABLE_SETTER = "this.$1L = $1L";
+
+    static {
+        ID_GETTER = MethodSpec.methodBuilder("getId")
+                .returns(NAMESPACE_ID_CLASS)
+                .addAnnotation(NotNull.class)
+                .addStatement("return this.id")
+                .addModifiers(Modifier.PUBLIC)
+                .build();
+        ADVENTURE_KEY_METHOD = MethodSpec.methodBuilder("key")
+                .returns(ADVENTURE_KEY)
+                .addAnnotation(Override.class)
+                .addAnnotation(NotNull.class)
+                .addStatement("return this.id")
+                .addModifiers(Modifier.PUBLIC)
+                .build();
+        TO_STRING = MethodSpec.methodBuilder("toString")
+                .addAnnotation(NotNull.class)
+                .addAnnotation(Override.class)
+                .returns(String.class)
+                // this resolves to [Namespace]
+                .addStatement("return \"[\" + this.id + \"]\"")
+                .addModifiers(Modifier.PUBLIC)
+                .build();
+    }
+
+    private GenerationHelper() { }
+}
